{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "f3u-aY1jMqDF"
      },
      "outputs": [],
      "source": [
        "!pip install -q google-cloud-storage"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import auth\n",
        "auth.authenticate_user()"
      ],
      "metadata": {
        "id": "ui1PgUfSRZJH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "!sudo apt-get install samtools\n",
        "!sudo apt-get install bwa\n",
        "!sudo apt-get update\n",
        "!sudo apt-get install openjdk-17-jdk"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OQHo9YROHxIr",
        "outputId": "2e45e1ca-45ea-4588-b3f0-fa37765cb6fb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libhts3 libhtscodecs2\n",
            "Suggested packages:\n",
            "  cwltool\n",
            "The following NEW packages will be installed:\n",
            "  libhts3 libhtscodecs2 samtools\n",
            "0 upgraded, 3 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 963 kB of archives.\n",
            "After this operation, 2,270 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
            "Fetched 963 kB in 0s (2,175 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 3.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "(Reading database ... 123630 files and directories currently installed.)\n",
            "Preparing to unpack .../libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../samtools_1.13-4_amd64.deb ...\n",
            "Unpacking samtools (1.13-4) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up samtools (1.13-4) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  bwa\n",
            "0 upgraded, 1 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 195 kB of archives.\n",
            "After this operation, 466 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bwa amd64 0.7.17-6 [195 kB]\n",
            "Fetched 195 kB in 0s (896 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 1.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package bwa.\n",
            "(Reading database ... 123729 files and directories currently installed.)\n",
            "Preparing to unpack .../bwa_0.7.17-6_amd64.deb ...\n",
            "Unpacking bwa (0.7.17-6) ...\n",
            "Setting up bwa (0.7.17-6) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Hit:1 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Get:2 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "Get:4 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]\n",
            "Get:5 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]\n",
            "Get:6 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n",
            "Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Get:9 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [1,172 kB]\n",
            "Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Hit:11 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Get:12 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [3,323 kB]\n",
            "Get:13 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,452 kB]\n",
            "Get:14 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,223 kB]\n",
            "Get:15 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [8,498 kB]\n",
            "Get:16 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,613 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [3,413 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,512 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [2,734 kB]\n",
            "Fetched 27.3 MB in 5s (5,672 kB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  fonts-dejavu-core fonts-dejavu-extra libatk-wrapper-java\n",
            "  libatk-wrapper-java-jni libfontenc1 libice-dev libsm-dev libxkbfile1\n",
            "  libxt-dev libxtst6 libxxf86dga1 openjdk-17-jdk-headless openjdk-17-jre\n",
            "  openjdk-17-jre-headless x11-utils\n",
            "Suggested packages:\n",
            "  libice-doc libsm-doc libxt-doc openjdk-17-demo openjdk-17-source visualvm\n",
            "  libnss-mdns fonts-ipafont-gothic fonts-ipafont-mincho fonts-wqy-microhei\n",
            "  | fonts-wqy-zenhei fonts-indic mesa-utils\n",
            "The following NEW packages will be installed:\n",
            "  fonts-dejavu-core fonts-dejavu-extra libatk-wrapper-java\n",
            "  libatk-wrapper-java-jni libfontenc1 libice-dev libsm-dev libxkbfile1\n",
            "  libxt-dev libxtst6 libxxf86dga1 openjdk-17-jdk openjdk-17-jdk-headless\n",
            "  openjdk-17-jre openjdk-17-jre-headless x11-utils\n",
            "0 upgraded, 16 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 126 MB of archives.\n",
            "After this operation, 289 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-dejavu-core all 2.37-2build1 [1,041 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-dejavu-extra all 2.37-2build1 [2,041 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/main amd64 libfontenc1 amd64 1:1.1.4-1build3 [14.7 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxkbfile1 amd64 1:1.1.0-1build3 [71.8 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxtst6 amd64 2:1.2.3-1build4 [13.4 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxxf86dga1 amd64 2:1.1.5-0ubuntu3 [12.6 kB]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu jammy/main amd64 x11-utils amd64 7.7+5build2 [206 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu jammy/main amd64 libatk-wrapper-java all 0.38.0-5build1 [53.1 kB]\n",
            "Get:9 http://archive.ubuntu.com/ubuntu jammy/main amd64 libatk-wrapper-java-jni amd64 0.38.0-5build1 [49.0 kB]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu jammy/main amd64 libice-dev amd64 2:1.0.10-1build2 [51.4 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu jammy/main amd64 libsm-dev amd64 2:1.2.3-1build2 [18.1 kB]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxt-dev amd64 1:1.2.1-1 [396 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 openjdk-17-jre-headless amd64 17.0.13+11-2ubuntu1~22.04 [48.4 MB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 openjdk-17-jre amd64 17.0.13+11-2ubuntu1~22.04 [234 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 openjdk-17-jdk-headless amd64 17.0.13+11-2ubuntu1~22.04 [71.4 MB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 openjdk-17-jdk amd64 17.0.13+11-2ubuntu1~22.04 [2,366 kB]\n",
            "Fetched 126 MB in 2s (58.4 MB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 16.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package fonts-dejavu-core.\n",
            "(Reading database ... 123746 files and directories currently installed.)\n",
            "Preparing to unpack .../00-fonts-dejavu-core_2.37-2build1_all.deb ...\n",
            "Unpacking fonts-dejavu-core (2.37-2build1) ...\n",
            "Selecting previously unselected package fonts-dejavu-extra.\n",
            "Preparing to unpack .../01-fonts-dejavu-extra_2.37-2build1_all.deb ...\n",
            "Unpacking fonts-dejavu-extra (2.37-2build1) ...\n",
            "Selecting previously unselected package libfontenc1:amd64.\n",
            "Preparing to unpack .../02-libfontenc1_1%3a1.1.4-1build3_amd64.deb ...\n",
            "Unpacking libfontenc1:amd64 (1:1.1.4-1build3) ...\n",
            "Selecting previously unselected package libxkbfile1:amd64.\n",
            "Preparing to unpack .../03-libxkbfile1_1%3a1.1.0-1build3_amd64.deb ...\n",
            "Unpacking libxkbfile1:amd64 (1:1.1.0-1build3) ...\n",
            "Selecting previously unselected package libxtst6:amd64.\n",
            "Preparing to unpack .../04-libxtst6_2%3a1.2.3-1build4_amd64.deb ...\n",
            "Unpacking libxtst6:amd64 (2:1.2.3-1build4) ...\n",
            "Selecting previously unselected package libxxf86dga1:amd64.\n",
            "Preparing to unpack .../05-libxxf86dga1_2%3a1.1.5-0ubuntu3_amd64.deb ...\n",
            "Unpacking libxxf86dga1:amd64 (2:1.1.5-0ubuntu3) ...\n",
            "Selecting previously unselected package x11-utils.\n",
            "Preparing to unpack .../06-x11-utils_7.7+5build2_amd64.deb ...\n",
            "Unpacking x11-utils (7.7+5build2) ...\n",
            "Selecting previously unselected package libatk-wrapper-java.\n",
            "Preparing to unpack .../07-libatk-wrapper-java_0.38.0-5build1_all.deb ...\n",
            "Unpacking libatk-wrapper-java (0.38.0-5build1) ...\n",
            "Selecting previously unselected package libatk-wrapper-java-jni:amd64.\n",
            "Preparing to unpack .../08-libatk-wrapper-java-jni_0.38.0-5build1_amd64.deb ...\n",
            "Unpacking libatk-wrapper-java-jni:amd64 (0.38.0-5build1) ...\n",
            "Selecting previously unselected package libice-dev:amd64.\n",
            "Preparing to unpack .../09-libice-dev_2%3a1.0.10-1build2_amd64.deb ...\n",
            "Unpacking libice-dev:amd64 (2:1.0.10-1build2) ...\n",
            "Selecting previously unselected package libsm-dev:amd64.\n",
            "Preparing to unpack .../10-libsm-dev_2%3a1.2.3-1build2_amd64.deb ...\n",
            "Unpacking libsm-dev:amd64 (2:1.2.3-1build2) ...\n",
            "Selecting previously unselected package libxt-dev:amd64.\n",
            "Preparing to unpack .../11-libxt-dev_1%3a1.2.1-1_amd64.deb ...\n",
            "Unpacking libxt-dev:amd64 (1:1.2.1-1) ...\n",
            "Selecting previously unselected package openjdk-17-jre-headless:amd64.\n",
            "Preparing to unpack .../12-openjdk-17-jre-headless_17.0.13+11-2ubuntu1~22.04_amd64.deb ...\n",
            "Unpacking openjdk-17-jre-headless:amd64 (17.0.13+11-2ubuntu1~22.04) ...\n",
            "Selecting previously unselected package openjdk-17-jre:amd64.\n",
            "Preparing to unpack .../13-openjdk-17-jre_17.0.13+11-2ubuntu1~22.04_amd64.deb ...\n",
            "Unpacking openjdk-17-jre:amd64 (17.0.13+11-2ubuntu1~22.04) ...\n",
            "Selecting previously unselected package openjdk-17-jdk-headless:amd64.\n",
            "Preparing to unpack .../14-openjdk-17-jdk-headless_17.0.13+11-2ubuntu1~22.04_amd64.deb ...\n",
            "Unpacking openjdk-17-jdk-headless:amd64 (17.0.13+11-2ubuntu1~22.04) ...\n",
            "Selecting previously unselected package openjdk-17-jdk:amd64.\n",
            "Preparing to unpack .../15-openjdk-17-jdk_17.0.13+11-2ubuntu1~22.04_amd64.deb ...\n",
            "Unpacking openjdk-17-jdk:amd64 (17.0.13+11-2ubuntu1~22.04) ...\n",
            "Setting up libice-dev:amd64 (2:1.0.10-1build2) ...\n",
            "Setting up libsm-dev:amd64 (2:1.2.3-1build2) ...\n",
            "Setting up libxtst6:amd64 (2:1.2.3-1build4) ...\n",
            "Setting up libxxf86dga1:amd64 (2:1.1.5-0ubuntu3) ...\n",
            "Setting up libfontenc1:amd64 (1:1.1.4-1build3) ...\n",
            "Setting up libxt-dev:amd64 (1:1.2.1-1) ...\n",
            "Setting up fonts-dejavu-core (2.37-2build1) ...\n",
            "Setting up openjdk-17-jre-headless:amd64 (17.0.13+11-2ubuntu1~22.04) ...\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/java to provide /usr/bin/java (java) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jpackage to provide /usr/bin/jpackage (jpackage) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/keytool to provide /usr/bin/keytool (keytool) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/rmiregistry to provide /usr/bin/rmiregistry (rmiregistry) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/lib/jexec to provide /usr/bin/jexec (jexec) in auto mode\n",
            "Setting up fonts-dejavu-extra (2.37-2build1) ...\n",
            "Setting up libxkbfile1:amd64 (1:1.1.0-1build3) ...\n",
            "Setting up openjdk-17-jre:amd64 (17.0.13+11-2ubuntu1~22.04) ...\n",
            "Setting up openjdk-17-jdk-headless:amd64 (17.0.13+11-2ubuntu1~22.04) ...\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jar to provide /usr/bin/jar (jar) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jarsigner to provide /usr/bin/jarsigner (jarsigner) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/javac to provide /usr/bin/javac (javac) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/javadoc to provide /usr/bin/javadoc (javadoc) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/javap to provide /usr/bin/javap (javap) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jcmd to provide /usr/bin/jcmd (jcmd) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jdb to provide /usr/bin/jdb (jdb) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jdeprscan to provide /usr/bin/jdeprscan (jdeprscan) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jdeps to provide /usr/bin/jdeps (jdeps) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jfr to provide /usr/bin/jfr (jfr) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jimage to provide /usr/bin/jimage (jimage) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jinfo to provide /usr/bin/jinfo (jinfo) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jlink to provide /usr/bin/jlink (jlink) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jmap to provide /usr/bin/jmap (jmap) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jmod to provide /usr/bin/jmod (jmod) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jps to provide /usr/bin/jps (jps) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jrunscript to provide /usr/bin/jrunscript (jrunscript) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jshell to provide /usr/bin/jshell (jshell) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jstack to provide /usr/bin/jstack (jstack) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jstat to provide /usr/bin/jstat (jstat) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jstatd to provide /usr/bin/jstatd (jstatd) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/serialver to provide /usr/bin/serialver (serialver) in auto mode\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jhsdb to provide /usr/bin/jhsdb (jhsdb) in auto mode\n",
            "Setting up x11-utils (7.7+5build2) ...\n",
            "Setting up libatk-wrapper-java (0.38.0-5build1) ...\n",
            "Setting up openjdk-17-jdk:amd64 (17.0.13+11-2ubuntu1~22.04) ...\n",
            "update-alternatives: using /usr/lib/jvm/java-17-openjdk-amd64/bin/jconsole to provide /usr/bin/jconsole (jconsole) in auto mode\n",
            "Setting up libatk-wrapper-java-jni:amd64 (0.38.0-5build1) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!chmod +x /content/drive/MyDrive/GenomeAnalysis/tools/samtools-1.18/samtools"
      ],
      "metadata": {
        "id": "PkZDqaCsOsCg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "os.environ['DATA'] = \"/content/drive/MyDrive/GenomeAnalysis/data\"\n",
        "os.environ['TOOLS'] = \"/content/drive/MyDrive/GenomeAnalysis/tools\""
      ],
      "metadata": {
        "id": "K3oAwmDXRuQk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!gsutil ls gs://gatk-tutorials/workshop_2002/2-germline/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ae-H2sWUSid4",
        "outputId": "52f10a59-2729-4200-9bbc-47090ee547aa"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gs://gatk-tutorials/workshop_2002/2-germline/trio.ped\n",
            "gs://gatk-tutorials/workshop_2002/2-germline/CNNScoreVariants/\n",
            "gs://gatk-tutorials/workshop_2002/2-germline/bams/\n",
            "gs://gatk-tutorials/workshop_2002/2-germline/gvcfs/\n",
            "gs://gatk-tutorials/workshop_2002/2-germline/illumina_platinum/\n",
            "gs://gatk-tutorials/workshop_2002/2-germline/intervals/\n",
            "gs://gatk-tutorials/workshop_2002/2-germline/ref/\n",
            "gs://gatk-tutorials/workshop_2002/2-germline/resources/\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gsutil ls gs://gatk-test-data/wgs_fastq/NA12878_20k/H06HDADXX130110.1.ATCACGAT.20k_reads_1.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Skl4ucSTeBQz",
        "outputId": "b3f6b11b-a73a-49ec-e4a9-54e65384283f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gs://gatk-test-data/wgs_fastq/NA12878_20k/H06HDADXX130110.1.ATCACGAT.20k_reads_1.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gsutil cp gs://gatk-tutorials/workshop_2002/2-germline/trio.ped ${DATA}/resources\n",
        "!gsutil cp gs://gatk-tutorials/workshop_2002/2-germline/resources/* ${DATA}/resources\n",
        "!gsutil cp gs://gatk-tutorials/workshop_2002/2-germline/bams/* ${DATA}/bams"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lXxZJBztSzWG",
        "outputId": "3d1ad3e6-ccfd-43aa-bedb-8a36c66f7f1b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/trio.ped...\n",
            "/ [1 files][  181.0 B/  181.0 B]                                                \n",
            "Operation completed over 1 objects/181.0 B.                                      \n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/resources/af-only-gnomad.chr20subset.b37.vcf.gz...\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/resources/af-only-gnomad.chr20subset.b37.vcf.gz.tbi...\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/resources/dbsnp.vcf...\n",
            "==> NOTE: You are downloading one or more large file(s), which would\n",
            "run significantly faster if you enabled sliced object downloads. This\n",
            "feature is enabled by default but requires that compiled crcmod be\n",
            "installed (see \"gsutil help crcmod\").\n",
            "\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/resources/dbsnp.vcf.idx...\n",
            "/ [4 files][230.4 MiB/230.4 MiB]   24.8 MiB/s                                   \n",
            "==> NOTE: You are performing a sequence of gsutil operations that may\n",
            "run significantly faster if you instead use gsutil -m cp ... Please\n",
            "see the -m section under \"gsutil help options\" for further information\n",
            "about when gsutil -m can be advantageous.\n",
            "\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/resources/motherGIAB.vcf.gz...\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/resources/motherGIAB.vcf.gz.tbi...\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/resources/motherGIABsnps.vcf.gz...\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/resources/motherGIABsnps.vcf.gz.tbi...\n",
            "/ [8 files][235.2 MiB/235.2 MiB]    2.3 MiB/s                                   \n",
            "Operation completed over 8 objects/235.2 MiB.                                    \n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/bams/father.bai...\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/bams/father.bam...\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/bams/mother.bai...\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/bams/mother.bam...\n",
            "| [4 files][ 33.5 MiB/ 33.5 MiB]                                                \n",
            "==> NOTE: You are performing a sequence of gsutil operations that may\n",
            "run significantly faster if you instead use gsutil -m cp ... Please\n",
            "see the -m section under \"gsutil help options\" for further information\n",
            "about when gsutil -m can be advantageous.\n",
            "\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/bams/son.bai...\n",
            "Copying gs://gatk-tutorials/workshop_2002/2-germline/bams/son.bam...\n",
            "\\ [6 files][ 43.0 MiB/ 43.0 MiB]                                                \n",
            "Operation completed over 6 objects/43.0 MiB.                                     \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!export OutputFolder=${DATA}/variant-discovery"
      ],
      "metadata": {
        "id": "ZlS2fCP5V0T7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/samtools-1.18/samtools view ${DATA}/bams/mother.bam | head -n 5"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gKjCV-U8WDuA",
        "outputId": "5ec590cc-1e0b-4fdc-c534-5c0a6e7de525"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "H0164ALXX140820:2:1108:8135:38509\t99\t20\t9999879\t60\t118M2I31M\t=\t10000037\t296\tCCCATTAGATGGTCTTGAAACGTTGTTAAAATTCAACTGACCATAGGTGTATTGGTTTATTTCTGTACTCTTAGTAGATTCCATTGACCTATATCTCTATCCTTATGCCAGTACCACACTGTTTTGTTTACTACAGCTTTGTAGTAAATTT\t===>?<?>=><>>>?=>>>=>8>=>>==><=>==>==>====;=>>>9>>?>;>>>====<<=>==?>?=><==>?>=><=<>=<>>>>@?><=>>>?@?=?@>@??@?@:?@<?@?@?<>??<?9@>?A@AA@<AAB;?;AB@<A<=<:;\tMC:Z:13S138M\tMD:Z:149\tPG:Z:MarkDuplicates\tRG:Z:H0164.2\tNM:i:2\tMQ:i:60\tOQ:Z:FFFFFAJFAJFFJJJFJJJFJJJJJJJFJFJJJJJJJJJJJJFJJJJAJJJJFJJJJJFJJJJJJJJJJJJJFFJJJFJJJJJJJJJJJJJJFFJJJJJJFJJJJJJJJJAJJFJJJJJAFJJFJAJFJJJJJJAJJJAFAJJFAJAJJJJ\tUQ:i:0\tAS:i:141\n",
            "H0164ALXX140820:2:1108:8247:38350\t1123\t20\t9999879\t60\t118M2I31M\t=\t10000037\t296\tCCCATTAGATGGTCTTGAAACGTTGTTAAAATTCAACTGACCATAGGTGTATTGGTTTATTTCTGTACTCTTAGTAGATTCCATTGACCTATATCTCTATCCTTATGCCAGTACCACACTGTTTTGTTTACTACAGCTTTGTAGTAAATTT\t===>?>??>><>>>@=>>>>>6>=>>=?>==>==>==<====;=>><>>>?>=>>===?=<<=>==?>?=><=?>?>>><=<;=<;>>>?<>?>>>>?:???@>@??@?@@<@????@?@?:?<>>>??;=A?@?AA=?@AABBA=/;;;;\tMC:Z:13S138M\tMD:Z:149\tPG:Z:MarkDuplicates\tRG:Z:H0164.2\tNM:i:2\tMQ:i:60\tOQ:Z:FFFFFFJJJFFJJFFJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJFJJJJJJJJFFJJJJJJJJJJJJJJJFJJJJJJJJJFJJFJJJFFJJJJJJJAJJJJJJJJJJJJFJJJFJJJJJAJFFFFJJAFJFJFJJ<JJJJJJJA-AAAJ\tUQ:i:0\tAS:i:141\n",
            "H0164ALXX140820:2:1223:12347:21719\t163\t20\t9999896\t60\t101M2I48M\t=\t10000096\t351\tAAACGTTGTTAAAATTCAACTGACCATAGGTGTATTGGTTTATTTCTGTACTCTTAGTAGATTCCATTGACCTATATCTCTATCCTTATGCCAGTACCACACTGTTTTGTTTACTACAGCTTTGTAGTAAATTTTGAATTCTAAAGTGTTA\t;1;;6=====<=;7=<=>:=?==5=-,;>=.==2;8>>77:;>==>?>>=<>>?==?9?=<=>?9@=87?*??=?==?<=@>??;@>>?+@=6@=4?;?=?@@:5>??@@?>>>A@=??>98>9?,:?@@.=@><:8/-><B-@/8>8;:9\tMC:Z:151M\tMD:Z:136C12\tPG:Z:MarkDuplicates\tRG:Z:H0164.2\tNM:i:3\tMQ:i:60\tOQ:Z:A-FFFFFJFJFJA<JJJJAJFJJ<J--AJJ-FF7A<JJ<<AAJJJJJJJFFFJJJFJ<JFFFJJ<JF<<J-JFFJFFJ<FJFJJAFJFJ-JF7JF7JAFFFJFA7FJJJJFFFFJJJFFF7<FAJ-AFFF-AFFAAF--FAF-F-<F<FAA\tUQ:i:12\tAS:i:136\n",
            "H0164ALXX140820:2:1222:11840:58673\t99\t20\t9999905\t60\t92M2I57M\t=\t10000151\t397\tTAAAATTCAACTGACCATAGGTGTATTGGTTTATTTCTGTACTCTTAGTAGATTCCATTGACCTATATCTCTATCCTTATGCCAGTACCACACTGTTTTGTTTACTACAGCTTTGTAGTAAATTTTGAACTCTAAAGTGTTAGTTCTCTAC\t====;?>>?>7@>>>=?>>?;=>>?>=>><8=?<<<=>==>=9=><?8>??>>8>>>;=====>;=?><>=>?=<>9=<==7:>>>?;,@>?>?<===>59>>@?@>?@@;>=:?;3=:::-@>>?+@@=@@A?>>;:??9=?;?=@==<#\tMC:Z:151M\tMD:Z:148A0\tPG:Z:MarkDuplicates\tRG:Z:H0164.2\tNM:i:3\tMQ:i:60\tOQ:Z:AFFF<JFJJJAFJFJFJJFJAFJJJJJJJFAJJFJJJJJJJJAJJJJAJJJJJAJJFFJJJJJJFJJJFJJJJFFJ<FFJJAFJJJJF-JJJJFFJJJJ<<JJJJJFJJJAFFAJA7FAA<-JJFJ-JJFFJJFFFA<FF<AF<FAFFJF#\tUQ:i:2\tAS:i:140\n",
            "H0164ALXX140820:2:1111:9434:70997\t99\t20\t9999906\t60\t91M2I58M\t=\t10000135\t367\tAAAATTCAACTGACCATAGGTGTATTGGTTTATTTCTGTACTCTTAGTAGATTCCATTGACCTATATCTCTATCCTTATGCCAGTACCACACTGTTTTGTTTACTACAGCTTTGTAGTAAATTTTGAACTCTAAAGTGTTAGTTCTCTAAC\t====?>>?<>?>>>>>>=?>;:==0==><,<?><<=>:=;=>=><>?>??><=>>?>87:==>>=:>>?=>:>>>?<?>=;<;/=?>>??:>?>>7==?=>>@<@@?@=@?>995>@;+??=?>?2@><A>AA?@A?:<?9B?-A?<>6<8\tMC:Z:13S138M\tMD:Z:149\tPG:Z:MarkDuplicates\tRG:Z:H0164.2\tNM:i:2\tMQ:i:60\tOQ:Z:FFFFFFJJ<JJJJJFFJFJJ<AFF-JFJF-JJJFJJJFJFJJJJJJJJJJJFJJJJJA<AJJJJJFJJJJJFJJJJJJJJFJF7JJJJJJAJJJJAFJJFJJJFJJJJFJFJAA7FJA-JJFJJJ7JFFJFFJFJJFAAF<JF-JAAJ7FA\tUQ:i:0\tAS:i:141\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!echo \"Statistics\"\n",
        "!samtools view ${DATA}/bams/mother.bam | cut -f 3 | sort | uniq -c | sort -nr | sed -e 's/^ *//;s/ /\\t/' | awk 'OFS=\"\\t\" {print $2,$1}' | sort -n -k1,1"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "x7bbCArHWNw3",
        "outputId": "62e939c0-2457-41a0-b8fb-38aeec0a4282"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Statistics\n",
            "20\t133237\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Xác định độ phủ (depth/coverage)\n",
        "!${TOOLS}/samtools-1.18/samtools depth ${DATA}/bams/mother.bam > ${DATA}/bams/mother.depth.txt"
      ],
      "metadata": {
        "id": "9Iv6knHzWsvs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "![a4b71c57-1572-4810-b89b-1b8bf0df27ed.jpg](data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAUFBQUFBQUGBgUICAcICAsKCQkKCxEMDQwNDBEaEBMQEBMQGhcbFhUWGxcpIBwcICkvJyUnLzkzMzlHREddXX0BBQUFBQUFBQYGBQgIBwgICwoJCQoLEQwNDA0MERoQExAQExAaFxsWFRYbFykgHBwgKS8nJScvOTMzOUdER11dff/CABEIAlEELAMBIgACEQEDEQH/xAAvAAEBAQADAQEAAAAAAAAAAAAAAgUDBAYBBwEBAQEAAAAAAAAAAAAAAAAAAAEC/9oADAMBAAIQAxAAAAL2QAAADLiNdwc9gKdTjO+ARFuHmsOp21MjXAADod8AAAAAAAAMjnjQYWvZzBQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEXCWFAAAxdrIRmdH0fCZncnnrocve5E6noMzTl8l2evp10L7M2dX1GNs5vkuz3+qdHc6O/Zl8Hfz163L3Rkdjv8EcfQ2LMeNzOr5xdnvR0+v2flnW4dHtS43za6VmN7fI1zixeDal4+rw61naCgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIuEsKAAAABwc4ABHDzFAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIuEsKAABxRcAC4sgAAAAAAAAFTUgAAAAAAAFxcAAAAAAAACpokAAAAAAAC4sgAAAAAAACpokAAAAAAAAFxcAAC4s5YuEsKAfOA7Dp9w4ouABcWQAAAAAAAACpqQAAAAAAAC4uAAAAAAAABU0SAAAAAAABcWQAAAAAAABU0SAAAAAAAAC87RxTXAAuLOWLhLCgdDr6/FHQ73zns69oW0C74bCMs5dLFs10C0C0C0C0C0C0Dk+fJLQLQLQLQLQLQLQLQOaEFoFoFoFoFoFoFoFoF/eOj6gWgWgWgWgWgWgWgXfDYQLQLQLQLQLQLQLQL+8dH1AtAtAtAtAtAtAvN7+QbSBzYmx5s9I6XaLQLvhs5YuEsKAABxRcAC4sxXDzGN96nbPXgAAAAAAqakAAAAAAAAuLgAAAAAAAAVNEgAAAAAAAXFkAAAAAAAAVNEgAAAAAAYur5U9ZWTzGj5zv456Ht9TtgC4s5YuEsKAABxRcAC4s8bsYQ4efpcx7oAAAAAAFTUgAAAAAAAFxcAAAAAAAACpokAAAAAAAC4sgAAAAAAACpokAAAAAAGRj6WQamf9yD1vifU+HP1Pt52iALizli4SwoAAHFFwALizwWP2c80+50+4e6AAAAAABU1IAAAAAAABcXAAAAAAAAAqaJAAAAAAAAuLIAAAAAAAAqaJAAAAAAB5zC1sM18Lnyz1Hku/wBA/R9fI1wBcWcsXCWFAAA4o5JJUJv51j806fc6Z39Xzvoj3ShKhKhKhKhKhKgm/hKhKhKhKhKhKhKhKh9jkklQlQlQlQlQlQlQlQmn0hQlQlQlQlQlQlQlQm/lHGoSoSoSoSoSoSoSoTT6QoSoSoSoSoSoSoeO8/6HyJ9gAP0fXydglQm/lHJFwlhQAAOKLgAZGvkH5wB6TzfpD3QAAAAAAKmpAAAAAAAALi4AAAAAAAAFTRIAAAAAAAFxZAAAAAAAAFTRIAAAAAAPIeQ9f5AAA/R9fI1wBcWcsXCWFAAA4ouABka+QfnAHpPN+kPdAAAAAAAqakAAAAAAAAuLgAAAAAAAAVNEgAAAAAAAXFkAAAAAAAAVNEgAAAAAA8h5D1/kAAD9H18jXAFxZyxcJYUAADii4AGRr5B+cAek836Q90AAAAAACpqQAAAAAAAC4uAAAAAAAABU0SAAAAAAABcWQAAAAAAABU0SAAAAAADyHkPX+QAAP0fXyNcAXFnLFwlhQB0o7rN0bOOKLKhORs45+bgek836Q90oSoSoSoSoSoSoJv4SoSoSoSoSoSoSoSofY5JJUJUJUJUJUJUJUJUJp9IUJUJUJUJUJUJUJUJv5RxqEqEqEqEqEqEqEqE0+kKEqEqEqEqEqEqHjvIem8yAAfo+vk7BKhN/LLi4SwoHWzdtGLs/VnC+wtJFY+tkH5wB6TzfpD3iRSRSRSRSRSRSRfxJSRSRSRSRSRSRSRSRyT9gpIpIpIpIpIpIpIpIr7FBIpIpIpIpIpIpIpIquOz4kUkUkUkUkUkUkUkV9igkUkUkUkUkUkUkeF836TzYAB+kbGNrlJFXxWcsXCWFAAA4ouABka+QfnAHpPN+kPdAAAAAAAqakAAAAAAAAuLgAAAAAAAAVNEgAAAAAAAXFkAAAAAAAAVNEgAAAAAA8L5v0nmwAD9H18jXAFxZyxcJZny9fSxdizsBQOKLgAZGvkH5wB6TzfpD3QAAAAAAKmpAAAAAAAALi4AAAAAAAAFTRIAAAAAAAFxZAAAAAAAAFTRIAAAAAAPC+b9J5sAA/R9fI1wBcWcsXCX0e9xTXkvTYzWfSiUDii4AGRr5B+cAek836Q90AAAAAACpqQAAAAAAAC4uAAAAAAAABU0SAAAAAAABcWQAAAAAAABU0SAAAAAADwvm/SebAAP0fXyNcAXFnLFwl/PvyXzur5r1upzCUCI+wUkVj62QfnAHpPN+kPeJFJFJFJFJFJFJF/ElJFJFJFJFJFJFJFJHJP2CkikikikikikikikivsUEikikikikikikikiq47PiRSRSRSRSRSRSRSRX2KCRSRSRSRSRSRSR4XzfpPNgAH6RsY2uUkVfFZyxcJeNs9CXvZV9quyADii4AGRr5B+cAek836Q90AAAAAACpqQAAAAAAAC4uAAAAAAAABU0SAAAAAAABcWQAAAAAAABU0SAAAAAADwvm/SebAAP0fXyNcAXFnLFwlhQAAOKLgAZGvkH5wB6TzfpD3QAAAAAAKmpAAAAAAAALi4AAAAAAAAFTRIAAAAAAAFxZAAAAAAAAFTRIAAAAAAPC+b9J5sAA/R9fI1wBcWcsXCWFAAA4ouABka+QfnAHpPN+kPdAAAAAAAqakAAAAAAAAuLgAAAAAAAAVNEgAAAAAAAXFkAAAAAAAAVNEgAAAAAA8L5v0nmwAD9H18jXAFxZyxcJYUA4+lGiy+9Z9hawsRkbWOfmwHpPN+lPcrELELELELELELHyeT4QsQsQsQsQsQsQsQsI5oIWIWIWIWIWIWIWIWIr79ONYhYhYhYhYhYhYhYi1nCsQsQsQsQsQsQsQsRX36caxCxCxCxCxCxCx4LzfpfNAAH6Pr5WwQsRayouEsKB1unqoyNcs4ouFAZGvkH5w5eIek836Q90AAAAAACpqQAAAAAAAC4uAAAAAAAABU0SAAAAAAABcWQAAAAAAABU0SAAAAAADw3mvaeLBZB2z3mvka4AuLOWLhLCgAAcUXAAyNfIPE52xjj0nm/VnswAAAAAAVNSAAAAAAAAXFwAAAAAAAAKmiQAAAAAAALiyAAAAAAAAKmiQAAAAAAec8J7vwg7nT7h0/Seb9Ubuvka4AuLOWLhLCgAAcUXAAyNfIPIY+xjj1/mfTHrwAAAAAAVNSAAAAAAAAXFwAAAAAAAAKmiQAAAAAAALiyAAAAAAAAKmiQAAAAAAec8J+ifnZ2e51u4Y/r/IevNjXyNcAXFnLFwlhQAAOKLgAZGvkHkMfZyDtejyNc9eAAAAAACpqQAAAAAAAC4uAAAAAAAABU0SAAAAAAABcWQAAAAAAABU0SAAAAAADI/OP0fwg7k0Y/r/PehNjXyNcAXFnLFwlhQAAOOfsFJFY+tkHnODn7ZkbuF682Uikikikikikiki/iSkikikikikikikikjkn7BSRSRSRSRSRSRSRSRX2KCRSRSRSRSRSRSRSRVcdnxIpIpIpIpIpIpIpIr7FBIpIpIpIpIpIpIyfB+78IaPFr5B1fRed9CbOxja5SRVcdnLFwlhQAAOKLgAZGvkHnNbJ9Gea9Lj7BrgAAAAAAqakAAAAAAAAuLgAAAAAAAAVNEgAAAAAAAXFkAAAAAAAAVNEgAAAAAAyPCe78IejzNPqGbu5OsbGvka4AuLOWLhLCgAAcUXAAyNfIPOej856Mx9js9Y1wAAAAAAVNSAAAAAAAAXFwAAAAAAAAKmiQAAAAAAALiyAAAAAAAAKmiQAAAAAAZHhPd+EPR9Tt8Z1e91e0bGvka4AuLOWLhLCgAAcUXAAyNfIMTbmjXyNfINcAAAAAAFTUgAAAAAAAFxcAAAAAAAACpokAAAAAAAC4sgAAAAAAACpokAAAAAAGR432WMauXsah4ntdXtGxr5GuALizli4SwoAmKYupZ9hawsRl6/CZl93skYm8MVtDFbQxW0MVtDFbQxW0MVtDG+bfwxW0MVtDFbQxW0MVtDFbQxW0MVtDHndgxW0MVtDFbQxW0MVtDFbQxW0MVtDF+7P0xG0MVtDFbQxW0MVtDFbQxW0MVtDFrYow20MVtDFbQxW0MVtDFbQxW0MVtDF+7P0xG0MVtDFbQxW0MVtDFbQxW0PP5vsvPHPr9LQPL9X2vmztanD2SFiLWVFwlhQAjN0izii4UBcWQAAAAAAAACpqQAAAAAAAC4uAAAAAAAABU0SAAAAAAABcWQAAAAAAABU0SAAAAAAAAC4uAABcWcsXCWFAAAAAdPuAAcZyAAAAAAAAAAAAAAAA6fcAAmgAAAAAAAAAAAAAAB0+4AD4PoAAAAAAAAAAAAAAAOn3AEFgARcJYUAAADN7HN8jpZ2vFcbSiOnHa5bOQKAAAAAAAAAAAAA6Pe4yM3t3HQvu9mzPdj7L19AsBQAAAAAAAAAAAE0jqR2LrF5u1zJn82h1JeDnvs2AoAAAAAAAAAAAAGbpcfIee5+5UcM6X2zD7vzurf0QFRcJYUAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABw83m49H9872K2mL2077yO0ajFiXdZOtYCgAAAAAAAAAAAAAAAAAAAAAAAAAAIuEsKAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAxtkfPP+hGD1vTjy+hsJPPcPp1uZpgAAAAAAAAAAAAAAAAAAAAAAAAAAAAi4SwoAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACBP/EAAL/2gAMAwEAAgADAAAAIfPPPOvfNPO/9NPPPPPPPPPPPMufPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPKfPPPEiaE3ZbPmF6LL9lmuUx9pXvPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPKfPPPOMFMMkMMMMMMOMMMMMMMMOMMMMMMMMNMMMMMMMMNMMMMMMMMNMMMMMMMMMOMMNKfPPPPPPPPPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPKfPNPPPPPPPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPPPKPPPKfPD5/PLOMPPPPPPPNPPPPPPPPHPPPPPPPPOPPPPPPPPLPPPPPPPPOPPPPPPPOPGPPLKfPPPPPPCFPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPJINFPPKfPPPPPPIEPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPGKPFPPKfPPPPPPEAPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPIOFFPPKfPPPOMMPJMMMMMMMEMMMMMMMMOMMMMMMMMIMMMMMMMMNMMMMMMMMIMMMMMMMOPPEMNKfPPPPPKPKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPPPPFPPKfPPPPPKPKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPPPPFPPKfPPPPPKPKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPPPPFPPKfPOvcMJPLMMMMMMMGMMMMMMMMOMMMMMMMMJMMMMMMMMNMMMMMMMMJMMMMMMMNPPGMMKfPHvTDGPODDDDDDDJDDDDDDDDLDDDDDDDDGDDDDDDDDHDDDDDDDDGDDDDDDDNPPJDDKfPPPPPKPKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPFPPFPPKfW/PPPKPKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPFPPFPPKbOPPPPKPKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPFPPFPPKeUfPDDGPODDDDDDDJDDDDDDDDLDDDDDDDDGDDDDDDDDHDDDDDDDDGDDDDDDDNPPJDDKbsPPPPKPKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPFPPFPPKfPPPPPKPKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPFPPFPPKfPPPPPKPKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPFPPFPPKfPNsfPLPKPPPPPPPHPPPPPPPPNPPPPPPPPLPPPPPPPPOPPPPPPPPLPPPPPPPFPPHPOKfPLnfPKNKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPNPOFPPKfPPPPPKFLPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPPPKFPPKfPPPPPKFPPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPNKPFPPKfPPPPPKGNPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPFKNFPPKfPPPLDCJIDDDDDDDBDDDDDDDDLDDDDDDDDCDDDDDDDDHDDDDDDDDCDDDDDDDFGFBDHKfPPPPPKAKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPFOPFPPKfPPPPPKAIPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPFLHFPPKfPPPPPKKKPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPAFPFPPKfPOufPPPLDDDDDDDDDDDDDDDDJDDDDDDDDDDDDDDDDDHDDDDDDDDHDDDDDDDOFEHPOKfPPvfPPPPPPPPPPPFPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPPPPPKPPPPPPPPPPPPPKfPPPLDBDADDDDDDDLDDDDDDDDDCDDDDDDDHDDDDDDDDDDDDDDDDDHDDDDDDDDBDBDHKfPPPPLtBjfPPPPPPPPPPPPPGPren/ADzzzzzzzzzzzy4DP53zzzzzzzzzzzzzzz5Fx3ynzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz7DFHb3zzzzzzzzzzzzzzzzzzzzzzzzzzzynzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzyxxy/zzzzzzzzzzzzzzzzzzzzzzzzzzzzzynzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzyH//EADgQAAEBBgMGBQIFBAMBAAAAAAIAAQMEBRESEyBAEBUwMTI1BjRBQlEURSEiM0RGJTZDYCNQcCT/2gAIAQEAAQcC/wDBfngzOMi4d7BOmn4gdMacFGO46HB9tio36eJgXDI17vcoPbFPScQsQ9gnxRMJDPkUbSYu4JBGzCJj4hzlZGvd7lB6E46PiYiIcwEecQUS4GYzSLE4iCiwjoZ2/wD9R+eDNO4STY7iSgXPiJ6w5DgXxj97HyqTvI2UuYKFeRcdDwsdGSR8+gwfz5jiEcjLp0UKXSSk0ocRsADxy8eBCTyCJ49KAkUHHQTmTO3ca+l8G+8QAJCwxIYCDgoaJnD+ClLmYQgRT6IevvD0cMBLYeHsiJ2/euYNg7l+lewz6JhmRfiFruEcjLp0ULAS6Hj4iap1DO5vGRygroKPi5fKZVDx0sA5ZLgmcvF67jolkgcKMlDmAhDipm8J+MtfyjdOIZQME6nIvowHjwIOewTJW5eSYYrdzuJlP10ZMD3dKBexUngsJ8ouLcQTkn0K4iMCaR8k7VBLw/5aK/1L54M2cRZv5c+bFTt4xohKAZLH0GDyfOnTIedO4hkHLAfjOI919JMIOIE5Y9dw8S2cfVlCv2zp1El0kpK9mjmABOpfEBBTFpy6JKXy1PnMzmjXLqMcRruaOI1QcG9F/OE4Gcy9z9I9lT8JI/hXTGi7dsmcG2OhCdFvqNJw6ZCv99lElCv2zp1EyuFfw7yZNJxHwEZEv4KEiseKjZNDPoSXO3Mnhn0LLwdOJS/OTfSvxnUe5+kiAmEN9M1xCRcTMXUa6czOVk/du5dEjAzJoQz5km+mCGfMk3027YpsBLG4k7iSAE9dTM5k8iYZ9GRGILl1OJe6KEl0H9DCO3P+o/PFi4N1GYGVv4qCg3cDDi4/3L54LTYxqxRWKKxRWKKYbGtWKKxRWKKxRWKKxRWKKxRWKKxRWKKxRWKKxRWKKxRWKKxRWIxYorFFYorFFYorFFYorFFYorFFYorFFYorFFYorFFYoq9lGrFFYorFFYorFFYorFFYorFFYorFFYorFFYorFFYorFFYorEYsUViisUViisUViisUViisUViisUViisUViisUViisUVeylcUViisUViisUViisUViisUViisUViisUViisUViisUViMWKKxRWKKxRWKKxRWKKxRWKKxRWKKxRWKKxRWKKxRWKKxRTTYxYorFFYorFFYorFFMNjW7Pngn1Nyh1M0bfTRM6T0Zc26L2aIebNGXtzh1M2fPBPqblDqZo2+miZ0noy5t0Xs0Q82aMvbnDqZs+eCfU3KHUzRt9NEzpPRlzbovZoh5s0Ze1S2JOMgnL/KHUzZ85X793Du2vHEa7ftIXUycPXggj6m5Q6maNvpomdJ6MubdF7NEPNmjL2qQ9qhcwdTNnzlmV1YNPDig+phnxAUJL2RML9TYnYYAsd3q9Xq9AVSYr1eomKehMJY6vV6vV6vV6vV6vV6vV6vV6vV3JXq9Xq9Xq9Xq9Xq9Xq9Xq9Xq9XphflJXq9Xq9Xq9Xq9Xq9Xq9Xq9Xq9Xq9NL8Wq9Xq9Xq9Xq9Xq9Xq9Xq9Xq9Xq9XflV6vV6vV6vV6vV6vV6vV6vV6vV6YX4sV6vV6vV6vV6vV6vV6vV6vV6iYp6Ewljq9Xoi5KQF/SYRXq9Xq9XoCqTNnzleOwfA0HEJDQtU7gIRy9xUfU3KHUzZiHvrDjO6yZYh76w+M300TOk9GXNui9miHmzQRndZNtL2qAJo+HGlAkRQUGWQOpmz54J9TcodTFB91nK+/KDJhTNwUh8zC8dvpomdJ6MubdF7NEPNmgiHgPZnJTN4DoWmJCQsJ68du8NQX9tGpf5CCyh1M2fPBPqblDqYnL904ms0X35ST9aWKQ+ZheO300TOk9GXNui9miHmzjm/dO3rl1Lv42p92qKUsfunkK4dTn7SnEXDw/h4Hcv8hBZQ6mbPngn1Nyh1MUV3V8vvykn60sXh56w4t2PGb6aJnSejLm3RezRDzZx4zusmUu/jamUSEZInz+WRzqEjMKdvRY9kzp/Eg9goBxL/IQWUOpmz54J9TcodTFMnuBHRr2InWBN372Rvax0A68Mefe8dvpomdJ6MubdF7NEPNnHmj3Ajpa9hXuCHhov4oh7rLlO3tZvJXSlD3GlsGWQOpmz54J9TcodTFO/wBaZqYefjVIe6wq8Mefe8dvpomdJ6MubdF7NEPNnHnn60InT2rfDrr+KKGe40ygC8RPcCbwD1SHtULlDqZs+eCfU3KHUxREScY4jX756T569eyHusKvDHn3vHb6aJnSejLm3RezRDzZx/Er1rooInHmZCnUc9OVxUE5ek5eunsfMH0xeg9Uh7VC5Q6mbPngn1Nyh1MQeQidktMnUVieGPPveO300TOk9GXNui9miHmzj+Kv2CgTI42WsEyBjxm2Q9qhcodTNnzwTb+ZqqqqqqowyGDjCDyETsYRD+Phfz71VVVVVVVVVVVVVVVVVVU1vJVVVVVVVVVVVVVVVVVVVVVVUxv5SVVVVVVVVVVVVVVVVVVVVVVVVTW/i1VVVVVVVVVVVVVVVVVVVVVVVX8iqqqqqqqqqqqqqqqqqqqqqqpjfxYqqqqqqqqqqqqqqqqqqq8V/sEJEJMLJIW/0mFVVVVVUDfzM2fPBPqbln3aorJ4Y8+947fTRM6T0Zc26L2aIebOP4q/YZpD2qFyh1M2fPBPqbln3aorJ4Y8+947fTRM6T0Zc26L2aIebOP4q/YZpD2qFyh1M2fPBPqbln3aorJ4Y8+947fTRM6T0Zc26L2aIebOP4q/YZpD2qFyh1M2fPBPqbln3aorJ4Y8+947fTRM6T0Zc26L2aIebOP4q/YZpD2qFyh1M2fPBPqbln3aorJ4Y8+947fTRM6T0Zc26L2aIebOP4q/YZpD2qFyh1M2fPBPqbln3aorJ4Y8+947fTRM6T0Zc26L2aIebOP4q/YZpD2qFyh1M2fPBPqbln3aorJ4Y8+947fTRM6T0Zc26L2aIebOP4q/YZpD2qFyh1M2fOVrWCxrYeOa9+rJ3FxjGuDeP3DmiuF7+e0laStJWkp8xu6YrJ4X8+9VpK0laStJWkrSVpK0laStJWkrSVpK0k1jfwVpK0laStJWkrSVpK0laStJWkrSVpK0laStJWkmMbaStJWkrSVpK0laStJWkrSVpK0laStJWkrSVpK0laSaxtWq0laStJWkrSVpK0laStJWkrSVpK0laStJWkrSVG2K0laStJWkrSVpK0laStJWkrSVpK0laStJWkrSTGNqxWkrSVpK0laStJWkrSVpK0laStJWkrSXihpfWuByyFjd0wqtJWkrSVpIGNuZs+csc5iHwALp3GPXk0D6kIsISGeOHD6itF1+S4lcSuJXEp81u6YrJ4X8+9VxK4lcSuJXEriVxK4lcSuJXEriVxK4k1rfwVxK4lcSuJXEriVxK4lcSuJXEriVxK4lcSuJXEmNbaSuJXEriVxK4lcSuJXEriVxK4lcSuJXEriVxK4lcSa1tWq4lcSuJXEriVxK4lcSuJXEriVxK4lcSuJXEriVW2K4lcSuJXEriVxK4lcSuJXEriVxK4lcSuJXEriTGtqxXEriVxK4lcSuJXEriVxK4lcSuJXEriXijz7rNIWt3TCq4lcSuJXEga25mz54J9Tcs+7VFZPDHn3vHb6aJnSejLm3RezRDzZx/E/n3WaQ9qhcodTNnzwT6m5Z92qKyeGPPveO300TOk9GXNui9miHmzj+J/Pus0h7VC5Q6mbPngn1Nyz7tUVk8Mefe8dvpomdJ6MubdF7NEPNnH8T+fdZpD2qFyh1M2fOwzF2BmE1jX4seQr5++BpZD6m5Z92qKyeGPPveO300TOk9GXNui9miHmzj+J/Pus0h7VC5Q6mbPnZMDjaGEEU7cQ7p1BP378DbkPqbln3aorJ4Y8+947fTRM6T0Zc26L2aIebOP4n8+6zSHtULlDqZs+dkxgCjXbWAUE4AXUtcQziHyn1Nyz7tUVk8Mefe8dvpomdJ6MubdF7NEPNnH8T+fdZpD2qFyh1M2fOyMYxsJEsl04g4eBdOZPMIZyJOsh9Tcs+7VFZPDHn3vHb6aJnSejLm3RezRDzZx/E/n3WaQ9qhcodTNnzsfvWOHL17BRMxfxbAlcU9inL1mQ2NuarSVpK0laSnzG7pisnhfz71WkrSVpK0laStJWkrSVpK0laStJWkrSTWN/BWkrSVpK0laStJWkrSVpK0laStJWkrSVpK0laSYxtpK0laStJWkrSVpK0laStJWkrSVpK0laStJWkrSVpJrG1arSVpK0laStJWkrSVpK0laStJWkrSVpK0laStJUbYrSVpK0laStJWkrSVpK0laStJWkrSVpK0laStJMY2rFaStJWkrSVpK0laStJWkrSVpK0laStJeKPPus0hY3dMKrSVpK0laSBjbmbPnY/eOXbo2gyAGIFsFCuoNwLvIfU3LPu1RWTwx597x2+miZ0noy5t0Xs0Q82cfxP591mkPaoXKHUzZ87IkAip1DuDcuXoYck/K5i3GQ+puWfdqisnhjz73jt9NEzpPRlzbovZoh5s4/ifz7rNIe1QuUOpmz52RUsZERLIndb1QMCECD0ch9Tcs+7VFZPDHn3vHb6aJnSejLm3RezRDzZx/E/n3WaQ9qhcodTNnzwT6m5Z92qKyeGPPveO300TOk9GXNui9miHmzj+J/Pus0h7VC5Q6mbPngn1Nyz7tUVk8Mefe8dvpomdJ6MubdF7NEPNnH8T+fdZpD2qFyh1M2fPBPqbln3aorJ4Y8+947fTRM6T0Zc26L2aIebOP4n8+6zSHtULlDqZs+eCfU3LPu1RWTwx597x2+miZ0noy5t0Xs0Q82cfxP591mkPaoXKHUzZ88E+puWfdqisnhjz73jt9NEzpPRlzbovZoh5s4/ifz7rNIe1QuUOpmz5yxP1FjGA/iId5EOxexzr6d9ExsLB2J28GKFj6wlYSsJWEp+JMlMXk8Lsa2PfKwlYSsJWErCVhKwlYSsJWErCVhKwlYSsL8FYSsJWErCVhKwlYSsJWErCVhKwlYSsJWErCVhJglaSsJWErCVhKwlYSsJWErCVhKwlYSsJWErCVhKwlYSaBVarCVhKwlYSsJWErCVhKwlYSsJWErCVhKwlYSsJWFZSwlYSsJWErCVhKwlYSsJWErCVhKwlYSsJWErCTAKrFYSsJWErCVhKwlYSsJWErCVhKwlYSsJeKGNZHuc0gAmymEVhKwlYSsJAJMJmz5yxz585cpxa9dRDhhP4lkLDo+puWfdqisnhjz73jt9NEzpPRlzbovZoh5s4/iUWlMHWaQ9qhcodTNnzwT6m5Z92qKTXRMdA92eGPPveO300TOk9GXNui9miHmzjzN019O3DrY8AnRm7QQb84V9FSHtULlDqZs+eCfU3LPu1RSfOiCXQBbPDHn3vHb6aJnSejLm3RezRDzZx4n+5ZftmHn43ZDf21MFIe1QuUOpmz54J9Tcs+7VFKM7VJtsidMczmOdcZvpomdJ6MubdF7NEPNnHif7ll+2Yefjdjt1h+F4gpD2qFyh1M2fPBPqbln3aopRnapNtlHfpnx2+miZ0noy5t0Xs0Q82ceJ/uWX7BAWwj95MgFhv3i/iikPaoXKHUzZ88E+puWfdqilGdqk2yDAXsXCu5R36Z8dvpomdJ6MubdF7NEPNnHIBPxA7ag8hEqZ/wCTZ/FFIe1QuUOpmz54J9Tcs+7VFKM7VJtkv8/BKUd+mfHb6aJnSejLm3RezRDzZx/v2wPIRKmf+TZ/FFIe1QuUOpmz54J9Tcs+7VFKKFpSySDaVrC+/KUd+mfHb6aJnSejLm3RezRDzZx/vyl/n4JRYsFk1GZ/5FH/AK4L+KKQ9qhcodTNnzwTZ+ZqoqKiop8z+kxSD9Hwun7l6/lslD7+pa5euZ/MWUVFRUVFRUVFRUVFRUVE1nJUVFRUVFRUVFRUVFRUVFRUVExn5SVFRUVFRUVFRUVFRUVFRUVFRUTWfi1UVFRUVFRUVFRUVFRUVFRUVFT8ioqKioqKioqKioqKioqKioqJjPxYqKioqKioqKioqKioqKip/X1L/PwKj3L0TmjuZ/5FH/rgv4mpCz+kwqoqKiogZ+Zmz54J9Tcs+7VFIP0fDCl38bX35ffuO300TOk9GXNui9miHmzj/flL/PwSiu6vlM/8ij/1wX8UUh7VC5Q6mbPngn1Nyz7tUUg/R8MJ3CtgoqQw8VCthZ+4X37jt9NEzpPRlzbovZoh5s4/35S/z8Eorur5Rrpr57hTV01zGtdfxRSHtULlDqZs+eCfU3LPu1RSD9HwwozusmU379LF9+47fTRM6T0Zc26L2aIebOP9+Uv8/BKK7q+R+fhlP+7RS/iikPaoXKHUzZ88E+puWfdqikH6PhhRndZMpv36WL79x2+miZ0noy5t0Xs0Q82cf78pf5+CUV3V8j8/DKauhfeIGuv4opD2qFyh1M2fPBPqbln3aopB+j4YUZ3WTJ/LoaJiHMT9+47fTRM6T0Zc26L2aIebOP8AflL/AD8Eorur5OnTHszcDH/3KC/iikPaoXKHUzZ88E+puWfdqilAgLxnhpkZ3WTbPv3Hb6aJnSejLm3RezRDzZx/vyloC03Dxw6dvprNlC91cqP/ALlBfxRSHtULlDqZs+eCfU3LPu1RSkQsbKoNRndZNs+/cdvpomdJ6MubdF7NEPNnH+/LwuLC+uUH3WcrDCtY/wDuUF/FFIe1QuUOpmz54J9Tcs6dm9lcWyRi0ZVCKKdPCmMqeqKJ5CzUYne63ut7re63ut7re63ut7re63ut7re63ut78lvdb3W91vdb3W91vdb3W91vdb3W91vdb3W91vdb3TJv+Ulvdb3W91vdb3W91vdb3W91vdb3W91vdb3W91vdb3W902b/AItW91vdb3W91vdb3W91vdb3W91vdb3W91vdb3W91vdb3/Kt7re63ut7re63ut7re63ut7re63ut7re63ut7re6ZN/xYt7re63ut7re63ut7re63ut7re63ut7re6hnhxM3a/wDCv79QfdZzsOFiGeJQf/xRSUCdyuEZkDqZs+c0S8a6h3x/TBCBCRMS9inViBpvAYVBVBVBVBUQ5x4eIdQbrCg4UKCqCqCqCqCqCqCqCqCqCqCqCqCqCqCqCqCqD+CoKoKoKoKoKoKoKoKoKoKoKoKoKoKoKoKoKYwbSVBVBVBVBVBVBVBVBVBVBVBVBVBVBVBVBVBVBTWDVqoKoKoKoKoKoKoKoKoKoKoKoKoKoKoKoKoKoNioKoKoKoKoKoKoKoKoKoKoKoKoKoKoKoKoKYwasVBVBVBVBVBVBVBVBVBVBVBVBVBVBVBUilr+AbFqHhSdzCYvqCiYP4Ldr/cBQUG6woOFCgqgqgqgoGDczZ85ngMeAYBCRjWuAR9TcodTNG300TOk9GXNui9miHmzRl7c4dTNnzwT6m5Q6maNvpomdJ6MubdF7NEPNmjL25w6mbPngtBjWrCFYQrCFYQqPelBQj1/hCsIVhCsIVhCsIVhCsIVhCsIVhCsIVhCsIVhCsIVhCsIVhsWEKwhWEKwhWEKwhWEKwhWEKwhWEKwhWEKwhWEKwhWEKfPSdRcE4whWEKwhWEKwhWEKwhWEKwhWEKwhWEKwhWEKwhWEKwhWEKw2LCFYQrCFYQrCFYQrCFYQrCFYQrCFYQrCFYQrCFYQrCFYpbw+kwhWEKwhWEKwhWEKwhWEKwhWEKwhWEKwhWEKwhWEKwhWGxYQrCFYQrCFYQrCFYQrCFYQrCFYQrCFYQrCFYQrCFYQrCFQb0or6pYQrCFYQrCFYQrCFMBjG7Pnh3DdagfOntUYA8FoprWCxrRfOSBpsi4RrWM1rCEqozB2y4DB4NzQAiEkbx26ZUnzkBYYREO9bbrWEJMqhITYwrAvxEb506ojfuXVE7fuXtdaRCLKonzkCYKEAC7ZjOb8MoqGAmiBg8G7J88Nw6wpkamTWsg3ieuXMJFQRb2P9aPjWgNsQ9+qdwwscunUfhshnD2OprHoNMGjLBYDuIGKcw52PRefRw8ZEwkwa2IduzmJfVOVaEVHCoR05ayLcy9y7a15Fax+6a9C2V+Qh1GRDu9kM7jWQ8rhlBzBpPHjsZiX1ZoHLqJjI1sMxw2AYcvh2OnTXmsmrq5wTxD+dkS9iJgItcO4aYtKHacDGk8vDAdbrbExoOsMWuHIQ7oHWT54eB/9TX5w7XwP3bmCeC9Y8ZAPh/I+ci+CyJh2RLuyGhTdPDeuHGBi65w4wGPmREE/fsh0yEfG7eu3MNEA8YTXIk+dvYmFN6YPQhMOGeOXTtjp27d611Dk5hWOChnDz8RlwDCg5h4d87Jpscix8T5/BPDek8KCFsH9LrYpx9S5J0Ym1rtHAPWk8RwLf8Aja4dE5CjpyLkSZu0q2PHF79w9y/P+hPI6CdE0HT9w/ZXax86IWm147YYu/8Avfn/AEIN3b1myD6LfMImRs1jGvHjydNZAsfQT2ZEdI18/fykyY+fjHwDpkbNYxrx4+nL1kCD9xGx4RrqG/7r5/0JxAu3symhu3DlwLRhX8TKXZQjpw/cQD8pfDUmDt42FiWyaKCryJmUviIV/Eyl2UI2CihhHRRbt4U1lZ/918/+DfP/AIN8r//EAAL/2gAMAwEAAgADAAAAEPPPPO//ADjzj3zTzzTzzzzzzzzLHzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzyvzzzzFoqomHA7Zx25jDHJk3k0lLzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzyvzzzzTzjy57zzzzzzTzzzzzzzzTzzzzzzzzjzzzzzzzzjzzzzzzzzjzzzzzzzzzTzzivzzzzzzzzzzzzzzzxTzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzyvzzTTzzzzzzzzzzzxTzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzyzzzyvzxQTzyzzzzzzzzzzTzzzzzzzxzzzzzzzzzzzzzzzzzyzzzzzzzzzzzzzzzzzzwjTyyvzzzzzzzTTzzzzzzxTzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzTyxTzyvzzzzzzzAjzzzzzzxTzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzByRzzyvzzzzzzwyjzzzzzzxTzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzxyxTzzyvzzzzjDRwjDDDDDDBjDDDDDDDDjDDDDDDDDTDDDDDDDDTDDDDDDDDTDDDDDDCTzzjDSvzzzzzzzyjzzzzzzxTzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzyvzzzzzzzyjzzzzzzxTzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzyvzzzzzzzyjzzzzzzxTzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzyvzz/PzzjyjzzzzzzxTzzzzzzzzzzzzzzzzzjzzzzzzzzzzzzzzzzzjzzzzzzzTzzTzivzy2/wA8s8o8888888U88888888888888888s88888888888888888s8888888U88c8sr88888888o8888888U8888888888888888888888888888888888888888888U88888rqe888888o8888888U8888888888888888888888888888888888888888888U88888r8l888888o8888888U8888888888888888888888888888888888888888888U88888ruB88c8s8o8888888U88888888888888888s88888888888888888s8888888U88c8sr7i888888o8888888U8888888888888888888888888888888888888888888U88888r88888888o8888888U8888888888888888888888888888888888888888888U88888r88888888o8888888U8888888888888888888888888888888888888888888U88888r887z8848s8888888c8888888808888888888888888848888888888888888c88084r88f/wDPPPKPPPPPPPFPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPHPOPPPK/PPPPPPHKPPPPPPPFPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPFKKPPPK/PPPPPPPNPPPPPPPFPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPHFPPPPK/PPPPPPPHPPPPPPPFPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPOHNPPPK/PPPLDHFLDDDDDDDJDDDDDDDDLDDDDDDDDHDDDDDDDDHDDDDDDDDHDDDDDDDOPBLDHK/PPPPPPBHPPPPPPPFPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPKOHPPPK/PPPPPPFNPPPPPPPFPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPKKFPPPK/PPPPPPPPPPPPPPPFPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPHFPPPK/POc9PLNLPPPPPPPPPPPPPPPPFPPPPPPPPPPPPPPPPPLPPPPPPPPLPPPPPPPHNFNPOK/PP3/PPPPPPPPPPPFPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPK/PPPHNKONPPPPPPPHPPPPPPOOFMPPPPPPPLPPPPPPPNIOPPPPPPPLPPPPPPPNNHMPLK/PPPPOyHl/PPPPPPPPPPPPPKE231fPPPPPPPPPPPPLeBRr/PPPPPPPPPPPPPGJdSH/K/PPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP2N6f9/PPPPPPPPPPPPPPPPPPPPPPPPPPPK/PPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPLPfD7/PPPPPPPPPPPPPPPPPPPPPPPPPPPPK/PPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPI//8QAOREBAAIABAIHBgQEBwEAAAAAAQIRAAMSITFBECJQUWBhgQQTcZGhwSAwMmIUQEKxIzNDUnKiw9H/2gAIAQIBAT8A/ASEZcIkmK/DCMWnoj19Nc5TifGEdT0SkQjKTwBXE4sJSi8YqPpg3B5N16NYN4QnykWfBLH16f6Yy5SuvRY/b8kSRd11NZfOPfgFFpKpR2QlweyT8MGsucab99NrhskcaaJQu5GVlQjKOz1I06V4b4ncv4j3VRm+0S0y5HVjz7jENDKgqHv89p7pQdOM1X2Xd3IZh/2lX0x7QLH2oDquQhHmz2rbm1e+JI+05iO/8XqZcnLs593HbEQcjLhXVHMGL/yeXwrEf0QJCy9xk6fKWnr78pO2I2e+V695ZqOawqaeS4yrJ90/ejb/ALKOfdxsxl9X+GrazPJ+rNL+lYjQezkjqmRECrDM0x3Tv4mEfdZ2refuDnbeqVeoYmjmZo75ejKZB/VUjV6pd4ed0x95HQVZwRU5G5XniVMoMh3ypAcdMtf3OHdjMJGT7Q3eaZOQWcWZGWpPO8Z1e+zzZyjOymqsYjK6DjyvH+mH71N+VHyMMHMyc1P15mVKMR20if3cMzMzc7ONoyggOzc5EuHlXZJ+PgVyu/VK+3QgiJs4VW17ZPAZ02b+XHHIe3zp03HMje0tX1jWGV6nnKbN8rAr6dvngM8BngM8BngM8BngM6bDV+2Op8jFP1p8k5dvnTp2zDlKKfOsMrvvZMn4tH27fPAZibIjLT+qmu68JG8wjuBtq41fgExnV7ubV1Fa9MbuZmxJalNpBtLc29ePgExmXonpkDTS8nHVY6iKR0yjEd29axb8otPgEwf5mVJLBl9YoPo4jtLNV/Vl18ZahH0L8AngM8BnTYXe1X60W4pLvksU7k5dvnTQmYPCQnphVu+LJkvevb54DPAZ4DPAZ4DPAZ4DOmUtMZS7hcJV3xjJi/Ep+/b50oSEeDi1+avmvPt8/J51z7aPxf0QEvTanJNa16mCyUbbOXlvzwbBZdZsZV5CP2xHZhvvGMm/Nlseg/nnVZtcctPWzFIZlfD4li+rjnB7h/s43ITqJZGYR71nqHEqGotxjsPkfnTLhIO7Gd15Zlbivrd4W8zNb4rT8T/7jbTEY3tM+Ft3jgzbtaL7+99fzzaNfvX0Q3wWQgd2gT4RT5Xww8Zb8cur89sUSraosjq9xzPXhhWSrxeg7BN51t+hl8kPvjaoJe8ZSfIjLTgLZdxz5O2rZ+DiHXjB4an5/wCEZlmBsinPRRzqaA/XFbReTdejX86dgm0l/ax+aP2xva6ncpxHqz1AURIh5EdIedYiEYwjxIn/AJ+6fpiKkoSW5RYbvMhIl9WO+DaMY92r6q/f+dPAZj//xAA5EQEAAgADBAgDBQcFAAAAAAABAhEAEiEDMUFREBMiUGBhgaFCcZEEIDAywRQjM0BSkrFDYoLR0//aAAgBAwEBPwD7iIxOKWHrX6YESzoezd/0xl6SllPfohFnKMTfJA9cRkSjGRuSzDoo7yr9S8OkpR4xafmaJ08ZHEr3L/BlcWt/aytcHk4WmvNL4Wbzul+7I7cG9OqiepKWLJVJNHbbWciXKTZYb8RqJ9nNpctmbEuKWvalonPDmyQzNz6jZRXjZtNfbEQNu0aLs3TnkjfvjZv7/ZMntH2q83KAt64A6jZkiz9my5eJPX6O7XCptpyvtJs3MeUS/e8cZJVddtc3nHN2PmBeEGGxA7GfbJHkXFjeJ6w5w6qRl/368OaVTwxOpftF6/wcnpkusT1n9okOrt5N3vhmdD2cWFZCo9fdVRWTXTkuKIkiCEo7baR2byEkR/4iGCr0i31cietLqVrxd+PhmR3dZFWqsyt++NlkdpsCg2Tttr2XcQWNX5Y2X8LYL+fqtsLubYxy/LXdijPYV2AdONvvjMbPawH8kNpFlWuZH/Bgiw2ex2W9hMVN1Rix977pfv76eRXpv6BpvHPzV+vfL4Denl544p3+9N1LZyDWOXy3SvBHLlL0jAiPOlb9+/3wG+A3wG+A3wG+A3wG9NLlObRiz2s8zv8AekaYPGNezeCIVW4iRPkd/vgNxAGRm/LZfOvLFySKlN61u8AuNlfWQBq5BdXjTq9m1QS1i8NHX03eAXEKzwzCllhxMS3hmuWaKpoVkCWnnLXwC4+Dal0oV/cL9TEtYwK1JjfIpHwC+A3wG9NOla3Xu1ix3aiCPMe/3pumCb4t+uACq3ESIeR3++A3wG+A3wG+A3wG+A3piZpETi1gkNVuYkj5Pf70iiJvMUe1ByO/38HhffT97/UnK6tjry7Jr6ONMv5a0BDjjV40sGN+aJh1z6aSYleRGl9a/HdSBymPpSY0vZ3w1+WP6/TFXKNujKKvII5XBdWlLqn4w04j2QDTTSuGHSMAN1aHk/8AWOM29Oz60GN5Eqi1rlyPT8d5+X+MOs5vPMnqn6Y+E0+O/TXGoqOoS7XNdz6YAADoe4fhXXeH1xSMhTRiD5yLrDoF8eHHfl3fPE3LKZy/9OrpwlMjlmF4XEVPbF6yOVe5f869w/DXmP0xZlqviJepZ+uJXKKLqqrzVtfJxIzM3cydf7+sMSc3WcCbNTzmJ7ZnDrJfIPp/OvgNx//EAE4QAAECAwMEDgUKBQIFBQAAAAEAEQIDEgQhYSAxQLIQExUwQVBRY3Fyk6Gx4RQiZHPCBSMyM1KBkbPB8EJgYpLRNHRDU2VwggYkVaLx/9oACAEBAAg/Av8AsLid5ssMqqfEYfnHbuUUmxzAP4YDGIj0Opdz3EHODyZG1vt8RD8jJodr9H2x+F3bIhZ4JcUQfAKMCqOARFsdja88ozKtizy5AkyIxDGY3qPQ2U0O1+j7Y/C7toVglSaZJpjmTiWfkFKtEoS58j6YGYg8IVis0k2eEmmsmuY3IoAwi4OQ8n8pYneeei2JX8M/1OtFcovlCZ6bS+3fOPUqmmx2iEVf1ZnVmmTIbTKFZm1EmPlqUcr/AFUPzl5vuVUUMqGxh4YSzwg5lIcSJlnrod2IKwVqMUwPFtcFRAgvw4VHMMfo8EVEUWemKFS5hg9JEIiiGekC9WIxQGCOETIaiRGDyuopTwxyNtN5+k+dRZiGK2n/AEkbyrzdcrXHMjtE0VCOoijqqbHVMkzdqMXLTEEKjPjg9eMxEu6lR0xzpkMsRclSsEdMyGP52uI/OQ8L4qKOIQeieuIS1QfMpDiRMs9dDuxBVoqihgtUdMNRAB5blaTFFIkR7VLlOQLuG5CZFFJ2nbZVV9PAytBji+ntYqIEF/AytsyZNN8MDxfQAQmnbY520Qx8IvVljmQWmSK9sqJrb7SnSZsdhil1TYJfKRwqwWmMerfIMRbpaJWyqOqZEJUNRAgA5GUcwx+jwmiKLPTEFHFGbRDZ64I6j6rC4BWiZMjtW0bZDMq+iwcMptoMsWgfPTRnpGf8V8mWmITYYxXB67TIeF6tidG0I78Ap0NE20QGmD7MIF33r+hcHpUxv5SxO82az7btMZiMNQh8VB8lQSif445wiA+4KKY8U14o5n9Z4V6FKiiApE/bPV6ac6jn1T/SYPnGb1vuUyyy5EEV0ydXU4/pCskgTPRSfm6qbmbOVHJpgNkEJvBaJ8yo+ZFmMFT8LrBSLJBaJcUUVPr0GC/hdTCI7VahGYqc2a4BQNBarLSQDmcZwrRZYbPIhjEUfr1mNuRlZ7MJ0O1bWYaxC1+e/YnQNLnx+rfnDKVZZc+GH6udXSw/qCg+cnzDXFwPES/CjnEIUMVMYIigP9QUySLLBDEDMmQTL424Ayo+Z9GoqfhdUfMizGCp+F1NgYTbTFHBfnCskiGfKnl4pdVBEXSVaxCJ02GiGCG+mEKdBTGDFc75zgp0FMYMVzvnOCm/NThMMcHCxe7MplllyRFdMnV1OMArEIZkqXBTFILQvyEFWiyQWbaoSKREIooyeUhWayw2iRHGY4PXoMD8rqY0VqtQiJEObNcFR876NRS/Cyo+d9GopfhZQUwWuysQIrxiLl6NBZIX9eZWJhI/pGxM+S9vggLSITOhhEOLcqtXyftELZ9tEb/gpFngnQOdqnGNqQeUIxVRXmKLlJ/lLE77MMQ2qYJgblGVLMRhD/Sz3/znid6vV6vV+xer1er1er1er1er1er1er1er1er9i9Xq9Xq9Xq9Xq9Xq9Xq9Xq9Xq/YvV6vV6vV6vV6vV6vV6vV6vV6vV+xer1er1er1er1er1er1er1er1fsXq9Xq9Xq9Xq9Xq9Xq9Xq9Xq9X7F6vV6vV6vV6vV6vV6vV6vV6vV6v2L1er1er1fs4njPo0PHQsdCx0PDecTxn0aHjoWOhY6HhvOJ4z6NDx0LHQsdDwUwARRVO2a4tvGJyZhuVEcuIB6ZgpLcqEEwVfQjihaGLo4s6NDx0LHQsdDwXX1jvGJyYTCPns8WYFlMnic9niiBENJH4KAhzHKp+5ekzpVL/VxUu/KqopjfxR3xFUhUhUhUhMFSFSFCwgm7dWGz0w3KkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkKkJgqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQqQmGdUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhUhQsIJu3Vhs9MNypCpCYZkw/j1iqQqQqQqQqQmGxicmZDVCeBSZQhfPw+KgkAR8u81mj0Oql7nrzr3+qqzR6HVS9z15+IejQ8dCx0LHQPf6uzgoSxEmeQR0lREkmTAST0ZeJ3/3GqvYfjUJcG220gjqBewx/nHiHo0PHQsdCx0CXGIoT6QxBcfRUyMQwjOSWChIIIcEKOMQ1NCHLOTwLmJ/6rmJfhl4nf5kTVx2aCHrGFew/GuftP5cK9hj/ADjxD0aHjoWOhY7/ABRevNqoHLTeV/u11NYKGL15UiTWOSqG5e2yVNjaKbKniAMby5XMS/DLxO/+3WLVK9h+Nc/afy4UP+HY4oT2lX68Q9Gh46FjoWO/+/1V/u1LBEMVLPnujZRwxEz5dkhhbqDP+K4YrVLi/t//AFAGqTtlXJ65e5cxL8MvE7+z0WuyRN0QFejPRBFIarkiz5k30Y50T9aDyXMHxHEPRoeOhY6Fjv7PRBaYm6IEz+vPh/vipX7+tX+01YU30TBE/Wi8thm9Sn+z1cvE7/z9m/LiXPzPFdfVK5g+I4h6NDx0LHQsd/5i2flpvox1P1pvkv39amZo5EP9jQpnolS4m6IjsdfWOXid/mACKKfIds10MQUTPHEYi2K6+qVzB8RxD0aHjoWOhY7+Psz4f7gIV1fzokYYdrly6geH6wf5ULPBEIg+CmwwAwwUerhsdfWOXid/5+T4R7EBaKGVOIOIllcwfEcQ9Gh46FjoWO/+8/RE3QTYBD0Vv+qBujDRdDv+mR19Y5eJ36EsRImEEZx6q5+Tqx7AJFxH4rmDrD+RseIfefooSQQXBGT19Y5eJ37qawyOYPiOIejQ8dCx0LHf/efpldfWOXid+6msMjmD4jiHo0PHQsdCx3/3n6ZXX1jl4nfuprDI5g+I4h6NDx0LHQsd/wDefpldfWOXid+6msMjmD4jiHo0PHQsdCx3/wB5+mV19Y5eJ37qawyOYPiOIejQ8dCx0LHf/efpldfWOXid+6msMjmD4jiHo0PHQsdCx3/3n6ZXX1jl4nfuprDI5g+I4h6NDx0LHQsd/wDefpldfWOXicklgM6igpglMYeVmdTpUsSZpADPVC+Z1NnQQPmqiZSyIoTmMN4TFMUxTFN9jWGRzB1gmKYpimKYpimKYpimKYpimKYpkxTFMUxTFMUxTFMUxTFMUxTFMUxTFMmKYpimKYpimKYpimKYpimKYpimKYpimTFMUxTFMUxTFMUxTFMUxTFMUxTFMU3CmKYpimKYpimKYpimKYpimKYpimKYpkxTFMUxTFMUxTFMUxTFMUxTFEltpBbpJym+3rFMUxTFMU2xicmSYPpPEI3Y/go9qeKAA0vnpuZ1AItsEcFYb6NCmyYI2zVQupYEMIzCG4JynKcpyn+xrDI5g6wTlOU5TlOU5TlOU5TlOU5TlOU6cpynKcpynKcpynKcpynKcpynKcpynTlOU5TlOU5TlOU5TlOU5TlOU5TlOU5TpynKcpynKcpynKcpynKcpynKcpynKfhTlOU5TlOU5TlOU5TlOU5TlOU5TlOU6cpynKcpynKcpynKcpynKcpyuYGscp/t6xTlOU5TlPsYnfuprDI5g+I4h6NDx0LHQsd/5geJyuvrHLxO/dTWGRzB8RxD0aHjoWOhY7/zA8TldfWOXid+6msMjmD4jiHo0PHQsdCx3/mB4nK6+scvE7ERaGEOfuUj5HmRS4voxGOGF1OspkRP9ExCLw3rqawyOYPiOIejQ8dCx0LHf+YHicrr6xy8TsSrJLmSYoPXMUyhCxyZkEIaCLbeBTpcEEUMdLQR1711NYZHMHxHEPRoeOhY6Fjv/MDxOV19Y5eJ2IbVNlGghoYvVL/aR+XLZ82KYtrcwBvuVnnGbDMiMZmEuYid66msMjmD4jiHo0PHQsdCx3/mB4nK6+scvE7BiI+bivHBco5MyGKXCxAln1sR0oiOEzp8UUMFBaCrMN66msMjmD4jiHo0PHQsdCx3/mB4nK6+scvE7BzQQmL8FMtQG3WPboAIA0sk3dKnNtsmbFKjIzEw8OSyYpimKYpvsawyOYOsExTFMUxTFMUxTFMUxTFMUxTFMmKYpimKYpimKYpimKYpimKYpimKYpkxTFMUxTFMUxTFMUxTFMUxTFMUxTFMUyYpimKYpimKYpimKYpimKYpimKYpim4UxTFMUxTFMUxTFMUxTFMUxTFMUxTFMmKYpimKYpimKYpimKYpimKYrmBrHKb7esUxTFMUxTbGJ2J8cMMvMTEWCsv/qDahRtcIigdoXdqipRJH0jES5iJ4d66msMjmD4jiHo0PHQsdCx3/mB4nK6+scvE7E6GqXLs5miE5qnZRyoYoPskXJ/VkWmOXB1RvXU1hkcwfEcQ9Gh46FjoWO/8wPE5XX1jl4nYFqnSoxBR6hGbOt1rZ/cP8KGbHHXGY4oo87neuprDI5g+I4h6NDx0LHQsd/5geJyuvrHLxO/dTWGRzB8RxD0aHjoWOhY7/wAwPE5XX1jl4nfuprDI5g+I4h6NDx0LHQsd/wCYHicrr6xy8Tv3U1hkcwfEcQ9Gh46FjoWO/wDMDxOV19Y5eJ37qawyOYPiOIejQ8dCx0LHf+YHicrr6xy8Tv3U1hkcwfEcQ9Gh46FjoWO/8wPE5XX1jl4nJs7VEtUf4RyqfNE4QStsqak9CnToYpc2IAwUtTVmvU+ZTU7XE5uhSTVBFmObNdwpkyZMm+xrDIH/ACDrBMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmTJuFMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmTJkyZMmR/5A1jlN9vWKZMmTJtjE5MmXFFHEW9WGpsULPaIY5kBqmTYWqKNkmQURQmZER6vqcmX1NYZHMHxHEPRoeOhY6Fjv4D/APt3/AnK6+scvE791NYK5oooof7W/wA7PMHxHEPRoeOhY6Fjv8OeOxzIQ+MMezGGihJBGI2BCNqlliX4bru9dfWOXid+6msEW+cmToh/9Yf02eYPiOIejQ8dCx0LHf8AmD8Wzz8zx2OfHwrr6xy8Tv3U1gvf62zDmghmQh8IxxD0aHjoWOhY7/zB+LZ5+Z47D/WR1dHriH9F19Y5eJ37qawXv9bZ99r8Q9Gh46FjoWO/8wfi2G9YTZYBwIi/wm9Y2y0gnAN/nY/f1q6+scvE791NYL3+tsRh4YpsAIwJXvtfiHo0PHQsdCx38i+CxvD01N+uxz8nwjXt1q+HY/f1q6+scvE791NYL3+tsc/L8V77X4h6NDx0LHQsd/8AYfj2Ofk+Ea9utXw7H7+tXX1jl4nfuprBQhydvAA66YsSz9C9u+Ne+1+IejQ8dCx0LHf/AGH41z8vxUIYC2wgAf8AmvbrV8K9ns/5UK/f1q6+scvE791NYLn4/wAwKVBVERaSAMC69v8AjUcLVQTIxiIox/I2PEPsHxrn5esqPXNulsBe9YjIXt9q+Fez2f8AKhX7+tXX1jl4nfuprBc/H+YF/u17d8a9h+PiHo0PHQsdCx3/ANh+Nc/L8V7dYtUr261fCvZ7P+VCv39auvrHLxO/dTWC5+P8wIxVUelX9N6MT7baIJo/8o17D8fEPRoeOhY6Fjv/ALD8a5+X4r26xapUOeP5RtMIfGlRZ4JUmEthLC/f1q6+scvE791NYLn4/wAwL3+qvc669h+PiHo0PHQsdCx3/wBh+Nc/L8V7dYtUr/q07xgWEvUC/f1q6+scvE791NYLn4/zAvf6q9zrr2H4+IejQ8dCx0LHf/YfjXPy/Fe3WLVK/wCrTvGBRZo45MJbGEL9/Wrr6xy8Tv3U1gufj/MC9/qqMHbJbUkHkNS9h+PiHo0PHQsdCx3/ANh+Nc/L8V7dYtUo/wDyNri/tEMS56z+EK/f1q6+scvE791NYKIOHtJ++EuF7/V2PYfj4h6NDx0LHQsd/wDYfjTesLZZgDgX/wAKOB6YrNGMDDDnXt1t1Ques/hCv39auvrHLxO/dTWCIzVkYesV7/V2PYfj4h6NDx0LHQsd/wDYfjRD/VH8HXuNVUB3qduFmf8ABc9Z/CFfv61dfWOXid+gDmkRfdCXKIb6R/GIoQPBBt1R5Kobtj0WdNlmy7X81DVfU63Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt/Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Lt/Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Lt/Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Mt3Y+a3Lt+f8A5PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2Pmty7f2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmtzLd2PmvRZ8qD0Sj52Cm+p17v9V7jV2Npi2owvW131bL9/WqIMWJ+6IuMvE5QzwwEqAxbbFHBthf6VakWTbnd/XEDfip0G1RnPA9TfeFX3KvuVfcq+5QRiqZKjhD4hlGWihlQCIYgKvuVfcq+5V9yr7lX3KvuVfcq+5V9yr7lX3KvuVfcq+5V9yr7lUq+5V9yr7lX3KvuVfcq+5V9yr7lX3KvuVfcq+5V9yr7lX3KvuVSr7lX3KvuVfcq+5V9yr7lX3KvuVfcq+5V9yr7lX3KvuVfcq+5V9yqVfcq+5V9yr7lX3KvuVfcq+5V9yr7lX3KvuVfcq+5V9yr7lX3KrhVfcq+5V9yr7lX3KvuVfcq+5V9yr7lX3KvuVfcq+5V9yr7lX3KpV9yr7lX3KvuVfcq+5V9yr7lX3KvuVfcq+5V9yr7lX3KfT69NJF+Z1FEKJwlUHqhiq+5VcC9Xbw93AWjq71GWihlQCIYgKvuVfcq+5V9yq2MTlHNEGU+dLMmUXDPVE2Z+LOjQ8dCx0LHQ8N5xPGfRoeOhY6FjoeG84ner1er1epbGKBmqzXller1er1er1er1er1er1er1er1er9i9Xq9Xq9Xq9Xq9Xq9Xq9Xq9Xq9BqZ+2VcvqB7ler1er1er1er1er1er1er1er1er9i9Xq9Xq9Xq9Xq9Xq9Xq9Xq9Xq9XUbRtr8L1Mr1er1er1er1er1er1er1er1er9i9Xq9Xq9Xq9Xq9Xq9Xq9Xq9Xq9XqNhtU+OUG5IVer1er1er9nE724dnbYlzYYmz0l1HAIoTwEONglgoZsBhGcg3IWqU/XGnAgsWKjjEI5TcoIhEOUXowAxQ/RLXh+TYmTIYRyksopsAhOYk3KXPgiPIIgdOhIIw2IYgRyhUCtmqa9uTYmTYYXzVFlMnQQvmqLKXOgjb7JfTYiAOU7Ec2CGI5gS2xDAIai5YZzy7G2wV/Ze9R2iXCeQxAKCMRDlBfJxO9mZFHEZDmI9KH8TQ/iWUmWIHqhibhDOvTIc/+n2s5utyqCTP+lB64h9X8UZMyCGOeIYoYwzjOoJcIgmSDXC11xUEiCGCQztCA8Z/xpsMwwPwjOhmFojCtB9SW5Y/R+9S5VEEcQ2uA3YVMvTBPEb/8Oikr0W1AUxPBRn+5TZTwwyBEIIxwxFRy4YpcucaREHAQlQwV3QABmg89N2yKEcNPCORYHxRnQy3DzCS3q8g6VBMhBMVFWehekekgQGMGGCmK7gZei2qna4fUozYsp0sRU0QwvwBnU+XDGJVYFQe6EowCGOb6xADNyDTTMiaGloOB3z7EfycZ7xx1xkgMB9lS7TtEvahFWYaz0IwRzjDHQ8qDPiyMmf8ATj9aKG4YKgbd9ZXwu6FnlmfOIhhJhBvPD9ygzQ5OJ3ur/h0N97qbMeCPMGalT7UZ1IaC5mdQW6OGQ/0AL+ipREs4N2F6qMJBeGIcBCmzzNmEM7MwVTmOYY/x06t65kUebM69Lvl3vtYLnlZWi1maImZoRBS3Qp1timCHNC1I+9s6veEED71KnmVMAZ2dwhMNUdVUZ5YuFD+GED8NOgm+sAWjb9FMky44m+kYQoY6Y4TUJkIYup1rimlmzUj8Ar6jCIfwUi1GSYg0dzuoY2F1+fTqqXa/oLqGZSAfWDO4Uq2RS5cZeKBnz52PArPPMmKGCjNU46FFOimREuYioSb4jFfivSovR6qtqb72fkVX1dV3TlYn+QplskwRDOIowCpU6CYP6Yn8MiGbCYQWJBuRjhrIcQveePsT/IVs9Hzy6dup5OCpfJ21ttce37V9BuDMrDIkbRDEYQZpLxtyMhJEM7btpihjPqwR4nkVplyIpZheGbIiu6L1LlSIIRayIgA2aK4/5VolSNtikzDFHCMzfZJ4FYZEjaIYjCDNJeNuRlLkDbtv2mZKi4DyK2ypI26GIyzLJup4C/HeJ/kK0WOGOEmXQY4H4OB1KlQSxyQhlH8nT5ohii2qOUKgQS9/Ip9gE+KfPMybJBzCLk5SFY7HabLIaLbhNuEXIwQkR1+lGMQNeRW69HmwQbRMeuFqelR/J0+aIYotqjlCoEEvfyKOSdtm/KEM6OCH1qQShBEYYRNqia4OOO8T/wBhsT/2GxK//8QALBAAAQIEAwkBAQEBAQEAAAAAAQARMWHw8SFAkRAgMEFRcYGhscFQYHDR4f/aAAgBAQABPyH/AIKK0+CdZewtMJ0OhrnvgYULog6DOJbjOu49ZouiMDoA6xDcEOQwgd7FDqA5A4cn2M9Fj+hZm2FVPubkd6IwOgDrEMlFJmQ6ToR6TBw8IlIB0oBRRluAR7wGJxBgfD/JCtPg1jpsCDh2wxGijI+jdLM65fu7kdT674aHIOGKeNuEGmASBFMzMmSagZiyLXj7IWTB17xGtUbRoiMGIoU/ExD4QCVDK68HwAp8Rkj7FjsU8bMLb94hROcDqCgAxm5HBnrijTDC4xAAMEX89COEEprQa/BtJwOCOyG9FzEppIEPLBO4jMZiQPmZHJFrx9kLJg6IWFrJMYWKI7Ry9xFgxWA4gCcMTuiMDQQjBQczoGP49AJZw0TMqPsxgFurJ6eERnMYcMU/szp54YGwRNkH/wB4R5golOOqbYMZihV8fEPpAJWHf2qOBGACJt3hBiAAGDJvT+LwowM5dObNySOgZAuAUFlCDmXQ5lOLm/Ni9AAZqSURZ5fD/wAkK0+C4F8+uOqw8rTvxVDTYo4mhBNhe4wQJAMrSWQc+LMwCPmABCPP5odDvIDhIkAIHpnlEcP1CLjAx1maK94uQySkYD1AobqoCSx90TuRWPSBikIkf3b0XkBBEI35ibA7Mm9hxwMILHhIGF1B2LJrMawQAkuRBKYARMBDElHgMUOgeAd0HI4wCgI2hHXZooRcYGOszRUBrAF3nghwVHiG5jkFHAGSTOUfmU5CyWogjmTnxZYiCOZNWd0SGD+YsCv2cOY3iHWOEDSNwIiKA4wQTLcqwWAV/ZUY5gULX5AJYElA5ReA6bPBA5ReA6bPBFpI4meqepYZ0JKGOTGDz2YdPkUOs+JHuS+J+XSgd6xPTTxZYUvNm5/yQrT4oUMWbAfquDhugwjqETeZAkDiPyA/2YrT4JUCCpCSEkJIQKABUhJCSEkJISQkhJCSEkJISQkhJCSEkJISQhC2BxUhJCSEkJISQkhJCSEkJISQkhJCSEkJIQGIxwUhJCSEkJISQkhJCSEkJISQkhJCSEkJISQhCSGOCkJISQkhJCSEkJISQkhJCSEkJISQkhJCcoLOykJISQkhJCSEkJISQkhJCSEkJISQkhJCApAY4qQkhJCSEkJISQkhJCSEkJISQkhJCSEkJIRkwOIdSEkJISQkhJCBQAOwVp8H5t37snB2/v8AFz22Sr8ZL02T+Xf+7YK0+D82792Tg7f3+LntslX4yXpsn8u/92wVp8H5t37snB2/v8XPbZKvxkvTZP5Vh/bBIxPM/Te+7YK091hwLDAOSTAAdUezv+xKSEozIcAOrZ82792Tg7f3+LntslX4yXpsn8u+192wVp7oa9U4MT+J7qFoWMGI6kaWwPq0VejHMuZ84+MxxKphUwqYVMJgN9gqYVMI38KfnIsQyphUwqYVMKmFTCphUwqYVMKmFTCphUwqYRw9L0VMKmFTCphUwqYVMKmFTCphUwqYVMKmFTCphUwsRhtyVMKmFTCphUwqYVMKmFTCphUwqYVMKmFTCphUwqYTfWeiphUwqYVMKmFTCphUwqYVMKmFTCphUwqYVMKmFze2ZUwqYVMKmFTCphUwqYVMKmFTCphUwqYVMKmFTCf6z0VMKmFTCphUwqYVMKmFTCphUwqYVMKmFTCN/Cn5yLEMqYVMJiPwckwJ+YKphUwqYVMKmEwG+w2CtPdHNiSQweYTkj5cjTFaI4P0BwGz5t37tmr+uWwdVXyrV/XLYOvHg7f3+LntslX4yXpshXy7flRtJ8DEEYoRGL4OSSOJ3fu2CtPg/Nu/cq+fZxeT4HBBxRkDUHb+/wAXPbZKvxkvTZDCk8YjMiFCo8wHwxJQFLiDggwIK6RZbkQPEnoqj1WidO792wVp8H5t37l0/wDsly2GGS5Y1B2/v8XPbZKvxkvTcdm/CBgeBlU87Gn5+EBgT3Kp7rEU/wAYBy7qidO792wVp8H5t37tzdywvC7n6mF446Dt/f4ue2yVfjJem49fKqnlYf2xowPI/RFhV+IIBaiIRhxbJ5MbFR3+PEB2CSidO792wVp8H5t37lOnszv2XZl1fH/+S709Rw28ZCXg7f3+LntslX4yXpuPOv8AM+Oy5x7V26jw+23+1/NhtnkLi13meW3fu2CtPg/Nu/dsLUTryLReDt/f4ue2yVfjJemyFPmv0Xw3jZuk+7v1XllMn8zvm32vu2CtPg/Nu/cmctoSJmfogMAPoHNy2RaLwdv7/Fz22Sr8ZL03HF6CTjehz0dgzeCRgcZj4oBgh9A5uHQmTQAkBiJ5k9d9r7tgrT4Pzbv3bQXVgdmIiOQLwdv7/Fz22Sr8ZL02Qt2tI4v1RmsxuoA3Xgmvu2CtPgmELRPloE+WgT5aBPloEfzh0iAiCCNrwySzEAtgDEeQiZevloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgRWhh0HVPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgRuXy5BPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBG5cTyCfLQJ8tAny0CfLQJ8tAny0CfLQJ8tAny0CfLQJ8tAny0CfLQJ8tAny0CfLQJ8tAndPiOifLQJ8tAny0CfLQJ8tAny0CfLQJ8tAny0CfLQJ8tAny0CfLQJ8tAny0CfLQJ8tAjcuI5BPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgT5aBPloE+WgROmBS4gxBECDumEHwU+WgT5aBPloE+WgRjC02CtPg/NmGi8Hb+/xc9tkq/GS9Nmra+7YK0+D82YaLwdv7/Fz22Sr8ZL02atr7tgrT4PzZhovB2/v8XPbZKvxkvTZq2vu2CtPg/NmGi8Hb+/xc9tkq/GS9Nmra+7YK0+D82YaLwdv7/Fz22Sr8ZL02atr7tgrT4PzZhovB2/v8XPbZKvxkvTZq2vu2CtPg/NmGi8Hb+/xc9tkq/GS9Nmra+7YK090d4AkjAALBeWGxHiHunQ737iYLnseCe3R1Mlk42GBCsSsSsSsSHdLcQEth1tiViViViViViViViViViViViViViTPHh0mrErErErErErErErErErErErErErErErErEovclYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlH+J5KxKxKxKxKxKxKxKxKxKxKxKxKxKxKxKxKxKGc8slYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlH+I5KxKxKxKxKxKxKxKxKxKxKxKxKxKxLC0RDgC6PrePZJKxKxKxKxIcxLTYK091jpAYIRyPQu4WBvhOqxz2HMNicrnseCa/R1Ilkw+OACuSuSuSuSPdLcRENlltyVyVyVyVyVyVyVyVyVyVyVyVyVyTPHh1mrkrkrkrkrkrkrkrkrkrkrkrkrkrkrkrkrkn+7krkrkrkrkrkrkrkrkrkrkrkrkrkrkrkrkrkrko/wATzVyVyVyVyVyVyVyVyVyVyVyVyVyVyVyVyVyUc555K5K5K5K5K5K5K5K5K5K5K5K5K5K5K5K5K5KP8RzVyVyVyVyVyVyVyVyVyVyVyVyVyVyREvl97R2SSuSuSuSuSPES12CtPg/NmGi8Hb+/xc9tkq/GS9Nmi7X3bBWnwfmzDReDt/f4ue2yVfjJemzRdr7tgrT4PzZhovB2/v8AFz22Sr8ZL02aLtfdsFaex9sZ+gxFDOc5MDqxQvYaHAOr7r5sw0Xg7f3+LntslX4yXps0Xa+7YK09jgrRgx8CNEFjNgB6XaKIaoQgBhzIgZbvzZhovB2/v8XPbZKvxkvTZou192wVp7IIsMSTosHSSvsIAiFPnm4TkSRu/NmGi8Hb+/xc9tkq/GS9Nmi7X3bBWnsAISMic9QAJ5lywGKDoFqptxxWbd+bMNF4O39/i57bJV+Ml6bNF2vu2CtPYA9y4droeGPg4LwJ0j2AUd908wLRWJWJWJWJDuluICWw62xKxKxKxKxKxKxKxKxKxKxKxKxKxJnjw6TViViViViViViViViViViViViViViViViViUXuSsSsSsSsSsSsSsSsSsSsSsSsSsSsSsSsSsSsSj/E8lYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlYlDOeWSsSsSsSsSsSsSsSsSsSsSsSsSsSsSsSsSsSj/EclYlYlYlYlYlYlYlYlYlYlYlYlYlYkBD4be09kkrErErErEhzEtNgrT2MZBBAcGxTJEJ4j4FnwU7VZiYJMz3fmzDReDt/f4ue2yVfjJemzRdr7tgrT2BcY9IyxyEclQMwOhGXAXtLtiAbvzZhovB2/v8XPbZKvxkvTZou192wVp7MTU2I7uoHZyIgToBIGMAOm782YaLwdv7/Fz22Sr8ZL02aLtfdsFafB+bMNF4O39/i57bJV+Ml6bNF2vu2CtPg/NmGi8Hb+/wAXPbZKvxkvTZou192wVp8H5sw0Xg7f3+LntslX4yXps0Xa+7YK0+D82YaLwdv7/Fz22Sr8ZL02aLtfdsFafB+bMNF4O39/i57bJV+Ml6bNF2vu2CtPdNtM5MNzGGKwkY8QTAJp1FjJuZprXfs5oD1WJZWHiiYMMQqxVYqsVWKKR3CGwLaxVYqsVWKrFViqxVYqsVWKrFViqxVYotylWKrFViqxVYqsVWKrFViqxVYqsVWKrFViqxVYoIERZlWKrFViqxVYqsVWKrFViqxVYqsVWKrFViqxVYqsUeYcyrFViqxVYqsVWKrFViqxVYqsVWKrFViqxVYqsVMGJViqxVYqsVWKrFViqxVYqsVWKrFViqxVYqsVWKHIeYKsVWKrFViqxVYqsVWKrFViqxVYqsVWKGwb2kIIqxVYqsVWKKBwbBWnujRDhxGebB6QmnjDw2Jx0Ray3Mzq832fNmGi8Hb+/wAXPbZKvxkvTccWHJjAHwJJ8DhtfdsFafB+beaIYs6ObiRyUvB2/v8AFz22Sr8ZL03HN4AcoHEx282w7MRiMNjaRCwxIgwR3jX3bBWnwfm3mjMMZ/QRPOSF4O39/i57bJV+Ml6bj0Hpt0Tr2UHrvWvu2CtPg/NvNV8u0mBJ6iYTHjwdv7/Fz22Sr8ZL03HoPTbonXsjDbbIw3jX3bBWnwfm3mq+XbTS8eDt/f4ue2yVfjJem49B6bJ6eXxiPRDp5dEx9cCa+7YK0+D8281Xy7OTQdnBBGCppePB2/v8XPbZKvxkvTcd2N2M788MC/NfdsFafB+bear5dlE6VTS8eDt/f4ue2yVfjJemzXBfmvu2CtPg/NvNG4ngDkkhgEQEPQGYEscPJ9nppePB2/v8XPbZKvxkvTZD0TpQOI+DAAQjg/oTX3bBWnwTGFqny1CfLUJ8tQny1CMIPkoVcU5HBHHehsL5O4TFAIKfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQitDDqOqfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQjcvlzCfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CNy4nmE+WoT5ahPlqE+WoT5ahPlqE+WoT5ahPlqE+WoT5ahPlqE+WoT5ahPlqE+WoT5ahO6fMdE+WoT5ahPlqE+WoT5ahPlqE+WoT5ahPlqE+WoT5ahPlqE+WoT5ahPlqE+WoT5ahG5cRzCfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQny1CfLUJ8tQndHkta6EUxErjmALdQUOy6DRKYwfJT5ahPlqE+WoT5ahGELXYK0+D82+1Gqecj/B2/v8XPbZKvxkvTZD0Tp4W79Ca+7YK0+D82+1GBLIsHBwDPRTFosHkIRyHg7f3+LntslX4yXpsh6J07NwuQA9QOY4o2gJ+gcfDfmvu2CtPg/NvtRq+VU0+Q8Hb+/xc9tkq/GS9NkPROnhbgWCa+7YK0+D82+1Gr5VTT5Dwdv7/Fz22Sr8ZL02Q9E6du4EZYg/RMLhvzX3bBWnwfm32o1fKuUa8BDOKjkPB2/v8XPbZKvxkvTZD0Tp2bjYkAYXqI9jgeJr7tgrT4PzbzTPYqiiFXy5Hwdv7/Fz22Sr8ZL02Q56eXRM/CHIsfeqAODu8TX3bBWnwfm3miaEkzSImXHgqvlyPg7f3+LntslX4yXpshxYjB2iHxIwVXzrR/D/AGoH6cDxNfdsFafB+bdYGhAhgzeAELDkzBDYEA+Qj4iQH5rTZFnkcgY45cdjGMYxjGMYxjDg6bJDGMYxjGMYxjGMYzBehkxjGMYxjGMYxjGMYzvRkhjGMYxjGMYxjGMZy8kBjGMYxjGMYxjGMZ3o44xjGMYxjGMYxje5xSbGuvn2F43gF0+RDEbJ/cdbifUd37tgrT3oA4u4CwnwlMtid16zzMJnWONXTEOQXGO7mZkNxGIDnxKEf6OwIIwyWZmZmZmZmZmYW5fTJZmZmZmZmZmZnQLB8MnmZmZmZmZmZmZ3p6ZLMzMzMzMzMzM7BzNkszMzMzMzMzMzvT0yGZmZmZmZmZg8MBilheP/ANT2hIJoiNmRFmDko6gLuSmMkFCP9HYEEYbuZmAMOPRtgrT3onJPscEPotu463Z82792Tg7f3+LntslX4yXpsn8u/wDdsFafB+bd+7Jwdv7/ABc9tkq/GS9Nk/l3/u2CtPglRJKmJMSYkxBDYFYsDyN1UxJiTEmJMSYkxJiTEmJMSYkxJiTEmJMSYhK2JwUxJiTEmJMSYkxJiTEmJMSYkxJiTEmJMQSZYyYMEkxJiTEmJMSYkxJiTEmJMSYkxJiTEmJMSYhKSXOKmJMSYkxJiTEmJMSYkxJiTEmJMSYkxJicg/Gb2UxJiTEmJMSYkxJiTEmJMSYkxJiTEmJMQBILnBTEmJMSYkxJiTEmJMSYkxJiTEmJMSYkxJiYAchN2mJd8VMSYkxJiTEmIFEE7BWnw9Qoxbq2xq9DFndlF7YMRsYHYMMAOScAFB0upBupRUAiwAISdc90hgF2I5FFIFiUB7QATYFZqE/el0DBF3J9mPYmZB7UG26guxKwyO7obxnhI0QJONhaTgRwdFpyyI+Lps7VMXdnXaoi7s6Y3ODLNM6RiqJGGyGpAyPYHZH1LYHIk3M7Pl39ULDYlg8FBBVgEGo3RWnwzCiHzUyiAkgSRDjGQiWAAuFmY3WCmH/wGBD0WQSOAkFg7nBDebtcA9CsPsxB+ECR5WNjYoxAMOQZ0nwZDYObOnXmU5c4JiDkniODhzkoAm4QmcgPRE+3RAIgB8IOE8QFhgnhiHYsiO8hSMQ464JshfAJD4P0dPtiQ8AQOHPOoz30jMclgoMSMHRq/mOfwnsqEWHES5A5tyRDOEV7snknHigCITnmQPVYKloXcgap8j8DjgwfsjpbZQCYLdBnTuCmmGBxddmCEfAGsABxwHRCLkDJODBmPl1iOkgSLOHmZPRyTjCTuCL4NBliSJw+vijzHG4xE45EPlhN3me+6K0+GGczHB6YjowiFgIW5HnjijLI4iI844RKOxMARo6IgExwmzGyKeISRiGKCjeJiIeQAUZfUWbs8Z4oRwAcEnwg+RGOIDvjLEpAlCxOEGNwtCoGCWfcmjfRFZubES8iCn2Q2JyYzEQhwJPY2eb8huTJLu5NrIgSxOqw1JlmSOfREsdgMGPonDOYOTE/6gBLtCLYRgUIFNgTDixcvCOe5iQ8D+gidjCwwUJIQG8EOLmXMhCP84CAYAlzRwdjqJ6DkJI2QD0xuVpFvrMo7TRhNiRZ3lvCtP8AwUBwSzuCUdDCJEO4N48CkCIgmSARFFAwIkD+8K0/8FTfLoEiYSF93C7oYi6u5sTy/KI+mGvNHpFi9yIX9Qx+QjGENvZYbmShykhtzKGIBc0MRdXc2J5flMzwST9UCPRQOIVZzD/bYrT/AMEx6C8DzvBHhmIY+kQZlg7xQ+ItSVgaOuMzWwFJRjMXZAsAFAHNi9MiYdJKIMywd4rzekShuiZDnhGCOeX9sVp/8GFaf/BhWmv/xAAsEAABAwIFBAMAAgMBAQAAAAABAGHwEfEhMUBR0RAgMEFQobFxkWCBwXDh/9oACAEBAAE/EP8AwXNDj8NTo1A4tjpwefQiRL2EvZ8xxtSBkaIiaXHOeFXB2BoW8JMwRQRggwPnIGoIAj098l2UUY9IcqfpxmPdImlxznhVwaJRk2s/yvD/ALGKk7lhWebg4Os2yYzr1X8v8SzQ4/DLbegpde5vtowTNSp/5PQMR2fjqmCkzGsCSQVgBCf8hdrWiTV9+Gf3K5fZUIxEBMBHUkTfu1RL6sjwtE0IxDc44vUZUMvlhwLi+bwijoRUL3l85VZW6hUc8egRCYyw9V2UKJEw2VF5IL79Xb+KHT0NR7jC+/DP7lYTqzYLWkdSjsaG2PPyKBjBtfDRrAUblLY1ECaYt86chDnU4HfYoaf0L7jMWpEsmE9X1SgKJkpdBf8A9Qq9NCht6Mc4KTb92qJvAatJo9fsCqvJRKOseBSrLCIQ5IURWswgJFBWAAZEVWFLTO9f2BEey1LwAEGziXuEN/sH+JZocfhzae1AiEU+hqNwakZHXziPH94IQU2FfAmVtDrRSmmBBp/C+WUQrrhEpc/BJgld1djjFTRK3LJGmYY6y+yr6OYfTCnHq6CrOaUNzTaKYpoO2CPMYXgIr1U5Y3TVbuxEkqGSlX+Rzl1zE0quPNPDSSYn4+rQAEYLM1dmqQQ6d85pFNPAx2tn1rcskaZhjrWnQdn8coIv2P8AQ4aVemws9x5Sr0wXpYoIF6k82scECpwApevRJoYikOG7+qWpuDgIwHAsWPW9MEBKrysBNVzjnGgWF7Trjv7PPpTrjv7PPoeUvSCaUTA9lR0xCkC3dJwlYxe8prfSnGTCaQBTEfvl6oxU1x2Q7Z/4lmhx+UYF2XDGOdoJfIgP+1Uh72BzqRf8zzQ4/DWiRsArEOVYhyrEOVYhyq0SdwFYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOUGPJVyG9FYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhygRo01wHtWIcqxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDlH7VIjIKxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDlWIcqxDno6rEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOUPtUgMgrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhyrEOVYhysS4KFAMirEOVYhyrEOVYhyrEOVYhyq0SdwOmaHH4Zzds5tH9v8ArRRn0cZv8JozfVfObpmhx+Gc3bObR/b/AK0UZ9HGb/CaM31Xzm6ZocfhnN2zm0f2/wCtFGfRxm/wmjN9J+HA0wp9AfcTm6ZocfafKaAVH0DZksTbLZ/INUK9ej0Zzm7ZzaP7f9aKM+jjN/hNGb6r15zdM0OPtJLRwFElBCJS4h4B6dRKLxYbPqzXEI9sqpTBf7ME9Pr1pWg7TMzzQlcaBy6mebxIuCxgNRoDMzMzMzMzM6ZM7/TidEZmZmZmZmZmZ16C9IwGp96MzMzMzMzMzMzOga9DFTopmZmZmZmZmZn/AF76Ms9EZmZmZmZmZmZ0DRqIqNBMzMzMzMzMzzeJFwWMBqOpnv8ADY1Ur6C/lie4zMzPNCVxoHLpmhx9obFUFwKbMoF6JCoj5jgAodh6E5u2c2jZQNrMoG/t/wBaKM+jjN/hNGb6Rbzd6GH6CIyclh/EhzJ7ZzdM0OPwzm7ZzdVuD/Ww/BoAb9v+tFGfRxm/wmjN9BlbNsqVE0vm4soBUgCpKAvIYP1AYEEZFfzExCf0SDvt/pzdM0OPwzm7ZzaONe9/3A37f9aKM+jjN/hNGb+f2BY/fgBFDfrK+QFn/vIilQ6T27NNb1E939Obpmhx+Gc3bObsX9wDPOn8q89+3/WijPo4zf4TRm+hWnDoaYGfg0Ys2kd6oELt2qsw9U+UWuB7v05umaHH4Zzds5umNN2/uG3/AB6l/E7w0Efvt/1ooz6OM3+E0ZvoMb7r6+vD2ypI1tPvq0D2zObpmhx+Gc3bObRuPy/77f8AWijPo4zf4TRm+g83XvTBD9MTh4Eo035ec3TNDj8M5u2cyo2NmFPoDQB+UhAmiAkmlToV/wB9v+tFGfRxm/wmjN/OTfeHTyiD966XN0MkoH5SEmKIAQaVCN3dXxUkY2vfXnN0zQ4/DObtnN1/h+DrhVSpoH77f9aKM+jjN/hNGb6AaNetQDKH9ypx7pAaTYssI4eFec3TNDj8IRA9MwPpMzMmZmTMzJmZl7O7WzxAR1gmJsgoidXD0Qgj2EVVGZmTMzJmZkzMyZmZMzMmZmTMzJmZkzMyZmZMzMmZmTMzLI3PEzMyZmZMzMmZmTMzJmZkzMyZmZMzMmZmTMzJmZkzMyZmZMzMmZmTMzIotMH3kzMyZmZMzMmZmTMzJmZkzMyZmZMzMmZmTMzJmZkzMyZmZMzMmZmTMzJmZkMID/ukzMyZmZMzMmZmTMzJmZkzMyZmZMzMmZmTMzJmZkzMyZmZMzMmZmTMzLOUrwIMzMmZmTMzJmZkzMyZmZMzMmZmTMzJmZkzMyZmZMzMmZmTMzJmZkzMyOFT/ukzMyZmZMzMmZmTMzJmZkzMyZmZMzMmZmTMzJmZkzMyZmZHWkXkMP6gMQQcj2hGn84rZmZMzMmZmTMzIIke2QD10zQ4/DObUL/vt/1ooz6OM3+E0ZvqhXnN0zQ4/DObUL/vt/1ooz6OM3+E0ZvqhXnN0zQ4/DObUL/vt/1ooz6OM3+E0ZvqhXnN0zQ4/DObUL/vt/1ooz6OM3+E0ZvqhXnN0zQ4/DObUL/vt/1ooz6OM3+E0ZvqhXnN0zQ4/DObUL/vt/1ooz6OM3+E0ZvqhXnN0zQ4/DObUL/vt/1ooz6OM3+E0ZvqhXnN0zQ4+06c6qAKkk7BGMa0ED//AF2ICq/pnGxnnKr9nxLxRWlU06pnTEg0IorkVyK5FciIofZw6cklyK5FciuRXIrkVyK5FciuRXIrkVyK5Fsx3tyuRXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyK5EIEjJwihxoVciuRXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyK5EeYFBpULdXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyK5Fci/kspVWiXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyK5FciHMQgVqlurkVyK5FciuRXIrkVyK5FciuRXIrkVyIP1KmVg7pFLXuRXIrkVyICmBXElt0zQ4+0TuSMMbFKSaswhB0n7MD9hUo+e6+Zl+z4l4JpWiad0zpgQKk1VyK5FciuRAUvs5dOSS5FciuRXIrkVyK5FciuRXIrkVyK5FciD9jvblciuRXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyIgquMJqcKlXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyK5FciuRDiAQKVC3VyK5FciuRXIrkVyK5FciuRXIrkVyK5FciuRXIv4LK1VolyK5FciuRXIrkVyK5FciuRXIrkVyK5FciuRXIjxEoNapbq5FciuRXIrkVyK5FciuRXIrkVyK5FciLpy7kAoa9yK5FciuREUxjgS26ZocfhnNqF/32/60UZ9HGb/AAmjN9V+XnN0zQ4/DObUL/vt/wBaKM+jjN/hNGb6r8vObpmhx+Gc2oX/AH2/60UZ9HGb/CaM31X5ec3TNDj6CzlJyCqdIhzK0Z1M9aYPbZzahf8Afb/rRRn0cZv8JozfVfl5zdM0OPoIKZ5oVpA8UkUHKVQWlqpGF+x7U5tQv++3/WijPo4zf4TRm+q/Lzm6ZocfQqzFnU7AUTVK8YSeKwsWm9c/CoUO2c2oX/fb/rRRn0cZv8JozfVfl5zdM0OPoZy8AYkDiBIXsBjnz9xRMeHwcYF2pzahf99v+tFGfRxm/wAJozfVfl5zdM0OPofSJbMgphUcj4+/560QIzoY7URTGGILZXIrkVyK5ERQ+zh05JLkVyK5FciuRXIrkVyK5FciuRXIrkVyLZjvblciuRXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyIQJGThFDjQq5FciuRXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyI8wKDSoW6uRXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyK5F/JZSqtEuRXIrkVyK5FciuRXIrkVyK5FciuRXIrkVyK5EOYhArVLdXIrkVyK5FciuRXIrkVyK5FciuRXIrkQdOXchFLXuRXIrkVyICmBXElt0zQ4+huvMKJmh3I6cGFWfRoCD/t/KDwj2pzahf8Afb/rRRn0cZv8JozfVfl5zdM0OPplDDvOLn/tOjH8kP8AX6RGDtM5tQv++3/WijPo4zf4TRm+q/Lzm6ZocfQDvubLxdNlEc+BGHaic2oX/fb/AK0UZ9HGb/CaM31X5ec3TNDj8M5tQv8Avt/1ooz6OM3+E0Zvqvy85umaHH4Zzahf99v+tFGfRxm/wmjN9V+XnN0zQ4/DObUL/vt/1ooz6OM3+E0Zvqvy85umaHH4Zzahf99v+tFGfRxm/wAJozfVfl5zdM0OPwzm1C/77f8AWijPo4zf4TRm+q/Lzm6ZocfaSG6+Y/8AlI9BH8UCj3DL8fSHU6fyzhY1E/MOvSPZQ4Yg4ACiJwU4KcFOHapvUIFJOCnBTgpwU4KcFOCnBTgpwU4KcFOCnBGUMtJxG5KnBTgpwU4KcFOCnBTgpwU4KcFOCnBTgpwU4KcEcfo1RjQqcFOCnBTgpwU4KcFOCnBTgpwU4KcFOCnBTgpwU4LAEkyMQpwU4KcFOCnBTgpwU4KcFOCnBTgpwU4KcFOCnBf0DVGVFOCnBTgpwU4KcFOCnBTgpwU4KcFOCnBTgpwU4LAMgxxCnBTgpwU4KcFOCnBTgpwU4KcFOCnBTgqBA96CbzgpwU4KcFRTBWpqNumaHH2jx5SU/wD8YUessato4u7BP+26SyuOk5tQv++3/WijPo4zf4TRm/nFf5xRDx/hec3TNDj8M5u5ek1WiaTcRsQFNF++3/WijPo4zf4TRm/nwN9CCInqxedsKqrKgaEdAYlco3VbCdu5ec3TNDj8M5u5cz8NMwFS0V/fb/rRRn0cZv8ACaM31Vv9bXnN0zQ4/DObwrrY2+tDET532/60UZ9HGb/CaM31Vv8APbhq85umaHH4ZzahdZP7f9aKM+jjN/hNGb6G3Vh1XHCi7LEqp06tjiF/BK85umaHH4ZzeBdbH11xxUmgJ/b/AK0UZ9HGb/CaM389evZt6q/u8f8AIrzm6ZocfhnNqF1vyf2/60UZ9HGb/CaM31X/AJFec3TNDj8M5u5cn+tg+ACANGWVKIGQkAJHqo0PT+3/AFooz6OM3+E0Zvof+B/qYPiADwkeK85umaHH4SiD6ZgPSdmdOzOnZnTszoo1/jFZ0BkLCS+nlLXNUVFrzszp2Z07M6dmdOzOnZnTszp2Z07M6dmdOzOnZnTszp2Z1mbnidmdOzOnZnTszp2Z07M6dmdOzOnZnTszp2Z07M6dmdOzOnZnTszp2Z0EGuL7ydmdOzOnZnTszp2Z07M6dmdOzOnZnTszp2Z07M6dmdOzOnZnTszp2Z07M6ONX/uk7M6dmdOzOnZnTszp2Z07M6dmdOzOnZnTszp2Z07M6dmdOzOnZnTszrKVrxIOzOnZnTszp2Z07M6dmdOzOnZnTszp2Z07M6dmdOzOnZnTszp2Z07M6GMT/wB0nZnTszp2Z07M6dmdOzOnZnTszp2Z07M6dmdOzOnZnTszrKV/1VxIWgKEsahq2l0p5XZQo1/lBbszp2Z07M6dmdFEn2yA+umaHH4ZzahfxP8A9v8ArRRn0cZv8JozfVf8vI8V5zdM0OPwzm71/FPF2EgdKivt+BQ6Df8A2/60UZ9HGb/CaM30X/L8DvQgi5TGz0oZme8rzm6ZocfhnNqF/Cyf+3/WijPo4zf4TRm+q/5f3rq85umaHH4Zzahfwsn/ALf9aKM+jjN/hNGb6P8A5f3wkfaCdnvK85umaHH4ZzeBfwsOjxw4H9poH+3/AFooz6OM3+E0Zvov+Xg/2D4OK2V5zdM0OPwzm7l6cc3+WTRa3+3/AFooz6OM3+E0ZvoPXp1bHALqHO/aRzw6vbK85umaHH4Zzdy9SP8AAJdEUW/2/wCtFGfRxm/wmjN9BxV+IEA6ZlrqoGlB4GyvObpmhx+Gc3bjhGAUYzEyR+MUR/3mFi6qsBUOiPgnDiXArDziEIQhCEIQhCEr6IoQhCEIQhCEIQhCEp5jzjRhCEIQhCEIQhCEIQlUvs56IIQhCEIQhCEIQhCfv+iiEIQhCEIQhCEIQlEvo5ecIQhCEIQhCEIStjLFUB7IrVu9c0xs4+k0454qKip285umaHH3fd92iEPSRCeBFPzSqPSUesBgx1MhQAkRKIlESiJFr+OERgkK4Yo25j68REGhURKIlESiJREoiURKIlESiJREoiURKIlESiJREsYZcOPHEqIlESiJREoiURKIlESiJREoiURKIlESiJREoiQwhWWNXhioiURKIlESiJREoiURKIlESiJREoiURKIlESiJREijmk11FaiJREoiURKIlESiJREoiURKIlESiJREoiURKIllN7i50yooiURKIlESiJREoiURKIlESiJREoiURKIlESiJBHFZA0FeKiJREoiURKIlESiJREoiURKIlESiJREoiUvqTv0XMUJkk+qYhREjvVCGM1G6Iaid+6mubcx9eIiDQqIlESiJREihnlKo9dM0OPuqj4a50alTsVm/iYKdJzds5tH9v+tFGfRxm/wmjN9V85umaHH4Zzds5tH9v+tFGfRxm/wmjN9V85umaHH4a0SdiFchwrkOFchwrkOFhl2okrNQZZJchwrkOFchwrkOFchwrkOFchwrkOFchwrkOFchwrkOFchwrkOFchwrkOFchwrkOEGHJUzG9VchwrkOFchwrkOFchwrkOFchwrkOFchwrkOFchwrkOFchwrkOFchwrkOFchwsE+Rk0SuCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCP2qROYVyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHC/qLC4ssKuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4Q+1SBzCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHCuQ4VyHC9JzI+xFVjVyHCuQ4VyHCuQ4VyHCuQ4VaJG5HTNDj8f8AQ0s+m5SvtEgAklHyPlf2dNFTXVjQKK1AaEdCUbFwAMyScgqq8jnujQNAh9tFkMgANczClQvbUMiNlkdmU13NAVUe9FLTaoFkjKkDFMaGBp0GjPQLR2qYVNygr+ioqKFGNTmBt7oZ11Yi1EJQ0NCOmUx/6bDApIfMcceNOh8D50D+zoqq2pcnC8OiqMmAARL1b1nW5C0wBXDEnp9+UVsAnp/1NOll7iekYM+q9yQ8rciNsYgoaYGlQduaHH4y3U9FS04YG0Cpjxr/AO8AqrtAjrj+UClcGjxRgsf9Ig7M5wKcTMCVnJ88su8ECmbhhQ5ByQtA0tgdUQx1ooYOUiqPQIBIwqjAmRDGBAYkoQTY6kYcQBqCCBg6DOVIxzKANuSbY5AOqO5tSWNOEgp+XoWaI9RQEo8X1QmA3kHNjNY0oC8dbSQlVWFGNamIH2RigAIXV0MWo3RBYC1XErD6+SGRDA/cmhqgKkNJNUCzseShTzYP45YsiSlSFZ0AAUT6QU3GVCjkGtsSkeiqEfadD0oD7RYrWw1MHOpfUZj/AGIY58scBK5RxOzhGInKpCvUmMCWR2Mp3FExKCwTqj7L24Xbmhx+OLpxFGMoKUgMAZgGIoOks+YCVGuMKoGElAdXVODQ3qSAqDiAfYXsgT+KhRY+FlKr6Y/RZ+Gmc5BrvffKRjwVNUrc23cuVGiG/GrzqD1VKDMwxiClVRQ4wEV0UClLpyFV9bdKisoT8BYThApUAq1pxBWFzxprQVhBpVZUrrN/AomTC9KTTGZDUszmTzVrlF6BiukIpOAgUrVBdJ0aYSAIUt0KR6cGE+LEoAAAGt/0kZ6WSuxb3U5IMZ341CrHTH+YxCm8wuflIbo1Ijc/QPDbBVxhOBNeIUAwRwMOP/8AsAQKUMtnvEwo7s0OP/AqFa8UXYiiG6oGx2JI9h5MLJ6ADQEswgoXSHsSB7Pz2aHH/gW+SDAysW8FSEvhqekpBgWK6z18kUU05MHc9dUVT/fsMJnztEgrG5i2gJhqekpBgWJ3JrIZotFa/wAWnzYmaHH/AIFSCWs4pUjRbqbx3IEKr49z0h2caePBBFIj44uftCr0lhYiqljhuPHcAVKvj3PSGnpph9KPbQPbBMgj06vm80OP/wAGzQ4//Bs0ONf/2Q==)"
      ],
      "metadata": {
        "id": "4fF6_wDkV8Ji"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!echo \"Statistics\"\n",
        "!samtools view ${DATA}/bams/father.bam | cut -f 3 | sort | uniq -c | sort -nr | sed -e 's/^ *//;s/ /\\t/' | awk 'OFS=\"\\t\" {print $2,$1}' | sort -n -k1,1"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "n4z01HAOXOAL",
        "outputId": "4706c556-5a58-45d4-89b6-b1136181ab7a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Statistics\n",
            "20\t53803\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!echo \"Statistics\"\n",
        "!samtools view ${DATA}/bams/son.bam | cut -f 3 | sort | uniq -c | sort -nr | sed -e 's/^ *//;s/ /\\t/' | awk 'OFS=\"\\t\" {print $2,$1}' | sort -n -k1,1"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5aqOhjWxXan9",
        "outputId": "cdb4c3de-2d92-44ea-d91e-f0722557a9fb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Statistics\n",
            "20\t54214\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Single sample mode**"
      ],
      "metadata": {
        "id": "MW_bwQULXfyR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "os.environ['GATK'] = \"/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar\""
      ],
      "metadata": {
        "id": "q4BZSmYTSy9c"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!java -jar $GATK HaplotypeCaller\\\n",
        "-R ${DATA}/ref/Homo_sapiens_assembly19.fasta \\\n",
        "-I ${DATA}/bams/mother.bam \\\n",
        "-L 20:10,000,000-10,200,000 \\\n",
        "-O ${DATA}/variant-discovery/motherHC.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "h6LHL8PzXkR8",
        "outputId": "9a2bb05f-4b1a-410e-82ff-26ca4581d6b0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "08:17:52.124 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "08:17:52.338 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "08:17:52.339 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.2.0.0\n",
            "08:17:52.339 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
            "08:17:52.339 INFO  HaplotypeCaller - Executing as root@f6921e6d410d on Linux v6.1.85+ amd64\n",
            "08:17:52.339 INFO  HaplotypeCaller - Java runtime: OpenJDK 64-Bit Server VM v17.0.13+11-Ubuntu-2ubuntu122.04\n",
            "08:17:52.340 INFO  HaplotypeCaller - Start Date/Time: November 21, 2024 at 8:17:52 AM UTC\n",
            "08:17:52.340 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "08:17:52.340 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "08:17:52.340 INFO  HaplotypeCaller - HTSJDK Version: 2.24.0\n",
            "08:17:52.340 INFO  HaplotypeCaller - Picard Version: 2.25.0\n",
            "08:17:52.341 INFO  HaplotypeCaller - Built for Spark Version: 2.4.5\n",
            "08:17:52.341 INFO  HaplotypeCaller - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
            "08:17:52.341 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
            "08:17:52.341 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
            "08:17:52.341 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
            "08:17:52.341 INFO  HaplotypeCaller - Deflater: IntelDeflater\n",
            "08:17:52.341 INFO  HaplotypeCaller - Inflater: IntelInflater\n",
            "08:17:52.341 INFO  HaplotypeCaller - GCS max retries/reopens: 20\n",
            "08:17:52.341 INFO  HaplotypeCaller - Requester pays: disabled\n",
            "08:17:52.341 INFO  HaplotypeCaller - Initializing engine\n",
            "WARNING: BAM index file /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bai is older than BAM /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bam\n",
            "08:17:53.362 INFO  IntervalArgumentCollection - Processing 200001 bp from intervals\n",
            "08:17:53.367 INFO  HaplotypeCaller - Done initializing engine\n",
            "08:17:53.400 INFO  HaplotypeCallerEngine - Disabling physical phasing, which is supported only for reference-model confidence output\n",
            "08:17:53.417 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_utils.so\n",
            "08:17:53.420 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so\n",
            "08:17:53.463 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM\n",
            "08:17:53.464 INFO  IntelPairHmm - Available threads: 2\n",
            "08:17:53.464 INFO  IntelPairHmm - Requested threads: 4\n",
            "08:17:53.464 WARN  IntelPairHmm - Using 2 available threads, but 4 were requested\n",
            "08:17:53.464 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation\n",
            "08:17:53.805 INFO  ProgressMeter - Starting traversal\n",
            "08:17:53.806 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
            "08:17:55.268 WARN  InbreedingCoeff - InbreedingCoeff will not be calculated at position 20:10000117 and possibly subsequent; at least 10 samples must have called genotypes\n",
            "08:18:03.880 INFO  ProgressMeter -          20:10030111              0.2                   190           1131.6\n",
            "08:18:14.359 INFO  ProgressMeter -          20:10131943              0.3                   720           2101.9\n",
            "08:18:19.663 INFO  HaplotypeCaller - 351 read(s) filtered by: MappingQualityReadFilter \n",
            "0 read(s) filtered by: MappingQualityAvailableReadFilter \n",
            "0 read(s) filtered by: MappedReadFilter \n",
            "0 read(s) filtered by: NotSecondaryAlignmentReadFilter \n",
            "6866 read(s) filtered by: NotDuplicateReadFilter \n",
            "0 read(s) filtered by: PassesVendorQualityCheckReadFilter \n",
            "0 read(s) filtered by: NonZeroReferenceLengthAlignmentReadFilter \n",
            "0 read(s) filtered by: GoodCigarReadFilter \n",
            "0 read(s) filtered by: WellformedReadFilter \n",
            "7217 total reads filtered\n",
            "08:18:19.663 INFO  ProgressMeter -          20:10199356              0.4                  1042           2417.9\n",
            "08:18:19.663 INFO  ProgressMeter - Traversal complete. Processed 1042 total regions in 0.4 minutes.\n",
            "08:18:19.855 INFO  VectorLoglessPairHMM - Time spent in setup for JNI call : 0.036936688\n",
            "08:18:19.855 INFO  PairHMM - Total compute time in PairHMM computeLogLikelihoods() : 3.178507043\n",
            "08:18:19.855 INFO  SmithWatermanAligner - Total compute time in java Smith-Waterman : 2.99 sec\n",
            "08:18:19.856 INFO  HaplotypeCaller - Shutting down engine\n",
            "[November 21, 2024 at 8:18:19 AM UTC] org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller done. Elapsed time: 0.46 minutes.\n",
            "Runtime.totalMemory()=394264576\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cat ${DATA}/variant-discovery/motherHC.vcf |grep -v \"##\"|head -n 10"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MBkS12o4Xvp2",
        "outputId": "e7513e57-0cb8-4552-bc28-f8a5f1edcb8b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tNA12878\n",
            "20\t10000117\t.\tC\tT\t283.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-0.779;DP=23;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.47;MQRankSum=1.044;QD=12.33;ReadPosRankSum=0.154;SOR=0.446\tGT:AD:DP:GQ:PL\t0/1:11,12:23:99:291,0,292\n",
            "20\t10000211\t.\tC\tT\t347.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=2.627;DP=35;ExcessHet=3.0103;FS=1.313;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=9.93;ReadPosRankSum=-0.421;SOR=0.465\tGT:AD:DP:GQ:PL\t0/1:21,14:35:99:355,0,526\n",
            "20\t10000439\t.\tT\tG\t781.06\t.\tAC=2;AF=1.00;AN=2;DP=29;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=26.93;SOR=1.255\tGT:AD:DP:GQ:PL\t1/1:0,29:29:86:795,86,0\n",
            "20\t10000598\t.\tT\tA\t984.06\t.\tAC=2;AF=1.00;AN=2;DP=30;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=32.80;SOR=0.976\tGT:AD:DP:GQ:PL\t1/1:0,30:30:90:998,90,0\n",
            "20\t10000694\t.\tG\tA\t563.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.399;DP=39;ExcessHet=3.0103;FS=2.961;MLEAC=1;MLEAF=0.500;MQ=59.25;MQRankSum=-1.252;QD=14.83;ReadPosRankSum=0.118;SOR=0.307\tGT:AD:DP:GQ:PL\t0/1:17,21:38:99:571,0,398\n",
            "20\t10000758\t.\tT\tA\t1191.06\t.\tAC=2;AF=1.00;AN=2;DP=39;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.18;QD=30.54;SOR=0.850\tGT:AD:DP:GQ:PL\t1/1:0,39:39:99:1205,117,0\n",
            "20\t10001019\t.\tT\tG\t385.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=0.699;DP=35;ExcessHet=3.0103;FS=3.064;MLEAC=1;MLEAF=0.500;MQ=42.41;MQRankSum=-3.219;QD=11.69;ReadPosRankSum=-0.561;SOR=0.537\tGT:AD:DP:GQ:PL\t0/1:18,15:33:99:393,0,480\n",
            "20\t10001298\t.\tT\tA\t899.06\t.\tAC=2;AF=1.00;AN=2;DP=31;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=29.97;SOR=1.765\tGT:AD:DP:GQ:PL\t1/1:0,30:30:89:913,89,0\n",
            "20\t10001436\t.\tA\tAAGGCT\t1246.03\t.\tAC=2;AF=1.00;AN=2;DP=30;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=25.36;SOR=0.836\tGT:AD:DP:GQ:PL\t1/1:0,28:28:84:1260,84,0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Hiển thị realigned reads và assembled haplotypes"
      ],
      "metadata": {
        "id": "oZMaMLp2Zn-8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!java -jar $GATK HaplotypeCaller \\\n",
        "-R ${DATA}/ref/Homo_sapiens_assembly19.fasta \\\n",
        "-I ${DATA}/bams/mother.bam \\\n",
        "-O ${DATA}/variant-discovery/motherHCdebug.vcf \\\n",
        "-bamout ${DATA}/variant-discovery/motherHCdebug.bam \\\n",
        "-L 20:10,000,000-10,200,000"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2QwIIj0bY5M1",
        "outputId": "14689ad8-5bea-4c32-e96c-11ca6f3aaeae"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "08:18:46.010 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "08:18:46.256 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "08:18:46.257 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.2.0.0\n",
            "08:18:46.257 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
            "08:18:46.257 INFO  HaplotypeCaller - Executing as root@f6921e6d410d on Linux v6.1.85+ amd64\n",
            "08:18:46.257 INFO  HaplotypeCaller - Java runtime: OpenJDK 64-Bit Server VM v17.0.13+11-Ubuntu-2ubuntu122.04\n",
            "08:18:46.257 INFO  HaplotypeCaller - Start Date/Time: November 21, 2024 at 8:18:45 AM UTC\n",
            "08:18:46.257 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "08:18:46.257 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "08:18:46.261 INFO  HaplotypeCaller - HTSJDK Version: 2.24.0\n",
            "08:18:46.262 INFO  HaplotypeCaller - Picard Version: 2.25.0\n",
            "08:18:46.262 INFO  HaplotypeCaller - Built for Spark Version: 2.4.5\n",
            "08:18:46.262 INFO  HaplotypeCaller - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
            "08:18:46.262 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
            "08:18:46.262 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
            "08:18:46.262 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
            "08:18:46.262 INFO  HaplotypeCaller - Deflater: IntelDeflater\n",
            "08:18:46.262 INFO  HaplotypeCaller - Inflater: IntelInflater\n",
            "08:18:46.262 INFO  HaplotypeCaller - GCS max retries/reopens: 20\n",
            "08:18:46.263 INFO  HaplotypeCaller - Requester pays: disabled\n",
            "08:18:46.263 INFO  HaplotypeCaller - Initializing engine\n",
            "WARNING: BAM index file /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bai is older than BAM /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bam\n",
            "08:18:46.620 INFO  IntervalArgumentCollection - Processing 200001 bp from intervals\n",
            "08:18:46.630 INFO  HaplotypeCaller - Done initializing engine\n",
            "08:18:46.673 INFO  HaplotypeCallerEngine - Disabling physical phasing, which is supported only for reference-model confidence output\n",
            "08:18:47.086 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_utils.so\n",
            "08:18:47.089 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so\n",
            "08:18:47.127 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM\n",
            "08:18:47.128 INFO  IntelPairHmm - Available threads: 2\n",
            "08:18:47.128 INFO  IntelPairHmm - Requested threads: 4\n",
            "08:18:47.129 WARN  IntelPairHmm - Using 2 available threads, but 4 were requested\n",
            "08:18:47.129 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation\n",
            "08:18:47.170 INFO  ProgressMeter - Starting traversal\n",
            "08:18:47.171 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
            "08:18:48.022 WARN  InbreedingCoeff - InbreedingCoeff will not be calculated at position 20:10000117 and possibly subsequent; at least 10 samples must have called genotypes\n",
            "08:18:57.245 INFO  ProgressMeter -          20:10044240              0.2                   260           1548.5\n",
            "08:19:07.338 INFO  ProgressMeter -          20:10155743              0.3                   810           2409.9\n",
            "08:19:11.904 INFO  HaplotypeCaller - 351 read(s) filtered by: MappingQualityReadFilter \n",
            "0 read(s) filtered by: MappingQualityAvailableReadFilter \n",
            "0 read(s) filtered by: MappedReadFilter \n",
            "0 read(s) filtered by: NotSecondaryAlignmentReadFilter \n",
            "6866 read(s) filtered by: NotDuplicateReadFilter \n",
            "0 read(s) filtered by: PassesVendorQualityCheckReadFilter \n",
            "0 read(s) filtered by: NonZeroReferenceLengthAlignmentReadFilter \n",
            "0 read(s) filtered by: GoodCigarReadFilter \n",
            "0 read(s) filtered by: WellformedReadFilter \n",
            "7217 total reads filtered\n",
            "08:19:11.904 INFO  ProgressMeter -          20:10199356              0.4                  1042           2527.8\n",
            "08:19:11.904 INFO  ProgressMeter - Traversal complete. Processed 1042 total regions in 0.4 minutes.\n",
            "08:19:12.321 INFO  VectorLoglessPairHMM - Time spent in setup for JNI call : 0.044528416\n",
            "08:19:12.321 INFO  PairHMM - Total compute time in PairHMM computeLogLikelihoods() : 3.148089977\n",
            "08:19:12.322 INFO  SmithWatermanAligner - Total compute time in java Smith-Waterman : 2.62 sec\n",
            "08:19:12.941 INFO  HaplotypeCaller - Shutting down engine\n",
            "[November 21, 2024 at 8:19:12 AM UTC] org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller done. Elapsed time: 0.45 minutes.\n",
            "Runtime.totalMemory()=587202560\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get update\n",
        "!apt-get install -y tabix"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zs6w66esbErR",
        "outputId": "93936212-7d12-4ece-9615-8be3b60c215c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rHit:1 http://security.ubuntu.com/ubuntu jammy-security InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com] [Connected to cloud.r-project.org (18.154.101.94)] [Connecting\r                                                                                                    \rHit:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com] [Connected to r2u.stat.illinois.edu (192.17.190.167)] [Waiting\r                                                                                                    \rHit:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com (91.189.91.83)] [Connected to r2u.stat.illinois.edu (192.17.190\r                                                                                                    \rHit:4 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "\r0% [Waiting for headers] [Waiting for headers] [Connected to ppa.launchpadcontent.net (185.125.190.8\r                                                                                                    \rHit:5 http://archive.ubuntu.com/ubuntu jammy-updates InRelease\n",
            "\r0% [Waiting for headers] [Waiting for headers] [Connected to ppa.launchpadcontent.net (185.125.190.8\r0% [Waiting for headers] [Waiting for headers] [Connected to ppa.launchpadcontent.net (185.125.190.8\r                                                                                                    \rHit:6 http://archive.ubuntu.com/ubuntu jammy-backports InRelease\n",
            "Hit:7 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n",
            "Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  tabix\n",
            "0 upgraded, 1 newly installed, 0 to remove and 57 not upgraded.\n",
            "Need to get 351 kB of archives.\n",
            "After this operation, 1,179 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tabix amd64 1.13+ds-2build1 [351 kB]\n",
            "Fetched 351 kB in 0s (1,454 kB/s)\n",
            "Selecting previously unselected package tabix.\n",
            "(Reading database ... 124687 files and directories currently installed.)\n",
            "Preparing to unpack .../tabix_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking tabix (1.13+ds-2build1) ...\n",
            "Setting up tabix (1.13+ds-2build1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!which bgzip\n",
        "!which tabix"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "CNzQ8qwEbMH1",
        "outputId": "9f9c0133-93d9-4df3-ec40-837c6cbb7f3f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/usr/bin/bgzip\n",
            "/usr/bin/tabix\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!bgzip ${DATA}/variant-discovery/motherHCdebug.vcf\n",
        "!tabix -p vcf ${DATA}/variant-discovery/motherHCdebug.vcf.gz"
      ],
      "metadata": {
        "id": "CEA8NcvSamdG",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "877e0464-427b-4386-fe51-df5862db79f8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bgzip] /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/motherHCdebug.vcf.gz already exists; do you wish to overwrite (y or n)? ^C\n",
            "[tabix] the index file exists. Please use '-f' to overwrite.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!chmod +x /content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/bcftools"
      ],
      "metadata": {
        "id": "_1-_vB66e7ze"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Compile\n",
        "%cd /content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/\n",
        "!./configure\n",
        "!make\n",
        "!make install"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KPe-MOIF3Oqu",
        "outputId": "e916dadb-6841-4ad8-da8f-ebbeccdd0f17"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21\n",
            "/bin/bash: ./configure: /bin/sh: bad interpreter: Permission denied\n",
            "cd htslib-1.21 && make htslib.pc.tmp\n",
            "make[1]: Entering directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "make[1]: ./version.sh: Permission denied\n",
            "echo '# Default htscodecs.mk generated by Makefile' > htscodecs.mk\n",
            "echo 'include $(HTSPREFIX)htscodecs_bundled.mk' >> htscodecs.mk\n",
            "./hts_probe_cc.sh 'gcc' '-g -Wall -O2 -fvisibility=hidden ' '-fvisibility=hidden' >> htscodecs.mk\n",
            "/bin/sh: 1: ./hts_probe_cc.sh: Permission denied\n",
            "Makefile:149: htscodecs.mk: No such file or directory\n",
            "make[1]: *** [Makefile:132: htscodecs.mk] Error 126\n",
            "make[1]: Leaving directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "config.mk:34: htslib-1.21/htslib_static.mk: No such file or directory\n",
            "make: *** [htslib-1.21/htslib.mk:186: htslib-1.21/htslib.pc.tmp] Error 2\n",
            "cd htslib-1.21 && make htslib.pc.tmp\n",
            "make[1]: Entering directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "make[1]: ./version.sh: Permission denied\n",
            "sed -e '/^static_libs=/s/@static_LIBS@/-lz -lm -lbz2 -llzma -lcurl/;s#@[^-][^@]*@##g' htslib.pc.in > htslib.pc.tmp\n",
            "make[1]: Leaving directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "cd htslib-1.21 && make htslib_static.mk\n",
            "make[1]: Entering directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "make[1]: ./version.sh: Permission denied\n",
            "sed -n '/^static_libs=/s/[^=]*=/HTSLIB_static_LIBS = /p;/^static_ldflags=/s/[^=]*=/HTSLIB_static_LDFLAGS = /p' htslib.pc.tmp > htslib_static.mk\n",
            "make[1]: Leaving directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "echo '/* Basic config.h generated by Makefile */' > config.h\n",
            "echo '#define ENABLE_BCF_PLUGINS 1' >> config.h\n",
            "echo '#define PLUGIN_EXT \".so\"' >> config.h\n",
            "echo '#define BCFTOOLS_VERSION \"1.21\"' > version.h\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o main.o main.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfindex.o vcfindex.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o tabix.o tabix.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfstats.o vcfstats.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfisec.o vcfisec.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfmerge.o vcfmerge.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfquery.o vcfquery.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcffilter.o vcffilter.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21    -c -o filter.o filter.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfsom.o vcfsom.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfnorm.o vcfnorm.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfgtcheck.o vcfgtcheck.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfview.o vcfview.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfannotate.o vcfannotate.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfroh.o vcfroh.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfconcat.o vcfconcat.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfcall.o vcfcall.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o mcall.o mcall.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcmp.o vcmp.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o gvcf.o gvcf.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o reheader.o reheader.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o convert.o convert.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfconvert.o vcfconvert.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o tsv2vcf.o tsv2vcf.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfcnv.o vcfcnv.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfhead.o vcfhead.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o HMM.o HMM.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o consensus.o consensus.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o ploidy.o ploidy.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o bin.o bin.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o hclust.o hclust.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o version.o version.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o regidx.o regidx.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o smpl_ilist.o smpl_ilist.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o csq.o csq.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfbuf.o vcfbuf.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o mpileup.o mpileup.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o bam2bcf.o bam2bcf.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o bam2bcf_indel.o bam2bcf_indel.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o bam2bcf_iaux.o bam2bcf_iaux.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o bam2bcf_edlib.o bam2bcf_edlib.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o read_consensus.o read_consensus.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o bam_sample.o bam_sample.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o vcfsort.o vcfsort.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o cols.o cols.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o extsort.o extsort.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o dist.o dist.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o abuf.o abuf.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o ccall.o ccall.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o em.o em.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o prob1.o prob1.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o kmin.o kmin.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o str_finder.o str_finder.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o gff.o gff.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21   -c -o edlib.o edlib.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.21  -DPLUGINPATH='\"/usr/local/libexec/bcftools\"' -c -o vcfplugin.o vcfplugin.c\n",
            "cd htslib-1.21 && make config.h\n",
            "make[1]: Entering directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "make[1]: ./version.sh: Permission denied\n",
            "echo '/* Default config.h generated by Makefile */' > config.h\n",
            "echo '#ifndef _XOPEN_SOURCE' >> config.h\n",
            "echo '#define _XOPEN_SOURCE 600' >> config.h\n",
            "echo '#endif' >> config.h\n",
            "echo '#define HAVE_LIBBZ2 1' >> config.h\n",
            "echo '#define HAVE_LIBLZMA 1' >> config.h\n",
            "echo '#ifndef __APPLE__' >> config.h\n",
            "echo '#define HAVE_LZMA_H 1' >> config.h\n",
            "echo '#endif' >> config.h\n",
            "echo '#define HAVE_DRAND48 1' >> config.h\n",
            "echo '#define HAVE_LIBCURL 1' >> config.h\n",
            "if [ \"x\" != \"x\" ]; then \\\n",
            "    echo '#define HAVE_DECL___CPUID_COUNT 1' >> config.h ; \\\n",
            "    echo '#define HAVE_DECL___GET_CPUID_MAX 1' >> config.h ; \\\n",
            "fi\n",
            "if [ \"x\" != \"x\" ]; then \\\n",
            "    echo '#define HAVE_POPCNT 1' >> config.h ; \\\n",
            "    echo '#define HAVE_SSE4_1 1' >> config.h ; \\\n",
            "    echo '#define HAVE_SSSE3 1' >> config.h ; \\\n",
            "    echo '#if defined(HTS_ALLOW_UNALIGNED) && HTS_ALLOW_UNALIGNED == 0' >> config.h ; \\\n",
            "    echo '#define UBSAN 1' >> config.h ; \\\n",
            "    echo '#endif' >> config.h ; \\\n",
            "fi\n",
            "if [ \"x\" != \"x\" ] ; then \\\n",
            "    echo '#define HAVE_AVX2 1' >> config.h ; \\\n",
            "fi\n",
            "if [ \"x\" != \"x\" ] ; then \\\n",
            "    echo '#define HAVE_AVX512 1' >> config.h ; \\\n",
            "fi\n",
            "echo '#if defined __x86_64__ || defined __arm__ || defined __aarch64__' >> config.h\n",
            "echo '#define HAVE_ATTRIBUTE_CONSTRUCTOR 1' >> config.h\n",
            "echo '#endif' >> config.h\n",
            "echo '#if (defined(__x86_64__) || defined(_M_X64))' >> config.h\n",
            "echo '#define HAVE_ATTRIBUTE_TARGET 1' >> config.h\n",
            "echo '#define HAVE_BUILTIN_CPU_SUPPORT_SSSE3 1' >> config.h\n",
            "echo '#endif' >> config.h\n",
            "make[1]: Leaving directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "cd htslib-1.21 && make hts-object-files\n",
            "make[1]: Entering directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "make[1]: ./version.sh: Permission denied\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o kfunc.o kfunc.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o kstring.o kstring.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o bcf_sr_sort.o bcf_sr_sort.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o bgzf.o bgzf.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o errmod.o errmod.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o faidx.o faidx.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o header.o header.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o hfile.o hfile.c\n",
            "echo '#define HTS_VERSION_TEXT \"\"' > version.h\n",
            "printf '#define HTS_CC \"%s\"\\n#define HTS_CPPFLAGS \"%s\"\\n#define HTS_CFLAGS \"%s\"\\n#define HTS_LDFLAGS \"%s\"\\n#define HTS_LIBS \"%s\"\\n' \\\n",
            "'gcc' \\\n",
            "'' \\\n",
            "'-g -Wall -O2 -fvisibility=hidden' \\\n",
            "'-fvisibility=hidden' \\\n",
            "'-lz -lm -lbz2 -llzma -lcurl' > config_vars.h\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o hts.o hts.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o hts_expr.o hts_expr.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o hts_os.o hts_os.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o md5.o md5.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o multipart.o multipart.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o probaln.o probaln.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o realn.o realn.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o regidx.o regidx.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o region.o region.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o sam.o sam.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o sam_mods.o sam_mods.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o simd.o simd.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o synced_bcf_reader.o synced_bcf_reader.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o vcf_sweep.o vcf_sweep.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o tbx.o tbx.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o textutils.o textutils.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o thread_pool.o thread_pool.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o vcf.o vcf.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o vcfutils.o vcfutils.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/cram_codecs.o cram/cram_codecs.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/cram_decode.o cram/cram_decode.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/cram_encode.o cram/cram_encode.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/cram_external.o cram/cram_external.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/cram_index.o cram/cram_index.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/cram_io.o cram/cram_io.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/cram_stats.o cram/cram_stats.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/mFILE.o cram/mFILE.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/open_trace_file.o cram/open_trace_file.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/pooled_alloc.o cram/pooled_alloc.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o cram/string_alloc.o cram/string_alloc.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/arith_dynamic.o htscodecs/htscodecs/arith_dynamic.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/fqzcomp_qual.o htscodecs/htscodecs/fqzcomp_qual.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/htscodecs.o htscodecs/htscodecs/htscodecs.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/pack.o htscodecs/htscodecs/pack.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/rANS_static4x16pr.o htscodecs/htscodecs/rANS_static4x16pr.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/rANS_static32x16pr_avx2.o htscodecs/htscodecs/rANS_static32x16pr_avx2.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/rANS_static32x16pr_avx512.o htscodecs/htscodecs/rANS_static32x16pr_avx512.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/rANS_static32x16pr_sse4.o htscodecs/htscodecs/rANS_static32x16pr_sse4.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/rANS_static32x16pr_neon.o htscodecs/htscodecs/rANS_static32x16pr_neon.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/rANS_static32x16pr.o htscodecs/htscodecs/rANS_static32x16pr.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/rANS_static.o htscodecs/htscodecs/rANS_static.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/rle.o htscodecs/htscodecs/rle.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/tokenise_name3.o htscodecs/htscodecs/tokenise_name3.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o htscodecs/htscodecs/utils.o htscodecs/htscodecs/utils.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden  -I.  -c -o hfile_libcurl.o hfile_libcurl.c\n",
            "touch hts-object-files\n",
            "make[1]: Leaving directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "cd htslib-1.21 && make lib-static\n",
            "make[1]: Entering directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "make[1]: ./version.sh: Permission denied\n",
            "ar -rc libhts.a kfunc.o kstring.o bcf_sr_sort.o bgzf.o errmod.o faidx.o header.o hfile.o hts.o hts_expr.o hts_os.o md5.o multipart.o probaln.o realn.o regidx.o region.o sam.o sam_mods.o simd.o synced_bcf_reader.o vcf_sweep.o tbx.o textutils.o thread_pool.o vcf.o vcfutils.o cram/cram_codecs.o cram/cram_decode.o cram/cram_encode.o cram/cram_external.o cram/cram_index.o cram/cram_io.o cram/cram_stats.o cram/mFILE.o cram/open_trace_file.o cram/pooled_alloc.o cram/string_alloc.o htscodecs/htscodecs/arith_dynamic.o htscodecs/htscodecs/fqzcomp_qual.o htscodecs/htscodecs/htscodecs.o htscodecs/htscodecs/pack.o htscodecs/htscodecs/rANS_static4x16pr.o htscodecs/htscodecs/rANS_static32x16pr_avx2.o htscodecs/htscodecs/rANS_static32x16pr_avx512.o htscodecs/htscodecs/rANS_static32x16pr_sse4.o htscodecs/htscodecs/rANS_static32x16pr_neon.o htscodecs/htscodecs/rANS_static32x16pr.o htscodecs/htscodecs/rANS_static.o htscodecs/htscodecs/rle.o htscodecs/htscodecs/tokenise_name3.o htscodecs/htscodecs/utils.o hfile_libcurl.o\n",
            "ranlib libhts.a\n",
            "make[1]: Leaving directory '/content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/htslib-1.21'\n",
            "gcc -rdynamic   -o bcftools main.o vcfindex.o tabix.o vcfstats.o vcfisec.o vcfmerge.o vcfquery.o vcffilter.o filter.o vcfsom.o vcfnorm.o vcfgtcheck.o vcfview.o vcfannotate.o vcfroh.o vcfconcat.o vcfcall.o mcall.o vcmp.o gvcf.o reheader.o convert.o vcfconvert.o tsv2vcf.o vcfcnv.o vcfhead.o HMM.o consensus.o ploidy.o bin.o hclust.o version.o regidx.o smpl_ilist.o csq.o vcfbuf.o mpileup.o bam2bcf.o bam2bcf_indel.o bam2bcf_iaux.o bam2bcf_edlib.o read_consensus.o bam_sample.o vcfsort.o cols.o extsort.o dist.o abuf.o ccall.o em.o prob1.o kmin.o str_finder.o gff.o edlib.o vcfplugin.o htslib-1.21/libhts.a -lz -lm -lbz2 -llzma -lcurl -lm -lz -ldl    -lpthread\n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/ad-bias.so convert.c version.c plugins/ad-bias.c  \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/add-variantkey.so version.c plugins/add-variantkey.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/af-dist.so bin.c version.c plugins/af-dist.c  \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/allele-length.so version.c plugins/allele-length.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/check-ploidy.so version.c plugins/check-ploidy.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/check-sparsity.so version.c plugins/check-sparsity.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/color-chrs.so HMM.c version.c plugins/color-chrs.c  \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/contrast.so version.c plugins/contrast.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/counts.so version.c plugins/counts.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/dosage.so version.c plugins/dosage.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/fill-AN-AC.so version.c plugins/fill-AN-AC.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21     -o plugins/fill-from-fasta.so filter.c version.c plugins/fill-from-fasta.c  \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/fill-tags.so version.c plugins/fill-tags.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/fixploidy.so ploidy.c regidx.c version.c plugins/fixploidy.c  \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/fixref.so version.c plugins/fixref.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/frameshifts.so version.c plugins/frameshifts.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/GTisec.so version.c plugins/GTisec.c  \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/GTsubset.so version.c plugins/GTsubset.c  \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/guess-ploidy.so version.c plugins/guess-ploidy.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/gvcfz.so version.c plugins/gvcfz.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/impute-info.so version.c plugins/impute-info.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/indel-stats.so version.c plugins/indel-stats.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/isecGT.so smpl_ilist.c version.c plugins/isecGT.c  \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/mendelian2.so version.c plugins/mendelian2.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/missing2ref.so version.c plugins/missing2ref.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/parental-origin.so version.c plugins/parental-origin.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/prune.so version.c plugins/prune.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/remove-overlaps.so version.c plugins/remove-overlaps.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/scatter.so version.c plugins/scatter.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/setGT.so version.c plugins/setGT.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/smpl-stats.so version.c plugins/smpl-stats.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/split.so version.c plugins/split.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/split-vep.so version.c plugins/split-vep.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/tag2tag.so version.c plugins/tag2tag.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/trio-dnm2.so version.c plugins/trio-dnm2.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/trio-stats.so version.c plugins/trio-stats.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/trio-switch-rate.so version.c plugins/trio-switch-rate.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/variant-distance.so version.c plugins/variant-distance.c \n",
            "gcc -fPIC -shared -g -Wall -O2 -I. -Ihtslib-1.21    -o plugins/variantkey-hex.so version.c plugins/variantkey-hex.c \n",
            "mkdir -p -m 755 /usr/local/bin /usr/local/share/man/man1 /usr/local/libexec/bcftools\n",
            "install -p bcftools /usr/local/bin\n",
            "install -p misc/color-chrs.pl misc/gff2gff.py misc/guess-ploidy.py misc/plot-vcfstats misc/plot-roh.py misc/run-roh.pl misc/vcfutils.pl /usr/local/bin\n",
            "install -p -m 644 doc/bcftools.1 /usr/local/share/man/man1\n",
            "install -p plugins/*.so /usr/local/libexec/bcftools\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/bcftools-1.21/bcftools view \\\n",
        "-R ${DATA}/variant-discovery/variants.txt \\\n",
        "${DATA}/variant-discovery/motherHCdebug.vcf.gz \\\n",
        "> ${DATA}/variant-discovery/motherHCdebug_extracted.vcf"
      ],
      "metadata": {
        "id": "ZrWEcVjFbR-I"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/bcftools-1.21/bcftools view \\\n",
        "-R ${DATA}/variant-discovery/variants_woREF_ALT.txt \\\n",
        "${DATA}/variant-discovery/motherHCdebug.vcf.gz \\\n",
        "> ${DATA}/variant-discovery/motherHCdebug_extracted.vcf"
      ],
      "metadata": {
        "id": "xHK6WzBz8FCA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Cohort mode (Áp dụng với bộ ba (trio) mother-father-son)**"
      ],
      "metadata": {
        "id": "eFqrI6fP_nOg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!chmod +x /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk"
      ],
      "metadata": {
        "id": "aKhihSyqAn-I"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/gatk-4.2.0.0/gatk HaplotypeCaller \\\n",
        "-R ${DATA}/ref/Homo_sapiens_assembly19.fasta \\\n",
        "-I ${DATA}/bams/mother.bam \\\n",
        "-ERC GVCF \\\n",
        "-L 20:10,000,000-10,200,000 \\\n",
        "-O ${DATA}/variant-discovery/mother.g.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "vFJgsIhd_sH7",
        "outputId": "0d9e068e-243b-4a4e-d662-cd16df4417d9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using GATK jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar\n",
            "Running:\n",
            "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar HaplotypeCaller -R /content/drive/MyDrive/GenomeAnalysis/data/ref/Homo_sapiens_assembly19.fasta -I /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bam -ERC GVCF -L 20:10,000,000-10,200,000 -O /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/mother.g.vcf\n",
            "09:17:32.616 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "09:17:32.882 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "09:17:32.882 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.2.0.0\n",
            "09:17:32.882 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
            "09:17:32.883 INFO  HaplotypeCaller - Executing as root@f6921e6d410d on Linux v6.1.85+ amd64\n",
            "09:17:32.883 INFO  HaplotypeCaller - Java runtime: OpenJDK 64-Bit Server VM v17.0.13+11-Ubuntu-2ubuntu122.04\n",
            "09:17:32.883 INFO  HaplotypeCaller - Start Date/Time: November 21, 2024 at 9:17:32 AM UTC\n",
            "09:17:32.883 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "09:17:32.883 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "09:17:32.884 INFO  HaplotypeCaller - HTSJDK Version: 2.24.0\n",
            "09:17:32.884 INFO  HaplotypeCaller - Picard Version: 2.25.0\n",
            "09:17:32.884 INFO  HaplotypeCaller - Built for Spark Version: 2.4.5\n",
            "09:17:32.884 INFO  HaplotypeCaller - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
            "09:17:32.884 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
            "09:17:32.884 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
            "09:17:32.884 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
            "09:17:32.884 INFO  HaplotypeCaller - Deflater: IntelDeflater\n",
            "09:17:32.885 INFO  HaplotypeCaller - Inflater: IntelInflater\n",
            "09:17:32.885 INFO  HaplotypeCaller - GCS max retries/reopens: 20\n",
            "09:17:32.885 INFO  HaplotypeCaller - Requester pays: disabled\n",
            "09:17:32.885 INFO  HaplotypeCaller - Initializing engine\n",
            "WARNING: BAM index file /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bai is older than BAM /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bam\n",
            "09:17:33.225 INFO  IntervalArgumentCollection - Processing 200001 bp from intervals\n",
            "09:17:33.237 INFO  HaplotypeCaller - Done initializing engine\n",
            "09:17:33.239 INFO  HaplotypeCallerEngine - Tool is in reference confidence mode and the annotation, the following changes will be made to any specified annotations: 'StrandBiasBySample' will be enabled. 'ChromosomeCounts', 'FisherStrand', 'StrandOddsRatio' and 'QualByDepth' annotations have been disabled\n",
            "09:17:33.278 INFO  HaplotypeCallerEngine - Standard Emitting and Calling confidence set to 0.0 for reference-model confidence output\n",
            "09:17:33.278 INFO  HaplotypeCallerEngine - All sites annotated with PLs forced to true for reference-model confidence output\n",
            "09:17:33.307 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_utils.so\n",
            "09:17:33.312 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so\n",
            "09:17:33.364 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM\n",
            "09:17:33.364 INFO  IntelPairHmm - Available threads: 2\n",
            "09:17:33.365 INFO  IntelPairHmm - Requested threads: 4\n",
            "09:17:33.365 WARN  IntelPairHmm - Using 2 available threads, but 4 were requested\n",
            "09:17:33.365 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation\n",
            "09:17:33.443 INFO  ProgressMeter - Starting traversal\n",
            "09:17:33.444 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
            "09:17:34.477 WARN  InbreedingCoeff - InbreedingCoeff will not be calculated at position 20:10000117 and possibly subsequent; at least 10 samples must have called genotypes\n",
            "09:17:43.851 INFO  ProgressMeter -          20:10009010              0.2                    80            461.2\n",
            "09:17:54.094 INFO  ProgressMeter -          20:10076737              0.3                   410           1191.3\n",
            "09:18:04.140 INFO  ProgressMeter -          20:10131943              0.5                   720           1407.4\n",
            "09:18:11.656 INFO  HaplotypeCaller - 351 read(s) filtered by: MappingQualityReadFilter \n",
            "0 read(s) filtered by: MappingQualityAvailableReadFilter \n",
            "0 read(s) filtered by: MappedReadFilter \n",
            "0 read(s) filtered by: NotSecondaryAlignmentReadFilter \n",
            "6866 read(s) filtered by: NotDuplicateReadFilter \n",
            "0 read(s) filtered by: PassesVendorQualityCheckReadFilter \n",
            "0 read(s) filtered by: NonZeroReferenceLengthAlignmentReadFilter \n",
            "0 read(s) filtered by: GoodCigarReadFilter \n",
            "0 read(s) filtered by: WellformedReadFilter \n",
            "7217 total reads filtered\n",
            "09:18:11.657 INFO  ProgressMeter -          20:10199356              0.6                  1042           1636.1\n",
            "09:18:11.657 INFO  ProgressMeter - Traversal complete. Processed 1042 total regions in 0.6 minutes.\n",
            "09:18:11.718 INFO  VectorLoglessPairHMM - Time spent in setup for JNI call : 0.069272359\n",
            "09:18:11.718 INFO  PairHMM - Total compute time in PairHMM computeLogLikelihoods() : 3.3781765790000002\n",
            "09:18:11.719 INFO  SmithWatermanAligner - Total compute time in java Smith-Waterman : 2.95 sec\n",
            "09:18:11.721 INFO  HaplotypeCaller - Shutting down engine\n",
            "[November 21, 2024 at 9:18:11 AM UTC] org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller done. Elapsed time: 0.65 minutes.\n",
            "Runtime.totalMemory()=346030080\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/gatk-4.2.0.0/gatk HaplotypeCaller \\\n",
        "-R ${DATA}/ref/Homo_sapiens_assembly19.fasta \\\n",
        "-I ${DATA}/bams/father.bam \\\n",
        "-ERC GVCF \\\n",
        "-L 20:10,000,000-10,200,000 \\\n",
        "-O ${DATA}/variant-discovery/father.g.vcf \\"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hSczSCJbA_eJ",
        "outputId": "c218b175-f846-4dda-8492-8e5b314cf80d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using GATK jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar\n",
            "Running:\n",
            "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar HaplotypeCaller -R /content/drive/MyDrive/GenomeAnalysis/data/ref/Homo_sapiens_assembly19.fasta -I /content/drive/MyDrive/GenomeAnalysis/data/bams/father.bam -ERC GVCF -L 20:10,000,000-10,200,000 -O /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/father.g.vcf \\\n",
            "USAGE: HaplotypeCaller [arguments]\n",
            "\n",
            "Call germline SNPs and indels via local re-assembly of haplotypes\n",
            "Version:4.2.0.0\n",
            "\n",
            "\n",
            "Required Arguments:\n",
            "\n",
            "--input,-I <GATKPath>         BAM/SAM/CRAM file containing reads  This argument must be specified at least once.\n",
            "                              Required. \n",
            "\n",
            "--output,-O <GATKPath>        File to which variants should be written  Required. \n",
            "\n",
            "--reference,-R <GATKPath>     Reference sequence file  Required. \n",
            "\n",
            "\n",
            "Optional Arguments:\n",
            "\n",
            "--add-output-sam-program-record,-add-output-sam-program-record <Boolean>\n",
            "                              If true, adds a PG tag to created SAM/BAM/CRAM files.  Default value: true. Possible\n",
            "                              values: {true, false} \n",
            "\n",
            "--add-output-vcf-command-line,-add-output-vcf-command-line <Boolean>\n",
            "                              If true, adds a command line header line to created VCF files.  Default value: true.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "--alleles <FeatureInput>      The set of alleles to force-call regardless of evidence  Default value: null. \n",
            "\n",
            "--annotate-with-num-discovered-alleles <Boolean>\n",
            "                              If provided, we will annotate records with the number of alternate alleles that were\n",
            "                              discovered (but not necessarily genotyped) at a given site  Default value: false. Possible\n",
            "                              values: {true, false} \n",
            "\n",
            "--annotation,-A <String>      One or more specific annotations to add to variant calls  This argument may be specified 0\n",
            "                              or more times. Default value: null. Possible values: {AlleleFraction,\n",
            "                              AS_BaseQualityRankSumTest, AS_FisherStrand, AS_InbreedingCoeff,\n",
            "                              AS_MappingQualityRankSumTest, AS_QualByDepth, AS_ReadPosRankSumTest, AS_RMSMappingQuality,\n",
            "                              AS_StrandBiasMutectAnnotation, AS_StrandOddsRatio, BaseQuality, BaseQualityHistogram,\n",
            "                              BaseQualityRankSumTest, ChromosomeCounts, ClippingRankSumTest, CountNs, Coverage,\n",
            "                              DepthPerAlleleBySample, DepthPerSampleHC, ExcessHet, FisherStrand, FragmentLength,\n",
            "                              GenotypeSummaries, InbreedingCoeff, LikelihoodRankSumTest, MappingQuality,\n",
            "                              MappingQualityRankSumTest, MappingQualityZero, OrientationBiasReadCounts,\n",
            "                              OriginalAlignment, PossibleDeNovo, QualByDepth, ReadPosition, ReadPosRankSumTest,\n",
            "                              ReferenceBases, RMSMappingQuality, SampleList, StrandBiasBySample, StrandOddsRatio,\n",
            "                              TandemRepeat, UniqueAltReadCountPossible values: {\n",
            "\n",
            "--annotation-group,-G <String>One or more groups of annotations to apply to variant calls  This argument may be\n",
            "                              specified 0 or more times. Default value: null. Possible values:\n",
            "                              {AlleleSpecificAnnotation, AS_StandardAnnotation, ReducibleAnnotation, StandardAnnotation,\n",
            "                              StandardHCAnnotation, StandardMutectAnnotationPossible values: {\n",
            "\n",
            "--annotations-to-exclude,-AX <String>\n",
            "                              One or more specific annotations to exclude from variant calls  This argument may be\n",
            "                              specified 0 or more times. Default value: null. Possible values: {BaseQualityRankSumTest,\n",
            "                              ChromosomeCounts, Coverage, DepthPerAlleleBySample, DepthPerSampleHC, ExcessHet,\n",
            "                              FisherStrand, InbreedingCoeff, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest,\n",
            "                              RMSMappingQuality, StrandOddsRatioPossible values: {\n",
            "\n",
            "--arguments_file <File>       read one or more arguments files and add them to the command line  This argument may be\n",
            "                              specified 0 or more times. Default value: null. \n",
            "\n",
            "--assembly-region-out <String>Output the assembly region to this IGV formatted file  Default value: null. \n",
            "\n",
            "--assembly-region-padding <Integer>\n",
            "                              Number of additional bases of context to include around each assembly region  Default\n",
            "                              value: 100. \n",
            "\n",
            "--base-quality-score-threshold <Byte>\n",
            "                              Base qualities below this threshold will be reduced to the minimum (6)  Default value: 18.\n",
            "\n",
            "--cloud-index-prefetch-buffer,-CIPB <Integer>\n",
            "                              Size of the cloud-only prefetch buffer (in MB; 0 to disable). Defaults to\n",
            "                              cloudPrefetchBuffer if unset.  Default value: -1. \n",
            "\n",
            "--cloud-prefetch-buffer,-CPB <Integer>\n",
            "                              Size of the cloud-only prefetch buffer (in MB; 0 to disable).  Default value: 40. \n",
            "\n",
            "--contamination-fraction-to-filter,-contamination <Double>\n",
            "                              Fraction of contamination in sequencing data (for all samples) to aggressively remove \n",
            "                              Default value: 0.0. \n",
            "\n",
            "--create-output-bam-index,-OBI <Boolean>\n",
            "                              If true, create a BAM/CRAM index when writing a coordinate-sorted BAM/CRAM file.  Default\n",
            "                              value: true. Possible values: {true, false} \n",
            "\n",
            "--create-output-bam-md5,-OBM <Boolean>\n",
            "                              If true, create a MD5 digest for any BAM/SAM/CRAM file created  Default value: false.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "--create-output-variant-index,-OVI <Boolean>\n",
            "                              If true, create a VCF index when writing a coordinate-sorted VCF file.  Default value:\n",
            "                              true. Possible values: {true, false} \n",
            "\n",
            "--create-output-variant-md5,-OVM <Boolean>\n",
            "                              If true, create a a MD5 digest any VCF file created.  Default value: false. Possible\n",
            "                              values: {true, false} \n",
            "\n",
            "--dbsnp,-D <FeatureInput>     dbSNP file  Default value: null. \n",
            "\n",
            "--disable-bam-index-caching,-DBIC <Boolean>\n",
            "                              If true, don't cache bam indexes, this will reduce memory requirements but may harm\n",
            "                              performance if many intervals are specified.  Caching is automatically disabled if there\n",
            "                              are no intervals specified.  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--disable-read-filter,-DF <String>\n",
            "                              Read filters to be disabled before analysis  This argument may be specified 0 or more\n",
            "                              times. Default value: null. Possible values: {GoodCigarReadFilter, MappedReadFilter,\n",
            "                              MappingQualityAvailableReadFilter, MappingQualityReadFilter,\n",
            "                              NonZeroReferenceLengthAlignmentReadFilter, NotDuplicateReadFilter,\n",
            "                              NotSecondaryAlignmentReadFilter, PassesVendorQualityCheckReadFilter,\n",
            "                              WellformedReadFilterPossible values: {\n",
            "\n",
            "--disable-sequence-dictionary-validation,-disable-sequence-dictionary-validation <Boolean>\n",
            "                              If specified, do not check the sequence dictionaries from our inputs for compatibility.\n",
            "                              Use at your own risk!  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--dont-use-dragstr-pair-hmm-scores <Boolean>\n",
            "                              disable DRAGstr pair-hmm score even when dragstr-params-path was provided  Default value:\n",
            "                              false. Possible values: {true, false} \n",
            "\n",
            "--dragen-mode <Boolean>       Single argument for enabling the bulk of DRAGEN-GATK features. NOTE: THIS WILL OVERWRITE\n",
            "                              PROVIDED ARGUMENT CHECK TOOL INFO TO SEE WHICH ARGUMENTS ARE SET).  Default value: false.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "--dragstr-het-hom-ratio <Integer>\n",
            "                              het to hom prior ratio use with DRAGstr on  Default value: 2. \n",
            "\n",
            "--dragstr-params-path <GATKPath>\n",
            "                              location of the DRAGstr model parameters for STR error correction used in the Pair HMM.\n",
            "                              When provided, it overrides other PCR error correcting mechanisms  Default value: null. \n",
            "\n",
            "--enable-dynamic-read-disqualification-for-genotyping <Boolean>\n",
            "                              Will enable less strict read disqualification low base quality reads  Default value:\n",
            "                              false. Possible values: {true, false} \n",
            "\n",
            "--exclude-intervals,-XL <String>\n",
            "                              One or more genomic intervals to exclude from processing  This argument may be specified 0\n",
            "                              or more times. Default value: null. \n",
            "\n",
            "--founder-id,-founder-id <String>\n",
            "                              Samples representing the population \"founders\"  This argument may be specified 0 or more\n",
            "                              times. Default value: null. \n",
            "\n",
            "--gatk-config-file <String>   A configuration file to use with the GATK.  Default value: null. \n",
            "\n",
            "--gcs-max-retries,-gcs-retries <Integer>\n",
            "                              If the GCS bucket channel errors out, how many times it will attempt to re-initiate the\n",
            "                              connection  Default value: 20. \n",
            "\n",
            "--gcs-project-for-requester-pays <String>\n",
            "                              Project to bill when accessing \"requester pays\" buckets. If unset, these buckets cannot be\n",
            "                              accessed.  User must have storage.buckets.get permission on the bucket being accessed. \n",
            "                              Default value: . \n",
            "\n",
            "--genotype-assignment-method,-gam <GenotypeAssignmentMethod>\n",
            "                              How we assign genotypes  Default value: USE_PLS_TO_ASSIGN. Possible values:\n",
            "                              {SET_TO_NO_CALL, USE_PLS_TO_ASSIGN, SET_TO_NO_CALL_NO_ANNOTATIONS, BEST_MATCH_TO_ORIGINAL,\n",
            "                              DO_NOT_ASSIGN_GENOTYPES, USE_POSTERIOR_PROBABILITIES} \n",
            "\n",
            "--graph-output,-graph <String>Write debug assembly graph information to this file  Default value: null. \n",
            "\n",
            "--help,-h <Boolean>           display the help message  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--heterozygosity <Double>     Heterozygosity value used to compute prior likelihoods for any locus.  See the GATKDocs\n",
            "                              for full details on the meaning of this population genetics concept  Default value: 0.001.\n",
            "\n",
            "--heterozygosity-stdev <Double>\n",
            "                              Standard deviation of heterozygosity for SNP and indel calling.  Default value: 0.01. \n",
            "\n",
            "--indel-heterozygosity <Double>\n",
            "                              Heterozygosity for indel calling.  See the GATKDocs for heterozygosity for full details on\n",
            "                              the meaning of this population genetics concept  Default value: 1.25E-4. \n",
            "\n",
            "--interval-exclusion-padding,-ixp <Integer>\n",
            "                              Amount of padding (in bp) to add to each interval you are excluding.  Default value: 0. \n",
            "\n",
            "--interval-merging-rule,-imr <IntervalMergingRule>\n",
            "                              Interval merging rule for abutting intervals  Default value: ALL. Possible values: {ALL,\n",
            "                              OVERLAPPING_ONLY} \n",
            "\n",
            "--interval-padding,-ip <Integer>\n",
            "                              Amount of padding (in bp) to add to each interval you are including.  Default value: 0. \n",
            "\n",
            "--interval-set-rule,-isr <IntervalSetRule>\n",
            "                              Set merging approach to use for combining interval inputs  Default value: UNION. Possible\n",
            "                              values: {UNION, INTERSECTION} \n",
            "\n",
            "--intervals,-L <String>       One or more genomic intervals over which to operate  This argument may be specified 0 or\n",
            "                              more times. Default value: null. \n",
            "\n",
            "--lenient,-LE <Boolean>       Lenient processing of VCF files  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--max-assembly-region-size <Integer>\n",
            "                              Maximum size of an assembly region  Default value: 300. \n",
            "\n",
            "--max-reads-per-alignment-start <Integer>\n",
            "                              Maximum number of reads to retain per alignment start position. Reads above this threshold\n",
            "                              will be downsampled. Set to 0 to disable.  Default value: 50. \n",
            "\n",
            "--min-assembly-region-size <Integer>\n",
            "                              Minimum size of an assembly region  Default value: 50. \n",
            "\n",
            "--min-base-quality-score,-mbq <Byte>\n",
            "                              Minimum base quality required to consider a base for calling  Default value: 10. \n",
            "\n",
            "--native-pair-hmm-threads <Integer>\n",
            "                              How many threads should a native pairHMM implementation use  Default value: 4. \n",
            "\n",
            "--native-pair-hmm-use-double-precision <Boolean>\n",
            "                              use double precision in the native pairHmm. This is slower but matches the java\n",
            "                              implementation better  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--num-reference-samples-if-no-call <Integer>\n",
            "                              Number of hom-ref genotypes to infer at sites not present in a panel  Default value: 0. \n",
            "\n",
            "--output-mode <OutputMode>    Specifies which type of calls we should output  Default value: EMIT_VARIANTS_ONLY.\n",
            "                              Possible values: {EMIT_VARIANTS_ONLY, EMIT_ALL_CONFIDENT_SITES, EMIT_ALL_ACTIVE_SITES} \n",
            "\n",
            "--pedigree,-ped <GATKPath>    Pedigree file for determining the population \"founders\"  Default value: null. \n",
            "\n",
            "--population-callset,-population <FeatureInput>\n",
            "                              Callset to use in calculating genotype priors  Default value: null. \n",
            "\n",
            "--QUIET <Boolean>             Whether to suppress job-summary info on System.err.  Default value: false. Possible\n",
            "                              values: {true, false} \n",
            "\n",
            "--read-filter,-RF <String>    Read filters to be applied before analysis  This argument may be specified 0 or more\n",
            "                              times. Default value: null. Possible values: {AlignmentAgreesWithHeaderReadFilter,\n",
            "                              AllowAllReadsReadFilter, AmbiguousBaseReadFilter, CigarContainsNoNOperator,\n",
            "                              FirstOfPairReadFilter, FragmentLengthReadFilter, GoodCigarReadFilter,\n",
            "                              HasReadGroupReadFilter, IntervalOverlapReadFilter, LibraryReadFilter, MappedReadFilter,\n",
            "                              MappingQualityAvailableReadFilter, MappingQualityNotZeroReadFilter,\n",
            "                              MappingQualityReadFilter, MatchingBasesAndQualsReadFilter, MateDifferentStrandReadFilter,\n",
            "                              MateDistantReadFilter, MateOnSameContigOrNoMappedMateReadFilter,\n",
            "                              MateUnmappedAndUnmappedReadFilter, MetricsReadFilter,\n",
            "                              NonChimericOriginalAlignmentReadFilter, NonZeroFragmentLengthReadFilter,\n",
            "                              NonZeroReferenceLengthAlignmentReadFilter, NotDuplicateReadFilter,\n",
            "                              NotOpticalDuplicateReadFilter, NotProperlyPairedReadFilter,\n",
            "                              NotSecondaryAlignmentReadFilter, NotSupplementaryAlignmentReadFilter,\n",
            "                              OverclippedReadFilter, PairedReadFilter, PassesVendorQualityCheckReadFilter,\n",
            "                              PlatformReadFilter, PlatformUnitReadFilter, PrimaryLineReadFilter,\n",
            "                              ProperlyPairedReadFilter, ReadGroupBlackListReadFilter, ReadGroupReadFilter,\n",
            "                              ReadLengthEqualsCigarLengthReadFilter, ReadLengthReadFilter, ReadNameReadFilter,\n",
            "                              ReadStrandFilter, SampleReadFilter, SecondOfPairReadFilter, SeqIsStoredReadFilter,\n",
            "                              SoftClippedReadFilter, ValidAlignmentEndReadFilter, ValidAlignmentStartReadFilter,\n",
            "                              WellformedReadFilterPossible values: {\n",
            "\n",
            "--read-index,-read-index <GATKPath>\n",
            "                              Indices to use for the read inputs. If specified, an index must be provided for every read\n",
            "                              input and in the same order as the read inputs. If this argument is not specified, the\n",
            "                              path to the index for each input will be inferred automatically.  This argument may be\n",
            "                              specified 0 or more times. Default value: null. \n",
            "\n",
            "--read-validation-stringency,-VS <ValidationStringency>\n",
            "                              Validation stringency for all SAM/BAM/CRAM/SRA files read by this program.  The default\n",
            "                              stringency value SILENT can improve performance when processing a BAM file in which\n",
            "                              variable-length data (read, qualities, tags) do not otherwise need to be decoded.  Default\n",
            "                              value: SILENT. Possible values: {STRICT, LENIENT, SILENT} \n",
            "\n",
            "--recover-dangling-heads <Boolean>\n",
            "                              This argument is deprecated since version 3.3  Default value: false. Possible values:\n",
            "                              {true, false} \n",
            "\n",
            "--sample-name,-ALIAS <String> Name of single sample to use from a multi-sample bam  Default value: null. \n",
            "\n",
            "--sample-ploidy,-ploidy <Integer>\n",
            "                              Ploidy (number of chromosomes) per sample. For pooled data, set to (Number of samples in\n",
            "                              each pool * Sample Ploidy).  Default value: 2. \n",
            "\n",
            "--seconds-between-progress-updates,-seconds-between-progress-updates <Double>\n",
            "                              Output traversal statistics every time this many seconds elapse  Default value: 10.0. \n",
            "\n",
            "--sequence-dictionary,-sequence-dictionary <GATKPath>\n",
            "                              Use the given sequence dictionary as the master/canonical sequence dictionary.  Must be a\n",
            "                              .dict file.  Default value: null. \n",
            "\n",
            "--sites-only-vcf-output <Boolean>\n",
            "                              If true, don't emit genotype fields when writing vcf file output.  Default value: false.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "--standard-min-confidence-threshold-for-calling,-stand-call-conf <Double>\n",
            "                              The minimum phred-scaled confidence threshold at which variants should be called  Default\n",
            "                              value: 30.0. \n",
            "\n",
            "--tmp-dir <GATKPath>          Temp directory to use.  Default value: null. \n",
            "\n",
            "--use-jdk-deflater,-jdk-deflater <Boolean>\n",
            "                              Whether to use the JdkDeflater (as opposed to IntelDeflater)  Default value: false.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "--use-jdk-inflater,-jdk-inflater <Boolean>\n",
            "                              Whether to use the JdkInflater (as opposed to IntelInflater)  Default value: false.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "--use-new-qual-calculator,-new-qual <Boolean>\n",
            "                              Use the new AF model instead of the so-called exact model  Default value: true. Possible\n",
            "                              values: {true, false} \n",
            "\n",
            "--use-posteriors-to-calculate-qual,-gp-qual <Boolean>\n",
            "                              if available, use the genotype posterior probabilities to calculate the site QUAL  Default\n",
            "                              value: false. Possible values: {true, false} \n",
            "\n",
            "--verbosity,-verbosity <LogLevel>\n",
            "                              Control verbosity of logging.  Default value: INFO. Possible values: {ERROR, WARNING,\n",
            "                              INFO, DEBUG} \n",
            "\n",
            "--version <Boolean>           display the version number for this tool  Default value: false. Possible values: {true,\n",
            "                              false} \n",
            "\n",
            "\n",
            "Advanced Arguments:\n",
            "\n",
            "--active-probability-threshold <Double>\n",
            "                              Minimum probability for a locus to be considered active.  Default value: 0.002. \n",
            "\n",
            "--adaptive-pruning <Boolean>  Use Mutect2's adaptive graph pruning algorithm  Default value: false. Possible values:\n",
            "                              {true, false} \n",
            "\n",
            "--adaptive-pruning-initial-error-rate <Double>\n",
            "                              Initial base error rate estimate for adaptive pruning  Default value: 0.001. \n",
            "\n",
            "--all-site-pls <Boolean>      Annotate all sites with PLs  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--allele-informative-reads-overlap-margin <Integer>\n",
            "                              Likelihood and read-based annotations will only take into consideration reads that overlap\n",
            "                              the variant or any base no further than this distance expressed in base pairs  Default\n",
            "                              value: 2. \n",
            "\n",
            "--allow-non-unique-kmers-in-ref <Boolean>\n",
            "                              Allow graphs that have non-unique kmers in the reference  Default value: false. Possible\n",
            "                              values: {true, false} \n",
            "\n",
            "--apply-bqd <Boolean>         If enabled this argument will apply the DRAGEN-GATK BaseQualityDropout model to the\n",
            "                              genotyping model for filtering sites due to Linked Error mode.  Default value: false.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "--apply-frd <Boolean>         If enabled this argument will apply the DRAGEN-GATK ForeignReadDetection model to the\n",
            "                              genotyping model for filtering sites.  Default value: false. Possible values: {true,\n",
            "                              false} \n",
            "\n",
            "--bam-output,-bamout <String> File to which assembled haplotypes should be written  Default value: null. \n",
            "\n",
            "--bam-writer-type <WriterType>Which haplotypes should be written to the BAM  Default value: CALLED_HAPLOTYPES. Possible\n",
            "                              values: {ALL_POSSIBLE_HAPLOTYPES, CALLED_HAPLOTYPES, NO_HAPLOTYPES} \n",
            "\n",
            "--comparison,-comp <FeatureInput>\n",
            "                              Comparison VCF file(s)  This argument may be specified 0 or more times. Default value:\n",
            "                              null. \n",
            "\n",
            "--contamination-fraction-per-sample-file,-contamination-file <File>\n",
            "                              Tab-separated File containing fraction of contamination in sequencing data (per sample) to\n",
            "                              aggressively remove. Format should be \"<SampleID><TAB><Contamination>\" (Contamination is\n",
            "                              double) per line; No header.  Default value: null. \n",
            "\n",
            "--debug-assembly,-debug <Boolean>\n",
            "                              Print out verbose debug information about each assembly region  Default value: false.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "--disable-cap-base-qualities-to-map-quality <Boolean>\n",
            "                              If false this disables capping of base qualities in the HMM to the mapping quality of the\n",
            "                              read  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--disable-optimizations <Boolean>\n",
            "                              Don't skip calculations in ActiveRegions with no variants  Default value: false. Possible\n",
            "                              values: {true, false} \n",
            "\n",
            "--disable-spanning-event-genotyping <Boolean>\n",
            "                              If enabled this argument will disable inclusion of the '*' spanning event when genotyping\n",
            "                              events that overlap deletions  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--disable-symmetric-hmm-normalizing <Boolean>\n",
            "                              Toggle to revive legacy behavior of asymmetrically normalizing the arguments to the\n",
            "                              reference haplotype  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--disable-tool-default-annotations,-disable-tool-default-annotations <Boolean>\n",
            "                              Disable all tool default annotations  Default value: false. Possible values: {true, false}\n",
            "\n",
            "--disable-tool-default-read-filters,-disable-tool-default-read-filters <Boolean>\n",
            "                              Disable all tool default read filters (WARNING: many tools will not function correctly\n",
            "                              without their default read filters on)  Default value: false. Possible values: {true,\n",
            "                              false} \n",
            "\n",
            "--do-not-correct-overlapping-quality <Boolean>\n",
            "                              Disable overlapping base quality correction  Default value: false. Possible values: {true,\n",
            "                              false} \n",
            "\n",
            "--do-not-run-physical-phasing <Boolean>\n",
            "                              Disable physical phasing  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--dont-increase-kmer-sizes-for-cycles <Boolean>\n",
            "                              Disable iterating over kmer sizes when graph cycles are detected  Default value: false.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "--dont-use-dragstr-priors <Boolean>\n",
            "                              Forfeit the use of the DRAGstr model to calculate genotype priors. This argument does not\n",
            "                              have any effect in the absence of DRAGstr model parameters (--dragstr-model-params) \n",
            "                              Default value: false. Possible values: {true, false} \n",
            "\n",
            "--dont-use-soft-clipped-bases <Boolean>\n",
            "                              Do not analyze soft clipped bases in the reads  Default value: false. Possible values:\n",
            "                              {true, false} \n",
            "\n",
            "--emit-ref-confidence,-ERC <ReferenceConfidenceMode>\n",
            "                              Mode for emitting reference confidence scores (For Mutect2, this is a BETA feature) \n",
            "                              Default value: NONE. Possible values: {NONE, BP_RESOLUTION, GVCF} \n",
            "\n",
            "--enable-all-annotations <Boolean>\n",
            "                              Use all possible annotations (not for the faint of heart)  Default value: false. Possible\n",
            "                              values: {true, false} \n",
            "\n",
            "--expected-mismatch-rate-for-read-disqualification <Double>\n",
            "                              Error rate used to set expectation for post HMM read disqualification based on mismatches \n",
            "                              Default value: 0.02. \n",
            "\n",
            "--floor-blocks <Boolean>      Output the band lower bound for each GQ block regardless of the data it represents \n",
            "                              Default value: false. Possible values: {true, false} \n",
            "\n",
            "--force-active <Boolean>      If provided, all regions will be marked as active  Default value: false. Possible values:\n",
            "                              {true, false} \n",
            "\n",
            "--force-call-filtered-alleles,-genotype-filtered-alleles <Boolean>\n",
            "                              Force-call filtered alleles included in the resource specified by --alleles  Default\n",
            "                              value: false. Possible values: {true, false} \n",
            "\n",
            "--gvcf-gq-bands,-GQB <Integer>Exclusive upper bounds for reference confidence GQ bands (must be in [1, 100] and\n",
            "                              specified in increasing order)  This argument may be specified 0 or more times. Default\n",
            "                              value: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,\n",
            "                              24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,\n",
            "                              46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 70, 80, 90, 99]. \n",
            "\n",
            "--indel-size-to-eliminate-in-ref-model <Integer>\n",
            "                              The size of an indel to check for in the reference model  Default value: 10. \n",
            "\n",
            "--kmer-size <Integer>         Kmer size to use in the read threading assembler  This argument may be specified 0 or more\n",
            "                              times. Default value: [10, 25]. \n",
            "\n",
            "--linked-de-bruijn-graph <Boolean>\n",
            "                              If enabled, the Assembly Engine will construct a Linked De Bruijn graph to recover better\n",
            "                              haplotypes  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--mapping-quality-threshold-for-genotyping <Integer>\n",
            "                              Control the threshold for discounting reads from the genotyper due to mapping quality\n",
            "                              after the active region detection and assembly steps but before genotyping. NOTE: this is\n",
            "                              in contrast to the --minimum-mapping-quality argument which filters reads from all parts\n",
            "                              of the HaplotypeCaller. If you would like to call genotypes with a different threshold\n",
            "                              both arguments must be set.  Default value: 20. \n",
            "\n",
            "--max-alternate-alleles <Integer>\n",
            "                              Maximum number of alternate alleles to genotype  Default value: 6. \n",
            "\n",
            "--max-effective-depth-adjustment-for-frd <Integer>\n",
            "                              Set the maximum depth to modify FRD adjustment to in the event of high depth sites (0 to\n",
            "                              disable)  Default value: 0. \n",
            "\n",
            "--max-genotype-count <Integer>Maximum number of genotypes to consider at any site  Default value: 1024. \n",
            "\n",
            "--max-mnp-distance,-mnp-dist <Integer>\n",
            "                              Two or more phased substitutions separated by this distance or less are merged into MNPs. \n",
            "                              Default value: 0. \n",
            "\n",
            "--max-num-haplotypes-in-population <Integer>\n",
            "                              Maximum number of haplotypes to consider for your population  Default value: 128. \n",
            "\n",
            "--max-prob-propagation-distance <Integer>\n",
            "                              Upper limit on how many bases away probability mass can be moved around when calculating\n",
            "                              the boundaries between active and inactive assembly regions  Default value: 50. \n",
            "\n",
            "--max-unpruned-variants <Integer>\n",
            "                              Maximum number of variants in graph the adaptive pruner will allow  Default value: 100. \n",
            "\n",
            "--min-dangling-branch-length <Integer>\n",
            "                              Minimum length of a dangling branch to attempt recovery  Default value: 4. \n",
            "\n",
            "--min-pruning <Integer>       Minimum support to not prune paths in the graph  Default value: 2. \n",
            "\n",
            "--num-pruning-samples <Integer>\n",
            "                              Number of samples that must pass the minPruning threshold  Default value: 1. \n",
            "\n",
            "--pair-hmm-gap-continuation-penalty <Integer>\n",
            "                              Flat gap continuation penalty for use in the Pair HMM  Default value: 10. \n",
            "\n",
            "--pair-hmm-implementation,-pairHMM <Implementation>\n",
            "                              The PairHMM implementation to use for genotype likelihood calculations  Default value:\n",
            "                              FASTEST_AVAILABLE. Possible values: {EXACT, ORIGINAL, LOGLESS_CACHING,\n",
            "                              AVX_LOGLESS_CACHING, AVX_LOGLESS_CACHING_OMP, EXPERIMENTAL_FPGA_LOGLESS_CACHING,\n",
            "                              FASTEST_AVAILABLE} \n",
            "\n",
            "--pcr-indel-model <PCRErrorModel>\n",
            "                              The PCR indel model to use  Default value: CONSERVATIVE. Possible values: {NONE, HOSTILE,\n",
            "                              AGGRESSIVE, CONSERVATIVE} \n",
            "\n",
            "--phred-scaled-global-read-mismapping-rate <Integer>\n",
            "                              The global assumed mismapping rate for reads  Default value: 45. \n",
            "\n",
            "--pruning-lod-threshold <Double>\n",
            "                              Ln likelihood ratio threshold for adaptive pruning algorithm  Default value:\n",
            "                              2.302585092994046. \n",
            "\n",
            "--pruning-seeding-lod-threshold <Double>\n",
            "                              Ln likelihood ratio threshold for seeding subgraph of good variation in adaptive pruning\n",
            "                              algorithm  Default value: 9.210340371976184. \n",
            "\n",
            "--recover-all-dangling-branches <Boolean>\n",
            "                              Recover all dangling branches  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--showHidden,-showHidden <Boolean>\n",
            "                              display hidden arguments  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--smith-waterman <Implementation>\n",
            "                              Which Smith-Waterman implementation to use, generally FASTEST_AVAILABLE is the right\n",
            "                              choice  Default value: JAVA. Possible values: {FASTEST_AVAILABLE, AVX_ENABLED, JAVA} \n",
            "\n",
            "--soft-clip-low-quality-ends <Boolean>\n",
            "                              If enabled will preserve low-quality read ends as softclips (used for DRAGEN-GATK BQD\n",
            "                              genotyper model)  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--transform-dragen-mapping-quality <Boolean>\n",
            "                              If enabled this argument will map DRAGEN aligner aligned reads with mapping quality <=250\n",
            "                              to scale up to MQ 50  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--use-filtered-reads-for-annotations <Boolean>\n",
            "                              Use the contamination-filtered read maps for the purposes of annotating variants  Default\n",
            "                              value: false. Possible values: {true, false} \n",
            "\n",
            "Conditional Arguments for annotation:\n",
            "\n",
            "Valid only if \"RMSMappingQuality\" is specified:\n",
            "--allow-old-rms-mapping-quality-annotation-data <Boolean>\n",
            "                              Override to allow old RMSMappingQuality annotated VCFs to function  Default value: false.\n",
            "                              Possible values: {true, false} \n",
            "\n",
            "Conditional Arguments for readFilter:\n",
            "\n",
            "Valid only if \"AmbiguousBaseReadFilter\" is specified:\n",
            "--ambig-filter-bases <Integer>Threshold number of ambiguous bases. If null, uses threshold fraction; otherwise,\n",
            "                              overrides threshold fraction.  Default value: null.  Cannot be used in conjunction with\n",
            "                              argument(s) maxAmbiguousBaseFraction\n",
            "\n",
            "--ambig-filter-frac <Double>  Threshold fraction of ambiguous bases  Default value: 0.05.  Cannot be used in conjunction\n",
            "                              with argument(s) maxAmbiguousBases\n",
            "\n",
            "Valid only if \"FragmentLengthReadFilter\" is specified:\n",
            "--max-fragment-length <Integer>\n",
            "                              Maximum length of fragment (insert size)  Default value: 1000000. \n",
            "\n",
            "--min-fragment-length <Integer>\n",
            "                              Minimum length of fragment (insert size)  Default value: 0. \n",
            "\n",
            "Valid only if \"IntervalOverlapReadFilter\" is specified:\n",
            "--keep-intervals <String>     One or more genomic intervals to keep  This argument must be specified at least once.\n",
            "                              Required. \n",
            "\n",
            "Valid only if \"LibraryReadFilter\" is specified:\n",
            "--library,-library <String>   Name of the library to keep  This argument must be specified at least once. Required. \n",
            "\n",
            "Valid only if \"MappingQualityReadFilter\" is specified:\n",
            "--maximum-mapping-quality <Integer>\n",
            "                              Maximum mapping quality to keep (inclusive)  Default value: null. \n",
            "\n",
            "--minimum-mapping-quality <Integer>\n",
            "                              Minimum mapping quality to keep (inclusive)  Default value: 20. \n",
            "\n",
            "Valid only if \"MateDistantReadFilter\" is specified:\n",
            "--mate-too-distant-length <Integer>\n",
            "                              Minimum start location difference at which mapped mates are considered distant  Default\n",
            "                              value: 1000. \n",
            "\n",
            "Valid only if \"OverclippedReadFilter\" is specified:\n",
            "--dont-require-soft-clips-both-ends <Boolean>\n",
            "                              Allow a read to be filtered out based on having only 1 soft-clipped block. By default,\n",
            "                              both ends must have a soft-clipped block, setting this flag requires only 1 soft-clipped\n",
            "                              block  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--filter-too-short <Integer>  Minimum number of aligned bases  Default value: 30. \n",
            "\n",
            "Valid only if \"PlatformReadFilter\" is specified:\n",
            "--platform-filter-name <String>\n",
            "                              Platform attribute (PL) to match  This argument must be specified at least once. Required.\n",
            "\n",
            "Valid only if \"PlatformUnitReadFilter\" is specified:\n",
            "--black-listed-lanes <String> Platform unit (PU) to filter out  This argument must be specified at least once. Required.\n",
            "\n",
            "Valid only if \"ReadGroupBlackListReadFilter\" is specified:\n",
            "--read-group-black-list <String>\n",
            "                              A read group filter expression in the form \"attribute:value\", where \"attribute\" is a two\n",
            "                              character read group attribute such as \"RG\" or \"PU\".  This argument must be specified at\n",
            "                              least once. Required. \n",
            "\n",
            "Valid only if \"ReadGroupReadFilter\" is specified:\n",
            "--keep-read-group <String>    The name of the read group to keep  Required. \n",
            "\n",
            "Valid only if \"ReadLengthReadFilter\" is specified:\n",
            "--max-read-length <Integer>   Keep only reads with length at most equal to the specified value  Required. \n",
            "\n",
            "--min-read-length <Integer>   Keep only reads with length at least equal to the specified value  Default value: 1. \n",
            "\n",
            "Valid only if \"ReadNameReadFilter\" is specified:\n",
            "--read-name <String>          Keep only reads with this read name  Required. \n",
            "\n",
            "Valid only if \"ReadStrandFilter\" is specified:\n",
            "--keep-reverse-strand-only <Boolean>\n",
            "                              Keep only reads on the reverse strand  Required. Possible values: {true, false} \n",
            "\n",
            "Valid only if \"SampleReadFilter\" is specified:\n",
            "--sample,-sample <String>     The name of the sample(s) to keep, filtering out all others  This argument must be\n",
            "                              specified at least once. Required. \n",
            "\n",
            "Valid only if \"SoftClippedReadFilter\" is specified:\n",
            "--invert-soft-clip-ratio-filter <Boolean>\n",
            "                              Inverts the results from this filter, causing all variants that would pass to fail and\n",
            "                              visa-versa.  Default value: false. Possible values: {true, false} \n",
            "\n",
            "--soft-clipped-leading-trailing-ratio <Double>\n",
            "                              Threshold ratio of soft clipped bases (leading / trailing the cigar string) to total bases\n",
            "                              in read for read to be filtered.  Default value: null.  Cannot be used in conjunction with\n",
            "                              argument(s) minimumSoftClippedRatio\n",
            "\n",
            "--soft-clipped-ratio-threshold <Double>\n",
            "                              Threshold ratio of soft clipped bases (anywhere in the cigar string) to total bases in\n",
            "                              read for read to be filtered.  Default value: null.  Cannot be used in conjunction with\n",
            "                              argument(s) minimumLeadingTrailingSoftClippedRatio\n",
            "\n",
            "\n",
            "***********************************************************************\n",
            "\n",
            "A USER ERROR has occurred: Illegal argument value: Positional arguments were provided ',\\}' but no positional argument is defined for this tool.\n",
            "\n",
            "***********************************************************************\n",
            "Set the system property GATK_STACKTRACE_ON_USER_EXCEPTION (--java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true') to print the stack trace.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/gatk-4.2.0.0/gatk HaplotypeCaller \\\n",
        "-R ${DATA}/ref/Homo_sapiens_assembly19.fasta \\\n",
        "-I ${DATA}/bams/son.bam \\\n",
        "-ERC GVCF \\\n",
        "-L 20:10,000,000-10,200,000 \\\n",
        "-O ${DATA}/variant-discovery/son.g.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xNklGWNYBN-D",
        "outputId": "d1c28399-6bd0-4257-cad1-d3d268b2ecea"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using GATK jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar\n",
            "Running:\n",
            "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar HaplotypeCaller -R /content/drive/MyDrive/GenomeAnalysis/data/ref/Homo_sapiens_assembly19.fasta -I /content/drive/MyDrive/GenomeAnalysis/data/bams/son.bam -ERC GVCF -L 20:10,000,000-10,200,000 -O /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/son.g.vcf\n",
            "09:19:48.917 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "09:19:49.160 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "09:19:49.160 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.2.0.0\n",
            "09:19:49.160 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
            "09:19:49.160 INFO  HaplotypeCaller - Executing as root@f6921e6d410d on Linux v6.1.85+ amd64\n",
            "09:19:49.161 INFO  HaplotypeCaller - Java runtime: OpenJDK 64-Bit Server VM v17.0.13+11-Ubuntu-2ubuntu122.04\n",
            "09:19:49.161 INFO  HaplotypeCaller - Start Date/Time: November 21, 2024 at 9:19:48 AM UTC\n",
            "09:19:49.161 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "09:19:49.161 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "09:19:49.162 INFO  HaplotypeCaller - HTSJDK Version: 2.24.0\n",
            "09:19:49.162 INFO  HaplotypeCaller - Picard Version: 2.25.0\n",
            "09:19:49.162 INFO  HaplotypeCaller - Built for Spark Version: 2.4.5\n",
            "09:19:49.162 INFO  HaplotypeCaller - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
            "09:19:49.162 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
            "09:19:49.162 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
            "09:19:49.162 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
            "09:19:49.162 INFO  HaplotypeCaller - Deflater: IntelDeflater\n",
            "09:19:49.163 INFO  HaplotypeCaller - Inflater: IntelInflater\n",
            "09:19:49.163 INFO  HaplotypeCaller - GCS max retries/reopens: 20\n",
            "09:19:49.163 INFO  HaplotypeCaller - Requester pays: disabled\n",
            "09:19:49.163 INFO  HaplotypeCaller - Initializing engine\n",
            "WARNING: BAM index file /content/drive/MyDrive/GenomeAnalysis/data/bams/son.bai is older than BAM /content/drive/MyDrive/GenomeAnalysis/data/bams/son.bam\n",
            "09:19:49.521 INFO  IntervalArgumentCollection - Processing 200001 bp from intervals\n",
            "09:19:49.539 INFO  HaplotypeCaller - Done initializing engine\n",
            "09:19:49.543 INFO  HaplotypeCallerEngine - Tool is in reference confidence mode and the annotation, the following changes will be made to any specified annotations: 'StrandBiasBySample' will be enabled. 'ChromosomeCounts', 'FisherStrand', 'StrandOddsRatio' and 'QualByDepth' annotations have been disabled\n",
            "09:19:49.568 INFO  HaplotypeCallerEngine - Standard Emitting and Calling confidence set to 0.0 for reference-model confidence output\n",
            "09:19:49.568 INFO  HaplotypeCallerEngine - All sites annotated with PLs forced to true for reference-model confidence output\n",
            "09:19:49.585 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_utils.so\n",
            "09:19:49.588 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so\n",
            "09:19:49.632 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM\n",
            "09:19:49.633 INFO  IntelPairHmm - Available threads: 2\n",
            "09:19:49.634 INFO  IntelPairHmm - Requested threads: 4\n",
            "09:19:49.634 WARN  IntelPairHmm - Using 2 available threads, but 4 were requested\n",
            "09:19:49.634 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation\n",
            "09:19:49.705 INFO  ProgressMeter - Starting traversal\n",
            "09:19:49.705 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
            "09:19:52.225 WARN  InbreedingCoeff - InbreedingCoeff will not be calculated at position 20:10000117 and possibly subsequent; at least 10 samples must have called genotypes\n",
            "09:20:00.058 INFO  ProgressMeter -          20:10013089              0.2                   110            637.5\n",
            "09:20:10.302 INFO  ProgressMeter -          20:10092753              0.3                   470           1369.1\n",
            "09:20:20.575 INFO  ProgressMeter -          20:10173215              0.5                   910           1768.7\n",
            "09:20:24.457 INFO  HaplotypeCaller - 420 read(s) filtered by: MappingQualityReadFilter \n",
            "0 read(s) filtered by: MappingQualityAvailableReadFilter \n",
            "0 read(s) filtered by: MappedReadFilter \n",
            "0 read(s) filtered by: NotSecondaryAlignmentReadFilter \n",
            "7031 read(s) filtered by: NotDuplicateReadFilter \n",
            "0 read(s) filtered by: PassesVendorQualityCheckReadFilter \n",
            "0 read(s) filtered by: NonZeroReferenceLengthAlignmentReadFilter \n",
            "0 read(s) filtered by: GoodCigarReadFilter \n",
            "0 read(s) filtered by: WellformedReadFilter \n",
            "7451 total reads filtered\n",
            "09:20:24.457 INFO  ProgressMeter -          20:10199742              0.6                  1050           1812.8\n",
            "09:20:24.458 INFO  ProgressMeter - Traversal complete. Processed 1050 total regions in 0.6 minutes.\n",
            "09:20:24.483 INFO  VectorLoglessPairHMM - Time spent in setup for JNI call : 0.043787881\n",
            "09:20:24.483 INFO  PairHMM - Total compute time in PairHMM computeLogLikelihoods() : 2.761567404\n",
            "09:20:24.483 INFO  SmithWatermanAligner - Total compute time in java Smith-Waterman : 2.33 sec\n",
            "09:20:24.483 INFO  HaplotypeCaller - Shutting down engine\n",
            "[November 21, 2024 at 9:20:24 AM UTC] org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller done. Elapsed time: 0.59 minutes.\n",
            "Runtime.totalMemory()=379584512\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cat ${DATA}/variant-discovery/mother.g.vcf |grep -v \"##\" |head -n 10"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7gu3eJcKBju0",
        "outputId": "1eb5d369-12ef-4729-86f9-254a51185f80"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tNA12878\n",
            "20\t10000000\t.\tT\t<NON_REF>\t.\t.\tEND=10000008\tGT:DP:GQ:MIN_DP:PL\t0/0:16:48:16:0,48,497\n",
            "20\t10000009\t.\tA\t<NON_REF>\t.\t.\tEND=10000009\tGT:DP:GQ:MIN_DP:PL\t0/0:16:32:16:0,32,485\n",
            "20\t10000010\t.\tC\t<NON_REF>\t.\t.\tEND=10000017\tGT:DP:GQ:MIN_DP:PL\t0/0:16:48:16:0,48,472\n",
            "20\t10000018\t.\tT\t<NON_REF>\t.\t.\tEND=10000022\tGT:DP:GQ:MIN_DP:PL\t0/0:16:45:16:0,45,675\n",
            "20\t10000023\t.\tA\t<NON_REF>\t.\t.\tEND=10000026\tGT:DP:GQ:MIN_DP:PL\t0/0:17:48:17:0,48,720\n",
            "20\t10000027\t.\tT\t<NON_REF>\t.\t.\tEND=10000027\tGT:DP:GQ:MIN_DP:PL\t0/0:18:51:18:0,51,765\n",
            "20\t10000028\t.\tT\t<NON_REF>\t.\t.\tEND=10000031\tGT:DP:GQ:MIN_DP:PL\t0/0:17:48:17:0,48,720\n",
            "20\t10000032\t.\tC\t<NON_REF>\t.\t.\tEND=10000032\tGT:DP:GQ:MIN_DP:PL\t0/0:17:34:17:0,34,448\n",
            "20\t10000033\t.\tT\t<NON_REF>\t.\t.\tEND=10000034\tGT:DP:GQ:MIN_DP:PL\t0/0:17:48:17:0,48,720\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Sử dụng GenomicsDBImport để lưu dần các file g.vcf (Consolidate GVCFs)**"
      ],
      "metadata": {
        "id": "eOORrAGnBrdr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/gatk-4.2.0.0/gatk GenomicsDBImport \\\n",
        "-V ${DATA}/variant-discovery/mother.g.vcf \\\n",
        "-V ${DATA}/variant-discovery/father.g.vcf \\\n",
        "--genomicsdb-workspace-path ${DATA}/variant-discovery/trio \\\n",
        "--intervals 20:10,000,000-10,200,000"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7QGLAZm3Bt9E",
        "outputId": "e7c2f5d1-2b0b-4084-add8-ee1b8170513c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using GATK jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar\n",
            "Running:\n",
            "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar GenomicsDBImport -V /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/mother.g.vcf -V /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/father.g.vcf --genomicsdb-workspace-path /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/trio --intervals 20:10,000,000-10,200,000\n",
            "09:21:56.020 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "09:21:56.390 INFO  GenomicsDBImport - ------------------------------------------------------------\n",
            "09:21:56.391 INFO  GenomicsDBImport - The Genome Analysis Toolkit (GATK) v4.2.0.0\n",
            "09:21:56.391 INFO  GenomicsDBImport - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
            "09:21:56.391 INFO  GenomicsDBImport - Executing as root@f6921e6d410d on Linux v6.1.85+ amd64\n",
            "09:21:56.391 INFO  GenomicsDBImport - Java runtime: OpenJDK 64-Bit Server VM v17.0.13+11-Ubuntu-2ubuntu122.04\n",
            "09:21:56.392 INFO  GenomicsDBImport - Start Date/Time: November 21, 2024 at 9:21:55 AM UTC\n",
            "09:21:56.392 INFO  GenomicsDBImport - ------------------------------------------------------------\n",
            "09:21:56.392 INFO  GenomicsDBImport - ------------------------------------------------------------\n",
            "09:21:56.393 INFO  GenomicsDBImport - HTSJDK Version: 2.24.0\n",
            "09:21:56.393 INFO  GenomicsDBImport - Picard Version: 2.25.0\n",
            "09:21:56.393 INFO  GenomicsDBImport - Built for Spark Version: 2.4.5\n",
            "09:21:56.393 INFO  GenomicsDBImport - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
            "09:21:56.393 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
            "09:21:56.393 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
            "09:21:56.393 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
            "09:21:56.393 INFO  GenomicsDBImport - Deflater: IntelDeflater\n",
            "09:21:56.393 INFO  GenomicsDBImport - Inflater: IntelInflater\n",
            "09:21:56.394 INFO  GenomicsDBImport - GCS max retries/reopens: 20\n",
            "09:21:56.394 INFO  GenomicsDBImport - Requester pays: disabled\n",
            "09:21:56.394 INFO  GenomicsDBImport - Initializing engine\n",
            "09:21:56.523 INFO  GenomicsDBImport - Shutting down engine\n",
            "[November 21, 2024 at 9:21:56 AM UTC] org.broadinstitute.hellbender.tools.genomicsdb.GenomicsDBImport done. Elapsed time: 0.01 minutes.\n",
            "Runtime.totalMemory()=119537664\n",
            "***********************************************************************\n",
            "\n",
            "A USER ERROR has occurred: Failed to create reader from file:///content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/father.g.vcf\n",
            "\n",
            "***********************************************************************\n",
            "Set the system property GATK_STACKTRACE_ON_USER_EXCEPTION (--java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true') to print the stack trace.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/gatk-4.2.0.0/gatk GenomicsDBImport \\\n",
        "-V ${DATA}/variant-discovery/son.g.vcf \\\n",
        "--genomicsdb-update-workspace-path ${DATA}/variant-discovery/trio \\\n",
        "--intervals 20:10,000,000-10,200,000"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4WIybyrsB5qv",
        "outputId": "b2b5bd68-16a2-47bd-f565-984373c96ae5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using GATK jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar\n",
            "Running:\n",
            "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar GenomicsDBImport -V /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/son.g.vcf --genomicsdb-update-workspace-path /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/trio --intervals 20:10,000,000-10,200,000\n",
            "09:22:34.057 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "09:22:34.364 INFO  GenomicsDBImport - ------------------------------------------------------------\n",
            "09:22:34.366 INFO  GenomicsDBImport - The Genome Analysis Toolkit (GATK) v4.2.0.0\n",
            "09:22:34.366 INFO  GenomicsDBImport - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
            "09:22:34.366 INFO  GenomicsDBImport - Executing as root@f6921e6d410d on Linux v6.1.85+ amd64\n",
            "09:22:34.366 INFO  GenomicsDBImport - Java runtime: OpenJDK 64-Bit Server VM v17.0.13+11-Ubuntu-2ubuntu122.04\n",
            "09:22:34.367 INFO  GenomicsDBImport - Start Date/Time: November 21, 2024 at 9:22:33 AM UTC\n",
            "09:22:34.367 INFO  GenomicsDBImport - ------------------------------------------------------------\n",
            "09:22:34.367 INFO  GenomicsDBImport - ------------------------------------------------------------\n",
            "09:22:34.371 INFO  GenomicsDBImport - HTSJDK Version: 2.24.0\n",
            "09:22:34.371 INFO  GenomicsDBImport - Picard Version: 2.25.0\n",
            "09:22:34.371 INFO  GenomicsDBImport - Built for Spark Version: 2.4.5\n",
            "09:22:34.371 INFO  GenomicsDBImport - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
            "09:22:34.372 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
            "09:22:34.372 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
            "09:22:34.372 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
            "09:22:34.372 INFO  GenomicsDBImport - Deflater: IntelDeflater\n",
            "09:22:34.372 INFO  GenomicsDBImport - Inflater: IntelInflater\n",
            "09:22:34.372 INFO  GenomicsDBImport - GCS max retries/reopens: 20\n",
            "09:22:34.372 INFO  GenomicsDBImport - Requester pays: disabled\n",
            "09:22:34.372 INFO  GenomicsDBImport - Initializing engine\n",
            "09:22:34.502 WARN  GenomicsDBImport - genomicsdb-update-workspace-path was set, so ignoring specified intervals.The tool will use the intervals specified by the initial import\n",
            "09:22:35.426 INFO  GenomicsDBLibLoader - GenomicsDB native library version : 1.3.2-e18fa63\n",
            "[TileDB::FileSystem] Error: (get_dirs) Cannot open directory; path=/content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/trio; errno=2(No such file or directory)\n",
            "09:22:35.779 INFO  IntervalArgumentCollection - Processing 200001 bp from intervals\n",
            "09:22:35.787 INFO  GenomicsDBImport - Done initializing engine\n",
            "09:22:37.024 INFO  GenomicsDBImport - Shutting down engine\n",
            "[November 21, 2024 at 9:22:37 AM UTC] org.broadinstitute.hellbender.tools.genomicsdb.GenomicsDBImport done. Elapsed time: 0.05 minutes.\n",
            "Runtime.totalMemory()=119537664\n",
            "***********************************************************************\n",
            "\n",
            "A USER ERROR has occurred: We require an existing valid workspace when incremental import is set\n",
            "\n",
            "***********************************************************************\n",
            "Set the system property GATK_STACKTRACE_ON_USER_EXCEPTION (--java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true') to print the stack trace.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ln -s ${DATA}/variant-discovery variant-discovery"
      ],
      "metadata": {
        "id": "ERXh_mVaCItx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/gatk-4.2.0.0/gatk SelectVariants \\\n",
        "-R ${DATA}/ref/Homo_sapiens_assembly19.fasta \\\n",
        "-V gendb://variant-discovery/trio \\\n",
        "-O ${DATA}/variant-discovery/trio_selectvariants.g.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hqoh2svDCQkH",
        "outputId": "313066c9-ef03-4337-ed0a-a4b29a9c68d9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using GATK jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar\n",
            "Running:\n",
            "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar SelectVariants -R /content/drive/MyDrive/GenomeAnalysis/data/ref/Homo_sapiens_assembly19.fasta -V gendb://variant-discovery/trio -O /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/trio_selectvariants.g.vcf\n",
            "09:24:04.378 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "09:24:04.727 INFO  SelectVariants - ------------------------------------------------------------\n",
            "09:24:04.728 INFO  SelectVariants - The Genome Analysis Toolkit (GATK) v4.2.0.0\n",
            "09:24:04.728 INFO  SelectVariants - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
            "09:24:04.728 INFO  SelectVariants - Executing as root@f6921e6d410d on Linux v6.1.85+ amd64\n",
            "09:24:04.728 INFO  SelectVariants - Java runtime: OpenJDK 64-Bit Server VM v17.0.13+11-Ubuntu-2ubuntu122.04\n",
            "09:24:04.728 INFO  SelectVariants - Start Date/Time: November 21, 2024 at 9:24:04 AM UTC\n",
            "09:24:04.728 INFO  SelectVariants - ------------------------------------------------------------\n",
            "09:24:04.729 INFO  SelectVariants - ------------------------------------------------------------\n",
            "09:24:04.730 INFO  SelectVariants - HTSJDK Version: 2.24.0\n",
            "09:24:04.730 INFO  SelectVariants - Picard Version: 2.25.0\n",
            "09:24:04.730 INFO  SelectVariants - Built for Spark Version: 2.4.5\n",
            "09:24:04.730 INFO  SelectVariants - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
            "09:24:04.732 INFO  SelectVariants - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
            "09:24:04.732 INFO  SelectVariants - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
            "09:24:04.732 INFO  SelectVariants - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
            "09:24:04.732 INFO  SelectVariants - Deflater: IntelDeflater\n",
            "09:24:04.733 INFO  SelectVariants - Inflater: IntelInflater\n",
            "09:24:04.733 INFO  SelectVariants - GCS max retries/reopens: 20\n",
            "09:24:04.733 INFO  SelectVariants - Requester pays: disabled\n",
            "09:24:04.733 INFO  SelectVariants - Initializing engine\n",
            "09:24:05.041 INFO  SelectVariants - Shutting down engine\n",
            "[November 21, 2024 at 9:24:05 AM UTC] org.broadinstitute.hellbender.tools.walkers.variantutils.SelectVariants done. Elapsed time: 0.01 minutes.\n",
            "Runtime.totalMemory()=119537664\n",
            "***********************************************************************\n",
            "\n",
            "A USER ERROR has occurred: Couldn't read file file:///content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/variant-discovery/trio/callset.json. Error was: It doesn't exist.\n",
            "\n",
            "***********************************************************************\n",
            "Set the system property GATK_STACKTRACE_ON_USER_EXCEPTION (--java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true') to print the stack trace.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/gatk-4.2.0.0/gatk GenotypeGVCFs \\\n",
        "-R ${DATA}/ref/Homo_sapiens_assembly19.fasta \\\n",
        "-V gendb://variant-discovery/trio \\\n",
        "-O ${DATA}/variant-discovery/trioGGVCF.vcf \\\n",
        "-L 20:10,000,000-10,200,000"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0YHaxAM_CSsP",
        "outputId": "48b8eaa2-bddd-41f3-b83a-fbb74639649c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using GATK jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar\n",
            "Running:\n",
            "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar GenotypeGVCFs -R /content/drive/MyDrive/GenomeAnalysis/data/ref/Homo_sapiens_assembly19.fasta -V gendb://variant-discovery/trio -O /content/drive/MyDrive/GenomeAnalysis/data/variant-discovery/trioGGVCF.vcf -L 20:10,000,000-10,200,000\n",
            "09:24:56.669 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "09:24:57.111 INFO  GenotypeGVCFs - ------------------------------------------------------------\n",
            "09:24:57.112 INFO  GenotypeGVCFs - The Genome Analysis Toolkit (GATK) v4.2.0.0\n",
            "09:24:57.112 INFO  GenotypeGVCFs - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
            "09:24:57.112 INFO  GenotypeGVCFs - Executing as root@f6921e6d410d on Linux v6.1.85+ amd64\n",
            "09:24:57.112 INFO  GenotypeGVCFs - Java runtime: OpenJDK 64-Bit Server VM v17.0.13+11-Ubuntu-2ubuntu122.04\n",
            "09:24:57.112 INFO  GenotypeGVCFs - Start Date/Time: November 21, 2024 at 9:24:56 AM UTC\n",
            "09:24:57.112 INFO  GenotypeGVCFs - ------------------------------------------------------------\n",
            "09:24:57.112 INFO  GenotypeGVCFs - ------------------------------------------------------------\n",
            "09:24:57.117 INFO  GenotypeGVCFs - HTSJDK Version: 2.24.0\n",
            "09:24:57.117 INFO  GenotypeGVCFs - Picard Version: 2.25.0\n",
            "09:24:57.117 INFO  GenotypeGVCFs - Built for Spark Version: 2.4.5\n",
            "09:24:57.117 INFO  GenotypeGVCFs - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
            "09:24:57.118 INFO  GenotypeGVCFs - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
            "09:24:57.118 INFO  GenotypeGVCFs - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
            "09:24:57.118 INFO  GenotypeGVCFs - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
            "09:24:57.118 INFO  GenotypeGVCFs - Deflater: IntelDeflater\n",
            "09:24:57.118 INFO  GenotypeGVCFs - Inflater: IntelInflater\n",
            "09:24:57.118 INFO  GenotypeGVCFs - GCS max retries/reopens: 20\n",
            "09:24:57.118 INFO  GenotypeGVCFs - Requester pays: disabled\n",
            "09:24:57.118 INFO  GenotypeGVCFs - Initializing engine\n",
            "09:24:57.554 INFO  GenotypeGVCFs - Shutting down engine\n",
            "[November 21, 2024 at 9:24:57 AM UTC] org.broadinstitute.hellbender.tools.walkers.GenotypeGVCFs done. Elapsed time: 0.02 minutes.\n",
            "Runtime.totalMemory()=130023424\n",
            "***********************************************************************\n",
            "\n",
            "A USER ERROR has occurred: Couldn't read file file:///content/drive/MyDrive/GenomeAnalysis/tools/bcftools-1.21/variant-discovery/trio/callset.json. Error was: It doesn't exist.\n",
            "\n",
            "***********************************************************************\n",
            "Set the system property GATK_STACKTRACE_ON_USER_EXCEPTION (--java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true') to print the stack trace.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!${TOOLS}/gatk-4.2.0.0/gatk HaplotypeCaller \\\n",
        "-R ${DATA}/ref/Homo_sapiens_assembly19.fasta \\\n",
        "-I ${DATA}/bams/mother.bam \\\n",
        "-I ${DATA}/bams/father.bam \\\n",
        "-I ${DATA}/bams/son.bam \\\n",
        "-O variant-discovery/trio_hcjoint_nq.vcf \\\n",
        "-L 20:10,000,000-10,200,000 \\\n",
        "-new-qual \\\n",
        "-bamout variant-discovery/trio_hcjoint_nq.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qZ3S6gJVFrgQ",
        "outputId": "52110e06-3805-47d5-bc3a-55d098e9a9c0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using GATK jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar\n",
            "Running:\n",
            "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar HaplotypeCaller -R /content/drive/MyDrive/GenomeAnalysis/data/ref/Homo_sapiens_assembly19.fasta -I /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bam -I /content/drive/MyDrive/GenomeAnalysis/data/bams/father.bam -I /content/drive/MyDrive/GenomeAnalysis/data/bams/son.bam -O variant-discovery/trio_hcjoint_nq.vcf -L 20:10,000,000-10,200,000 -new-qual -bamout variant-discovery/trio_hcjoint_nq.bam\n",
            "09:39:03.824 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "09:39:04.052 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "09:39:04.052 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.2.0.0\n",
            "09:39:04.053 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
            "09:39:04.053 INFO  HaplotypeCaller - Executing as root@f6921e6d410d on Linux v6.1.85+ amd64\n",
            "09:39:04.053 INFO  HaplotypeCaller - Java runtime: OpenJDK 64-Bit Server VM v17.0.13+11-Ubuntu-2ubuntu122.04\n",
            "09:39:04.053 INFO  HaplotypeCaller - Start Date/Time: November 21, 2024 at 9:39:03 AM UTC\n",
            "09:39:04.053 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "09:39:04.053 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
            "09:39:04.054 INFO  HaplotypeCaller - HTSJDK Version: 2.24.0\n",
            "09:39:04.054 INFO  HaplotypeCaller - Picard Version: 2.25.0\n",
            "09:39:04.054 INFO  HaplotypeCaller - Built for Spark Version: 2.4.5\n",
            "09:39:04.054 INFO  HaplotypeCaller - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
            "09:39:04.054 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
            "09:39:04.054 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
            "09:39:04.054 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
            "09:39:04.055 INFO  HaplotypeCaller - Deflater: IntelDeflater\n",
            "09:39:04.055 INFO  HaplotypeCaller - Inflater: IntelInflater\n",
            "09:39:04.055 INFO  HaplotypeCaller - GCS max retries/reopens: 20\n",
            "09:39:04.055 INFO  HaplotypeCaller - Requester pays: disabled\n",
            "09:39:04.055 INFO  HaplotypeCaller - Initializing engine\n",
            "WARNING: BAM index file /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bai is older than BAM /content/drive/MyDrive/GenomeAnalysis/data/bams/mother.bam\n",
            "WARNING: BAM index file /content/drive/MyDrive/GenomeAnalysis/data/bams/father.bai is older than BAM /content/drive/MyDrive/GenomeAnalysis/data/bams/father.bam\n",
            "WARNING: BAM index file /content/drive/MyDrive/GenomeAnalysis/data/bams/son.bai is older than BAM /content/drive/MyDrive/GenomeAnalysis/data/bams/son.bam\n",
            "09:39:04.507 INFO  IntervalArgumentCollection - Processing 200001 bp from intervals\n",
            "09:39:04.523 INFO  HaplotypeCaller - Done initializing engine\n",
            "09:39:04.565 INFO  HaplotypeCallerEngine - Disabling physical phasing, which is supported only for reference-model confidence output\n",
            "09:39:04.631 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_utils.so\n",
            "09:39:04.635 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/content/drive/MyDrive/GenomeAnalysis/tools/gatk-4.2.0.0/gatk-package-4.2.0.0-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so\n",
            "09:39:04.676 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM\n",
            "09:39:04.677 INFO  IntelPairHmm - Available threads: 2\n",
            "09:39:04.678 INFO  IntelPairHmm - Requested threads: 4\n",
            "09:39:04.678 WARN  IntelPairHmm - Using 2 available threads, but 4 were requested\n",
            "09:39:04.679 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation\n",
            "09:39:04.766 INFO  ProgressMeter - Starting traversal\n",
            "09:39:04.776 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
            "09:39:06.474 WARN  InbreedingCoeff - InbreedingCoeff will not be calculated at position 20:10000117 and possibly subsequent; at least 10 samples must have called genotypes\n",
            "09:39:15.127 INFO  ProgressMeter -          20:10005539              0.2                    50            289.8\n",
            "09:39:25.569 INFO  ProgressMeter -          20:10027828              0.3                   190            548.3\n",
            "09:39:35.817 INFO  ProgressMeter -          20:10061186              0.5                   360            695.9\n",
            "09:39:46.458 WARN  DepthPerSampleHC - Annotation will not be calculated at position 20:10088985 and possibly subsequent; genotype for sample NA12877 is not called\n",
            "09:39:46.472 INFO  ProgressMeter -          20:10088858              0.7                   540            777.1\n",
            "09:39:56.924 INFO  ProgressMeter -          20:10112907              0.9                   710            816.9\n",
            "09:40:07.250 INFO  ProgressMeter -          20:10172568              1.0                  1020            979.6\n",
            "09:40:12.702 INFO  HaplotypeCaller - 1176 read(s) filtered by: MappingQualityReadFilter \n",
            "0 read(s) filtered by: MappingQualityAvailableReadFilter \n",
            "0 read(s) filtered by: MappedReadFilter \n",
            "0 read(s) filtered by: NotSecondaryAlignmentReadFilter \n",
            "20525 read(s) filtered by: NotDuplicateReadFilter \n",
            "0 read(s) filtered by: PassesVendorQualityCheckReadFilter \n",
            "0 read(s) filtered by: NonZeroReferenceLengthAlignmentReadFilter \n",
            "0 read(s) filtered by: GoodCigarReadFilter \n",
            "0 read(s) filtered by: WellformedReadFilter \n",
            "21701 total reads filtered\n",
            "09:40:12.702 INFO  ProgressMeter -          20:10199356              1.1                  1172           1035.2\n",
            "09:40:12.702 INFO  ProgressMeter - Traversal complete. Processed 1172 total regions in 1.1 minutes.\n",
            "09:40:12.769 INFO  VectorLoglessPairHMM - Time spent in setup for JNI call : 0.10333578\n",
            "09:40:12.769 INFO  PairHMM - Total compute time in PairHMM computeLogLikelihoods() : 11.904412326000001\n",
            "09:40:12.770 INFO  SmithWatermanAligner - Total compute time in java Smith-Waterman : 7.08 sec\n",
            "09:40:14.251 INFO  HaplotypeCaller - Shutting down engine\n",
            "[November 21, 2024 at 9:40:14 AM UTC] org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller done. Elapsed time: 1.17 minutes.\n",
            "Runtime.totalMemory()=706740224\n"
          ]
        }
      ]
    }
  ]
}